// @theia/task/lib/node/process/process-task-runner.js
"use strict";
/********************************************************************************
 * Copyright (C) 2017-2019 Ericsson and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __values = (this && this.__values) || function (o) {
    var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
    if (m) return m.call(o);
    return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
};
Object.defineProperty(exports, "__esModule", { value: true });
var inversify_1 = require("inversify");
var core_1 = require("@theia/core");
var node_1 = require("@theia/core/lib/node");
var node_2 = require("@theia/process/lib/node");
var process_task_1 = require("./process-task");
var fs = require("fs");
var path = require("path");
/**
 * Task runner that runs a task as a process or a command inside a shell.
 */
var ProcessTaskRunner = /** @class */ (function () {
    function ProcessTaskRunner() {
    }
    /**
     * Runs a task from the given task configuration.
     * @param taskConfig task configuration to run a task from. The provided task configuration must have a shape of `CommandProperties`.
     */
    ProcessTaskRunner.prototype.run = function (taskConfig, ctx) {
        return __awaiter(this, void 0, void 0, function () {
            var command, args, options, cwd, cmd, proc, processType;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!taskConfig.command) {
                            throw new Error("Process task config must have 'command' property specified");
                        }
                        // on windows, prefer windows-specific options, if available
                        if (core_1.isWindows && taskConfig.windows !== undefined) {
                            command = taskConfig.windows.command;
                            args = taskConfig.windows.args;
                            options = taskConfig.windows.options;
                        }
                        else {
                            command = taskConfig.command;
                            args = taskConfig.args;
                            options = taskConfig.options;
                        }
                        // sanity checks:
                        // - we expect the cwd to be set by the client.
                        if (!taskConfig.cwd) {
                            throw new Error("Can't run a task when 'cwd' is not provided by the client");
                        }
                        cwd = this.asFsPath(taskConfig.cwd);
                        // Use task's cwd with spawned process and pass node env object to
                        // new process, so e.g. we can re-use the system path
                        options = {
                            cwd: cwd,
                            env: process.env
                        };
                        return [4 /*yield*/, this.findCommand(command, cwd)];
                    case 1:
                        cmd = _a.sent();
                        if (!cmd) {
                            throw new Error("Command not found: " + command);
                        }
                        try {
                            proc = void 0;
                            processType = taskConfig.type === 'process' ? 'process' : 'shell';
                            if (processType === 'process') {
                                this.logger.debug('Task: creating underlying raw process');
                                proc = this.rawProcessFactory({
                                    command: command,
                                    args: args,
                                    options: options
                                });
                            }
                            else {
                                // all Task types without specific TaskRunner will be run as a shell process e.g.: npm, gulp, etc.
                                this.logger.debug('Task: creating underlying terminal process');
                                proc = this.terminalProcessFactory({
                                    command: command,
                                    args: args,
                                    options: options
                                });
                            }
                            return [2 /*return*/, this.taskFactory({
                                    label: taskConfig.label,
                                    command: cmd,
                                    process: proc,
                                    processType: processType,
                                    context: ctx,
                                    config: taskConfig
                                })];
                        }
                        catch (error) {
                            this.logger.error("Error occurred while creating task: " + error);
                            throw error;
                        }
                        return [2 /*return*/];
                }
            });
        });
    };
    ProcessTaskRunner.prototype.asFsPath = function (uriOrPath) {
        return (uriOrPath.startsWith('file:/'))
            ? node_1.FileUri.fsPath(uriOrPath)
            : uriOrPath;
    };
    /**
     * Uses heuristics to look-for a command. Will look into the system path, if the command
     * is given without a path. Will resolve if a potential match is found, else reject. There
     * is no guarantee that a command we find will be the one executed, if multiple commands with
     * the same name exist.
     * @param command command name to look for
     * @param cwd current working directory (as a fs path, not URI)
     */
    ProcessTaskRunner.prototype.findCommand = function (command, cwd) {
        return __awaiter(this, void 0, void 0, function () {
            var e_1, _a, systemPath, pathDelimiter, resolvedCommand, pathArray, pathArray_1, pathArray_1_1, p, candidate, e_1_1;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        systemPath = process.env.PATH;
                        pathDelimiter = path.delimiter;
                        if (!path.isAbsolute(command)) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.executableFileExists(command)];
                    case 1:
                        if (_b.sent()) {
                            return [2 /*return*/, command];
                        }
                        return [3 /*break*/, 12];
                    case 2:
                        resolvedCommand = node_1.FileUri.fsPath(node_1.FileUri.create(cwd).resolve(command));
                        return [4 /*yield*/, this.executableFileExists(resolvedCommand)];
                    case 3:
                        if (!_b.sent()) return [3 /*break*/, 4];
                        return [2 /*return*/, resolvedCommand];
                    case 4:
                        if (!(path.basename(command) === command)) return [3 /*break*/, 12];
                        if (!(systemPath !== undefined)) return [3 /*break*/, 12];
                        pathArray = systemPath.split(pathDelimiter);
                        _b.label = 5;
                    case 5:
                        _b.trys.push([5, 10, 11, 12]);
                        pathArray_1 = __values(pathArray), pathArray_1_1 = pathArray_1.next();
                        _b.label = 6;
                    case 6:
                        if (!!pathArray_1_1.done) return [3 /*break*/, 9];
                        p = pathArray_1_1.value;
                        candidate = node_1.FileUri.fsPath(node_1.FileUri.create(p).resolve(command));
                        return [4 /*yield*/, this.executableFileExists(candidate)];
                    case 7:
                        if (_b.sent()) {
                            return [2 /*return*/, candidate];
                        }
                        _b.label = 8;
                    case 8:
                        pathArray_1_1 = pathArray_1.next();
                        return [3 /*break*/, 6];
                    case 9: return [3 /*break*/, 12];
                    case 10:
                        e_1_1 = _b.sent();
                        e_1 = { error: e_1_1 };
                        return [3 /*break*/, 12];
                    case 11:
                        try {
                            if (pathArray_1_1 && !pathArray_1_1.done && (_a = pathArray_1.return)) _a.call(pathArray_1);
                        }
                        finally { if (e_1) throw e_1.error; }
                        return [7 /*endfinally*/];
                    case 12: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Checks for the existence of a file, at the provided path, and make sure that
     * it's readable and executable.
     */
    ProcessTaskRunner.prototype.executableFileExists = function (filePath) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                return [2 /*return*/, new Promise(function (resolve, reject) { return __awaiter(_this, void 0, void 0, function () {
                        return __generator(this, function (_a) {
                            fs.access(filePath, fs.constants.F_OK | fs.constants.X_OK, function (err) {
                                resolve(err ? false : true);
                            });
                            return [2 /*return*/];
                        });
                    }); })];
            });
        });
    };
    __decorate([
        inversify_1.inject(core_1.ILogger), inversify_1.named('task'),
        __metadata("design:type", Object)
    ], ProcessTaskRunner.prototype, "logger", void 0);
    __decorate([
        inversify_1.inject(node_2.RawProcessFactory),
        __metadata("design:type", Function)
    ], ProcessTaskRunner.prototype, "rawProcessFactory", void 0);
    __decorate([
        inversify_1.inject(node_2.TerminalProcessFactory),
        __metadata("design:type", Function)
    ], ProcessTaskRunner.prototype, "terminalProcessFactory", void 0);
    __decorate([
        inversify_1.inject(process_task_1.TaskFactory),
        __metadata("design:type", Function)
    ], ProcessTaskRunner.prototype, "taskFactory", void 0);
    ProcessTaskRunner = __decorate([
        inversify_1.injectable()
    ], ProcessTaskRunner);
    return ProcessTaskRunner;
}());
exports.ProcessTaskRunner = ProcessTaskRunner;
