// vscode-ws-jsonrpc/lib/connection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var socket_1 = require("./socket");
var logger_1 = require("./logger");
function listen(options) {
    var webSocket = options.webSocket, onConnection = options.onConnection;
    var logger = options.logger || new logger_1.ConsoleLogger();
    webSocket.onopen = function () {
        var socket = toSocket(webSocket);
        var connection = socket_1.createWebSocketConnection(socket, logger);
        onConnection(connection);
    };
}
exports.listen = listen;
function toSocket(webSocket) {
    return {
        send: function (content) { return webSocket.send(content); },
        onMessage: function (cb) { return webSocket.onmessage = function (event) { return cb(event.data); }; },
        onError: function (cb) { return webSocket.onerror = function (event) {
            if ('message' in event) {
                cb(event.message);
            }
        }; },
        onClose: function (cb) { return webSocket.onclose = function (event) { return cb(event.code, event.reason); }; },
        dispose: function () { return webSocket.close(); }
    };
}
exports.toSocket = toSocket;
