// @theia/extension-manager/lib/common/extension-protocol.js
"use strict";
/********************************************************************************
 * Copyright (C) 2017 TypeFox and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.extensionPath = '/services/extensions';
/**
 * The raw extension information from the repository.
 */
var RawExtension = /** @class */ (function () {
    function RawExtension() {
    }
    return RawExtension;
}());
exports.RawExtension = RawExtension;
/**
 * The detailed extension information from the repository.
 */
var ResolvedRawExtension = /** @class */ (function (_super) {
    __extends(ResolvedRawExtension, _super);
    function ResolvedRawExtension() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return ResolvedRawExtension;
}(RawExtension));
exports.ResolvedRawExtension = ResolvedRawExtension;
/**
 * The extension consists of the raw information and the installation state.
 */
var Extension = /** @class */ (function (_super) {
    __extends(Extension, _super);
    function Extension() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return Extension;
}(RawExtension));
exports.Extension = Extension;
exports.ExtensionServer = Symbol('ExtensionServer');
exports.ExtensionClient = Symbol('ExtensionClient');
