// @theia/go/lib/browser/go-command-contribution.js
"use strict";
/*
 * Copyright (C) 2017 TypeFox and others.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 */
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
var browser_1 = require("@theia/editor/lib/browser");
var inversify_1 = require("inversify");
var go_commands_1 = require("./go-commands");
var browser_2 = require("@theia/core/lib/browser");
exports.SHOW_GO_REFERENCES = {
    id: 'go.editor.action.showReferences'
};
exports.GO_SHOW_COMMANDS = {
    id: 'go.show.commands'
};
var GoCommandContribution = /** @class */ (function () {
    function GoCommandContribution(keybindingRegistry, quickOpenService) {
        this.keybindingRegistry = keybindingRegistry;
        this.quickOpenService = quickOpenService;
    }
    GoCommandContribution.prototype.registerCommands = function (commands) {
        var _this = this;
        commands.registerCommand(exports.SHOW_GO_REFERENCES, {
            execute: function (uri, position, locations) {
                return commands.executeCommand(browser_1.EditorCommands.SHOW_REFERENCES.id, uri, position, locations);
            }
        });
        commands.registerCommand(exports.GO_SHOW_COMMANDS, {
            execute: function () { return _this.showCommands(commands); }
        });
    };
    GoCommandContribution.prototype.showCommands = function (commands) {
        var items = [];
        for (var commandId in go_commands_1.GoCommands.TITLES) {
            var command = commands.getCommand(commandId);
            if (command && command.label)
                items.push(new browser_2.CommandQuickOpenItem(command, commands, this.keybindingRegistry));
        }
        var model = {
            onType: function (lookFor, acceptor) {
                acceptor(items);
            }
        };
        this.quickOpenService.open(model);
    };
    GoCommandContribution = __decorate([
        inversify_1.injectable(),
        __param(0, inversify_1.inject(browser_2.KeybindingRegistry)),
        __param(1, inversify_1.inject(browser_2.QuickOpenService)),
        __metadata("design:paramtypes", [browser_2.KeybindingRegistry,
            browser_2.QuickOpenService])
    ], GoCommandContribution);
    return GoCommandContribution;
}());
exports.GoCommandContribution = GoCommandContribution;
