// @theia/debug/lib/browser/model/debug-breakpoint.js
"use strict";
/********************************************************************************
 * Copyright (C) 2018 TypeFox and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __values = (this && this.__values) || function (o) {
    var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
    if (m) return m.call(o);
    return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var uri_1 = require("@theia/core/lib/common/uri");
var browser_1 = require("@theia/core/lib/browser");
var DebugBreakpointData = /** @class */ (function () {
    function DebugBreakpointData() {
    }
    return DebugBreakpointData;
}());
exports.DebugBreakpointData = DebugBreakpointData;
var DebugBreakpoint = /** @class */ (function (_super) {
    __extends(DebugBreakpoint, _super);
    function DebugBreakpoint(origin, labelProvider, breakpoints, editorManager, session) {
        var _this = _super.call(this) || this;
        _this.labelProvider = labelProvider;
        _this.breakpoints = breakpoints;
        _this.editorManager = editorManager;
        _this.session = session;
        _this.setBreakpointEnabled = function (event) {
            _this.setEnabled(event.target.checked);
        };
        Object.assign(_this, { origins: [origin] });
        _this.uri = new uri_1.default(_this.origins[0].uri);
        return _this;
    }
    DebugBreakpoint.prototype.update = function (data) {
        Object.assign(this, data);
    };
    Object.defineProperty(DebugBreakpoint.prototype, "origin", {
        get: function () {
            return this.origins[0];
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DebugBreakpoint.prototype, "id", {
        get: function () {
            return this.raw && this.raw.id;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DebugBreakpoint.prototype, "enabled", {
        get: function () {
            return this.breakpoints.breakpointsEnabled && this.origin.enabled;
        },
        enumerable: true,
        configurable: true
    });
    DebugBreakpoint.prototype.setEnabled = function (enabled) {
        var e_1, _a;
        var _b = this, uri = _b.uri, raw = _b.raw;
        var shouldUpdate = false;
        var breakpoints = raw && this.doRemove(this.origins.filter(function (origin) { return origin.raw.line !== raw.line; }));
        if (breakpoints) {
            shouldUpdate = true;
        }
        else {
            breakpoints = this.breakpoints.getBreakpoints(uri);
        }
        try {
            for (var breakpoints_1 = __values(breakpoints), breakpoints_1_1 = breakpoints_1.next(); !breakpoints_1_1.done; breakpoints_1_1 = breakpoints_1.next()) {
                var breakpoint = breakpoints_1_1.value;
                if (breakpoint.raw.line === this.origin.raw.line && breakpoint.enabled !== enabled) {
                    breakpoint.enabled = enabled;
                    shouldUpdate = true;
                }
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (breakpoints_1_1 && !breakpoints_1_1.done && (_a = breakpoints_1.return)) _a.call(breakpoints_1);
            }
            finally { if (e_1) throw e_1.error; }
        }
        if (shouldUpdate) {
            this.breakpoints.setBreakpoints(this.uri, breakpoints);
        }
    };
    Object.defineProperty(DebugBreakpoint.prototype, "installed", {
        get: function () {
            return !!this.raw;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DebugBreakpoint.prototype, "verified", {
        get: function () {
            return !!this.raw ? this.raw.verified : true;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DebugBreakpoint.prototype, "message", {
        get: function () {
            return this.raw && this.raw.message || '';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DebugBreakpoint.prototype, "line", {
        /** 1-based */
        get: function () {
            return this.raw && this.raw.line || this.origins[0].raw.line;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DebugBreakpoint.prototype, "column", {
        get: function () {
            return this.raw && this.raw.column || this.origins[0].raw.column;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DebugBreakpoint.prototype, "endLine", {
        get: function () {
            return this.raw && this.raw.endLine;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DebugBreakpoint.prototype, "endColumn", {
        get: function () {
            return this.raw && this.raw.endColumn;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DebugBreakpoint.prototype, "source", {
        get: function () {
            return this.raw && this.raw.source && this.session && this.session.getSource(this.raw.source);
        },
        enumerable: true,
        configurable: true
    });
    DebugBreakpoint.prototype.open = function (options) {
        if (options === void 0) { options = {
            mode: 'reveal'
        }; }
        return __awaiter(this, void 0, void 0, function () {
            var _a, line, column, endLine, endColumn, selection;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _a = this, line = _a.line, column = _a.column, endLine = _a.endLine, endColumn = _a.endColumn;
                        selection = {
                            start: {
                                line: line - 1,
                                character: typeof column === 'number' ? column - 1 : undefined
                            }
                        };
                        if (typeof endLine === 'number') {
                            selection.end = {
                                line: endLine - 1,
                                character: typeof endColumn === 'number' ? endColumn - 1 : undefined
                            };
                        }
                        if (!this.source) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.source.open(__assign({}, options, { selection: selection }))];
                    case 1:
                        _b.sent();
                        return [3 /*break*/, 3];
                    case 2:
                        this.editorManager.open(this.uri, __assign({}, options, { selection: selection }));
                        _b.label = 3;
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    DebugBreakpoint.prototype.render = function () {
        var classNames = ['theia-source-breakpoint'];
        if (!this.breakpoints.breakpointsEnabled || !this.verified) {
            classNames.push(browser_1.DISABLED_CLASS);
        }
        return React.createElement("div", { title: this.message, className: classNames.join(' ') },
            React.createElement("input", { type: 'checkbox', checked: this.origins[0].enabled, onChange: this.setBreakpointEnabled }),
            React.createElement("span", { className: 'name' },
                this.labelProvider.getName(this.uri),
                " "),
            React.createElement("span", { className: 'path' },
                this.labelProvider.getLongName(this.uri.parent),
                " "),
            React.createElement("span", { className: 'line' }, this.line));
    };
    DebugBreakpoint.prototype.remove = function () {
        var breakpoints = this.doRemove(this.origins);
        if (breakpoints) {
            this.breakpoints.setBreakpoints(this.uri, breakpoints);
        }
    };
    DebugBreakpoint.prototype.doRemove = function (origins) {
        if (!origins.length) {
            return undefined;
        }
        var uri = this.uri;
        var toRemove = new Set();
        origins.forEach(function (origin) { return toRemove.add(origin.raw.line); });
        var shouldUpdate = false;
        var breakpoints = this.breakpoints.findMarkers({
            uri: uri,
            dataFilter: function (data) {
                var result = !toRemove.has(data.raw.line);
                shouldUpdate = shouldUpdate || !result;
                return result;
            }
        }).map(function (_a) {
            var data = _a.data;
            return data;
        });
        return shouldUpdate && breakpoints || undefined;
    };
    return DebugBreakpoint;
}(DebugBreakpointData));
exports.DebugBreakpoint = DebugBreakpoint;
