// @theia/extension-manager/lib/browser/extension-contribution.js
"use strict";
/********************************************************************************
 * Copyright (C) 2017 TypeFox and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var inversify_1 = require("inversify");
var core_1 = require("@theia/core");
var common_1 = require("../common");
var view_contribution_1 = require("@theia/core/lib/browser/shell/view-contribution");
exports.EXTENSIONS_WIDGET_FACTORY_ID = 'extensions';
var ExtensionContribution = /** @class */ (function (_super) {
    __extends(ExtensionContribution, _super);
    function ExtensionContribution() {
        return _super.call(this, {
            widgetId: exports.EXTENSIONS_WIDGET_FACTORY_ID,
            widgetName: 'Extensions',
            defaultWidgetOptions: {
                area: 'left',
                rank: 300
            },
            toggleCommandId: 'extensionsView:toggle',
            toggleKeybinding: 'ctrlcmd+shift+x'
        }) || this;
    }
    ExtensionContribution.prototype.onStart = function (app) {
        var _this = this;
        this.extensionManager.onWillStartInstallation(function (_a) {
            var reverting = _a.reverting;
            if (reverting) {
                _this.messageService.error('Failed to install extensions. Reverting...');
            }
            else {
                _this.messageService.info('Installing extensions...');
            }
        });
        this.extensionManager.onDidStopInstallation(function (_a) {
            var reverting = _a.reverting, failed = _a.failed;
            if (!failed) {
                var reloadMessage = !reverting ? 'Reload to complete the installation.' : 'Reload to revert the installation.';
                _this.messageService.info(reloadMessage).then(function () { return window.location.reload(true); });
            }
        });
    };
    __decorate([
        inversify_1.inject(common_1.ExtensionManager),
        __metadata("design:type", common_1.ExtensionManager)
    ], ExtensionContribution.prototype, "extensionManager", void 0);
    __decorate([
        inversify_1.inject(core_1.MessageService),
        __metadata("design:type", core_1.MessageService)
    ], ExtensionContribution.prototype, "messageService", void 0);
    ExtensionContribution = __decorate([
        inversify_1.injectable(),
        __metadata("design:paramtypes", [])
    ], ExtensionContribution);
    return ExtensionContribution;
}(view_contribution_1.AbstractViewContribution));
exports.ExtensionContribution = ExtensionContribution;
