// @theia/plugin-ext/lib/main/node/plugins-key-value-storage.js
"use strict";
/********************************************************************************
 * Copyright (C) 2018 Red Hat, Inc. and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var inversify_1 = require("inversify");
var fs = require("fs");
var path = require("path");
var promise_util_1 = require("@theia/core/lib/common/promise-util");
var common_1 = require("@theia/filesystem/lib/common");
var const_1 = require("./paths/const");
var plugin_paths_protocol_1 = require("../common/plugin-paths-protocol");
var PluginsKeyValueStorage = /** @class */ (function () {
    function PluginsKeyValueStorage(pluginPathsService, fileSystem) {
        var _this = this;
        this.pluginPathsService = pluginPathsService;
        this.fileSystem = fileSystem;
        this.deferredTheiaDirPath = new promise_util_1.Deferred();
        this.pluginPathsService.getTheiaDirPath().then(function (theiaDirPathUri) {
            return _this.fileSystem.getFsPath(theiaDirPathUri);
        }).then(function (theiaDirPath) {
            _this.theiaDirPath = theiaDirPath;
            _this.globalDataPath = path.join(_this.theiaDirPath, const_1.PluginPaths.PLUGINS_GLOBAL_STORAGE_DIR, 'global-state.json');
            if (!fs.existsSync(path.dirname(_this.globalDataPath))) {
                _this.createDirectories(path.dirname(_this.globalDataPath));
            }
            _this.deferredTheiaDirPath.resolve(_this.theiaDirPath);
        });
    }
    PluginsKeyValueStorage.prototype.createDirectories = function (pathString) {
        var toCreate = [];
        while (!fs.existsSync(pathString)) {
            toCreate.push(path.basename(pathString));
            pathString = path.dirname(pathString);
        }
        while (toCreate.length > 0) {
            pathString = path.join(pathString, toCreate.pop());
            fs.mkdirSync(pathString);
        }
    };
    PluginsKeyValueStorage.prototype.set = function (key, value, isGlobal) {
        return __awaiter(this, void 0, void 0, function () {
            var dataPath, data;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.getDataPath(isGlobal)];
                    case 1:
                        dataPath = _a.sent();
                        if (!dataPath) {
                            return [2 /*return*/, Promise.reject('Cannot save data: no opened workspace')];
                        }
                        data = this.readFromFile(dataPath);
                        if (value === undefined || value === {}) {
                            delete data[key];
                        }
                        else {
                            data[key] = value;
                        }
                        this.writeToFile(dataPath, data);
                        return [2 /*return*/, Promise.resolve(true)];
                }
            });
        });
    };
    PluginsKeyValueStorage.prototype.get = function (key, isGlobal) {
        return __awaiter(this, void 0, void 0, function () {
            var dataPath, data;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.getDataPath(isGlobal)];
                    case 1:
                        dataPath = _a.sent();
                        if (!dataPath) {
                            return [2 /*return*/, Promise.resolve({})];
                        }
                        data = this.readFromFile(dataPath);
                        return [2 /*return*/, Promise.resolve(data[key])];
                }
            });
        });
    };
    PluginsKeyValueStorage.prototype.getAll = function (isGlobal) {
        return __awaiter(this, void 0, void 0, function () {
            var dataPath, data;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.getDataPath(isGlobal)];
                    case 1:
                        dataPath = _a.sent();
                        if (!dataPath) {
                            return [2 /*return*/, Promise.resolve({})];
                        }
                        data = this.readFromFile(dataPath);
                        return [2 /*return*/, Promise.resolve(data)];
                }
            });
        });
    };
    PluginsKeyValueStorage.prototype.getDataPath = function (isGlobal) {
        return __awaiter(this, void 0, void 0, function () {
            var storagePath;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!(this.theiaDirPath === undefined)) return [3 /*break*/, 2];
                        // wait for Theia data directory path if it hasn't been initialized yet
                        return [4 /*yield*/, this.deferredTheiaDirPath.promise];
                    case 1:
                        // wait for Theia data directory path if it hasn't been initialized yet
                        _a.sent();
                        _a.label = 2;
                    case 2:
                        if (!isGlobal) return [3 /*break*/, 3];
                        return [2 /*return*/, this.globalDataPath];
                    case 3: return [4 /*yield*/, this.pluginPathsService.getLastStoragePath()];
                    case 4:
                        storagePath = _a.sent();
                        return [2 /*return*/, storagePath ? path.join(storagePath, 'workspace-state.json') : undefined];
                }
            });
        });
    };
    PluginsKeyValueStorage.prototype.readFromFile = function (pathToFile) {
        if (!fs.existsSync(pathToFile)) {
            return {};
        }
        var rawData = fs.readFileSync(pathToFile, 'utf8');
        try {
            return JSON.parse(rawData);
        }
        catch (error) {
            console.error('Failed to parse data from "', pathToFile, '". Reason:', error);
            return {};
        }
    };
    PluginsKeyValueStorage.prototype.writeToFile = function (pathToFile, data) {
        if (!fs.existsSync(path.dirname(pathToFile))) {
            fs.mkdirSync(path.dirname(pathToFile));
        }
        var rawData = JSON.stringify(data);
        fs.writeFileSync(pathToFile, rawData, 'utf8');
    };
    PluginsKeyValueStorage = __decorate([
        inversify_1.injectable(),
        __param(0, inversify_1.inject(plugin_paths_protocol_1.PluginPathsService)),
        __param(1, inversify_1.inject(common_1.FileSystem)),
        __metadata("design:paramtypes", [Object, Object])
    ], PluginsKeyValueStorage);
    return PluginsKeyValueStorage;
}());
exports.PluginsKeyValueStorage = PluginsKeyValueStorage;
