// vscode-base-languageclient/lib/foldingRange.js
/* --------------------------------------------------------------------------------------------
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for license information.
 * ------------------------------------------------------------------------------------------ */
'use strict';
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var UUID = require("./utils/uuid");
var Is = require("./utils/is");
var vscode_1 = require("vscode");
var vscode_languageserver_protocol_1 = require("vscode-languageserver-protocol");
var client_1 = require("./client");
function ensure(target, key) {
    if (target[key] === void 0) {
        target[key] = {};
    }
    return target[key];
}
var FoldingRangeFeature = /** @class */ (function (_super) {
    __extends(FoldingRangeFeature, _super);
    function FoldingRangeFeature(client) {
        return _super.call(this, client, vscode_languageserver_protocol_1.FoldingRangeRequest.type) || this;
    }
    FoldingRangeFeature.prototype.fillClientCapabilities = function (capabilites) {
        var capability = ensure(ensure(capabilites, 'textDocument'), 'foldingRange');
        capability.dynamicRegistration = true;
        capability.rangeLimit = 5000;
        capability.lineFoldingOnly = true;
    };
    FoldingRangeFeature.prototype.initialize = function (capabilities, documentSelector) {
        if (!capabilities.foldingRangeProvider) {
            return;
        }
        var implCapabilities = capabilities.foldingRangeProvider;
        var id = Is.string(implCapabilities.id) && implCapabilities.id.length > 0 ? implCapabilities.id : UUID.generateUuid();
        var selector = implCapabilities.documentSelector || documentSelector;
        if (selector) {
            this.register(this.messages, {
                id: id,
                registerOptions: Object.assign({}, { documentSelector: selector })
            });
        }
    };
    FoldingRangeFeature.prototype.registerLanguageProvider = function (options) {
        var _this = this;
        var client = this._client;
        var provideFoldingRanges = function (document, _, token) {
            var requestParams = {
                textDocument: client.code2ProtocolConverter.asTextDocumentIdentifier(document)
            };
            return client.sendRequest(vscode_languageserver_protocol_1.FoldingRangeRequest.type, requestParams, token).then(_this.asFoldingRanges.bind(_this), function (error) {
                client.logFailedRequest(vscode_languageserver_protocol_1.FoldingRangeRequest.type, error);
                return Promise.resolve(null);
            });
        };
        var middleware = client.clientOptions.middleware;
        return vscode_1.languages.registerFoldingRangeProvider(options.documentSelector, {
            provideFoldingRanges: function (document, context, token) {
                return middleware.provideFoldingRanges
                    ? middleware.provideFoldingRanges(document, context, token, provideFoldingRanges)
                    : provideFoldingRanges(document, context, token);
            }
        });
    };
    FoldingRangeFeature.prototype.asFoldingRangeKind = function (kind) {
        if (kind) {
            switch (kind) {
                case vscode_languageserver_protocol_1.FoldingRangeKind.Comment:
                    return vscode_1.FoldingRangeKind.Comment;
                case vscode_languageserver_protocol_1.FoldingRangeKind.Imports:
                    return vscode_1.FoldingRangeKind.Imports;
                case vscode_languageserver_protocol_1.FoldingRangeKind.Region:
                    return vscode_1.FoldingRangeKind.Region;
            }
        }
        return void 0;
    };
    FoldingRangeFeature.prototype.asFoldingRanges = function (foldingRanges) {
        var _this = this;
        if (Array.isArray(foldingRanges)) {
            return foldingRanges.map(function (r) {
                return new vscode_1.FoldingRange(r.startLine, r.endLine, _this.asFoldingRangeKind(r.kind));
            });
        }
        return [];
    };
    return FoldingRangeFeature;
}(client_1.TextDocumentFeature));
exports.FoldingRangeFeature = FoldingRangeFeature;
