// @theia/core/lib/browser/widgets/virtual-widget.js
"use strict";
/********************************************************************************
 * Copyright (C) 2017 TypeFox and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
var inversify_1 = require("inversify");
var common_1 = require("../../common");
var widget_1 = require("./widget");
var virtual_renderer_1 = require("./virtual-renderer");
/*
 * @deprecated use ReactWidget instead. VirtualWidget will be removed with the next major release.
 */
var VirtualWidget = /** @class */ (function (_super) {
    __extends(VirtualWidget, _super);
    function VirtualWidget() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.onRender = new common_1.DisposableCollection();
        _this.scrollOptions = {
            suppressScrollX: true
        };
        return _this;
    }
    VirtualWidget.prototype.onUpdateRequest = function (msg) {
        _super.prototype.onUpdateRequest.call(this, msg);
        var child = this.render();
        if (!this.childContainer) {
            // if we are adding scrolling, we need to wrap the contents in its own div, to not conflict with the virtual dom algo.
            if (this.scrollOptions) {
                this.childContainer = this.createChildContainer();
                this.node.appendChild(this.childContainer);
            }
            else {
                this.childContainer = this.node;
            }
        }
        virtual_renderer_1.VirtualRenderer.render(child, this.childContainer);
        this.onRender.dispose();
    };
    VirtualWidget.prototype.render = function () {
        // tslint:disable-next-line:no-null-keyword
        return null;
    };
    VirtualWidget.prototype.createChildContainer = function () {
        return document.createElement('div');
    };
    VirtualWidget = __decorate([
        inversify_1.injectable()
    ], VirtualWidget);
    return VirtualWidget;
}(widget_1.BaseWidget));
exports.VirtualWidget = VirtualWidget;
