// @theia/core/lib/browser/widgets/virtual-renderer.js
"use strict";
/********************************************************************************
 * Copyright (C) 2017 TypeFox and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spread = (this && this.__spread) || function () {
    for (var ar = [], i = 0; i < arguments.length; i++) ar = ar.concat(__read(arguments[i]));
    return ar;
};
var __values = (this && this.__values) || function (o) {
    var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
    if (m) return m.call(o);
    return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
};
Object.defineProperty(exports, "__esModule", { value: true });
var virtualdom_1 = require("@phosphor/virtualdom");
// Phosphor elements can be null, so we need to disable our no-null rule
// tslint:disable:no-null-keyword
/*
 * @deprecated use ReactRenderer instead. VirtualRenderer will be removed with the next major release.
 */
var VirtualRenderer = /** @class */ (function () {
    function VirtualRenderer(host) {
        this.host = host || document.createElement('div');
    }
    VirtualRenderer.prototype.render = function () {
        VirtualRenderer.render(this.doRender(), this.host);
    };
    VirtualRenderer.prototype.doRender = function () {
        return null;
    };
    return VirtualRenderer;
}());
exports.VirtualRenderer = VirtualRenderer;
(function (VirtualRenderer) {
    /*
     * @deprecated use ReactDOM.render instead. VirtualRenderer will be removed with the next major release.
     */
    function render(child, host) {
        var content = toContent(child);
        virtualdom_1.VirtualDOM.render(content, host);
    }
    VirtualRenderer.render = render;
    function flatten(children) {
        return children.reduce(function (prev, current) { return merge(prev, current); }, null);
    }
    VirtualRenderer.flatten = flatten;
    function merge(left, right) {
        if (!right) {
            return left || null;
        }
        if (!left) {
            return right;
        }
        var result = left instanceof Array ? left : [left];
        if (right instanceof Array) {
            result.push.apply(result, __spread(right));
        }
        else {
            result.push(right);
        }
        return result;
    }
    VirtualRenderer.merge = merge;
    function toContent(children) {
        var e_1, _a;
        if (!children) {
            return null;
        }
        if (typeof children === 'string') {
            return new virtualdom_1.VirtualText(children);
        }
        if (children instanceof Array) {
            var nodes = [];
            try {
                for (var children_1 = __values(children), children_1_1 = children_1.next(); !children_1_1.done; children_1_1 = children_1.next()) {
                    var child = children_1_1.value;
                    if (child) {
                        var node = typeof child === 'string' ? new virtualdom_1.VirtualText(child) : child;
                        nodes.push(node);
                    }
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (children_1_1 && !children_1_1.done && (_a = children_1.return)) _a.call(children_1);
                }
                finally { if (e_1) throw e_1.error; }
            }
            return nodes;
        }
        return children;
    }
    VirtualRenderer.toContent = toContent;
})(VirtualRenderer = exports.VirtualRenderer || (exports.VirtualRenderer = {}));
exports.VirtualRenderer = VirtualRenderer;
