// @theia/core/lib/browser/shell/shell-layout-restorer.js
"use strict";
/********************************************************************************
 * Copyright (C) 2017 TypeFox and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __values = (this && this.__values) || function (o) {
    var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
    if (m) return m.call(o);
    return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
};
Object.defineProperty(exports, "__esModule", { value: true });
var inversify_1 = require("inversify");
var widget_manager_1 = require("../widget-manager");
var storage_service_1 = require("../storage-service");
var logger_1 = require("../../common/logger");
var theming_1 = require("../theming");
var StatefulWidget;
(function (StatefulWidget) {
    // tslint:disable-next-line:no-any
    function is(arg) {
        return arg !== undefined && typeof arg['storeState'] === 'function' && typeof arg['restoreState'] === 'function';
    }
    StatefulWidget.is = is;
})(StatefulWidget = exports.StatefulWidget || (exports.StatefulWidget = {}));
var ShellLayoutRestorer = /** @class */ (function () {
    function ShellLayoutRestorer(widgetManager, logger, storageService) {
        this.widgetManager = widgetManager;
        this.logger = logger;
        this.storageService = storageService;
        this.storageKey = 'layout';
        this.shouldStoreLayout = true;
    }
    ShellLayoutRestorer.prototype.registerCommands = function (commands) {
        var _this = this;
        commands.registerCommand({
            id: 'reset.layout',
            category: 'View',
            label: 'Reset Workbench Layout'
        }, {
            execute: function () { return __awaiter(_this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    this.shouldStoreLayout = false;
                    this.storageService.setData(this.storageKey, undefined);
                    theming_1.ThemeService.get().reset(); // Theme service cannot use DI, so the current theme ID is stored elsewhere. Hence the explicit reset.
                    window.location.reload(true);
                    return [2 /*return*/];
                });
            }); }
        });
    };
    ShellLayoutRestorer.prototype.storeLayout = function (app) {
        if (this.shouldStoreLayout) {
            try {
                var layoutData = app.shell.getLayoutData();
                var serializedLayoutData = this.deflate(layoutData);
                this.storageService.setData(this.storageKey, serializedLayoutData);
            }
            catch (error) {
                this.storageService.setData(this.storageKey, undefined);
                this.logger.error('Error during serialization of layout data', error);
            }
        }
    };
    ShellLayoutRestorer.prototype.restoreLayout = function (app) {
        return __awaiter(this, void 0, void 0, function () {
            var serializedLayoutData, layoutData;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.storageService.getData(this.storageKey)];
                    case 1:
                        serializedLayoutData = _a.sent();
                        if (serializedLayoutData === undefined) {
                            return [2 /*return*/, false];
                        }
                        return [4 /*yield*/, this.inflate(serializedLayoutData)];
                    case 2:
                        layoutData = _a.sent();
                        return [4 /*yield*/, app.shell.setLayoutData(layoutData)];
                    case 3:
                        _a.sent();
                        return [2 /*return*/, true];
                }
            });
        });
    };
    ShellLayoutRestorer.prototype.isWidgetProperty = function (propertyName) {
        return propertyName === 'widget';
    };
    ShellLayoutRestorer.prototype.isWidgetsProperty = function (propertyName) {
        return propertyName === 'widgets';
    };
    /**
     * Turns the layout data to a string representation.
     */
    ShellLayoutRestorer.prototype.deflate = function (data) {
        var _this = this;
        return JSON.stringify(data, function (property, value) {
            var e_1, _a;
            if (_this.isWidgetProperty(property)) {
                var description = _this.convertToDescription(value);
                return description;
            }
            else if (_this.isWidgetsProperty(property)) {
                var descriptions = [];
                try {
                    for (var _b = __values(value), _c = _b.next(); !_c.done; _c = _b.next()) {
                        var widget = _c.value;
                        var description = _this.convertToDescription(widget);
                        if (description) {
                            descriptions.push(description);
                        }
                    }
                }
                catch (e_1_1) { e_1 = { error: e_1_1 }; }
                finally {
                    try {
                        if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                    }
                    finally { if (e_1) throw e_1.error; }
                }
                return descriptions;
            }
            return value;
        });
    };
    ShellLayoutRestorer.prototype.convertToDescription = function (widget) {
        var desc = this.widgetManager.getDescription(widget);
        if (desc) {
            var innerState = undefined;
            if (StatefulWidget.is(widget)) {
                innerState = widget.storeState();
            }
            return {
                constructionOptions: desc,
                innerWidgetState: innerState
            };
        }
    };
    /**
     * Creates the layout data from its string representation.
     */
    ShellLayoutRestorer.prototype.inflate = function (layoutData) {
        var _this = this;
        var pending = [];
        var result = JSON.parse(layoutData, function (property, value) {
            if (_this.isWidgetsProperty(property)) {
                var widgets_1 = [];
                var descs = value;
                var _loop_1 = function (i) {
                    var promise = _this.convertToWidget(descs[i]);
                    pending.push(promise.then(function (widget) {
                        widgets_1[i] = widget;
                        return widget;
                    }));
                };
                for (var i = 0; i < descs.length; i++) {
                    _loop_1(i);
                }
                return widgets_1;
            }
            else if (value && typeof value === 'object' && !Array.isArray(value)) {
                // tslint:disable-next-line:no-any
                var copy_1 = {};
                var _loop_2 = function (p) {
                    if (_this.isWidgetProperty(p)) {
                        var promise = _this.convertToWidget(value[p]);
                        pending.push(promise.then(function (widget) {
                            copy_1[p] = widget;
                            return widget;
                        }));
                    }
                    else {
                        copy_1[p] = value[p];
                    }
                };
                for (var p in value) {
                    _loop_2(p);
                }
                return copy_1;
            }
            return value;
        });
        return Promise.all(pending).then(function () { return result; });
    };
    ShellLayoutRestorer.prototype.convertToWidget = function (desc) {
        var _this = this;
        if (desc.constructionOptions) {
            return this.widgetManager.getOrCreateWidget(desc.constructionOptions.factoryId, desc.constructionOptions.options)
                .then(function (widget) {
                if (StatefulWidget.is(widget) && desc.innerWidgetState !== undefined) {
                    try {
                        widget.restoreState(desc.innerWidgetState);
                    }
                    catch (err) {
                        _this.logger.warn("Couldn't restore widget state for " + widget.id + ". Error: " + err + " ");
                    }
                }
                return widget;
            }, function (err) {
                _this.logger.warn("Couldn't restore widget for " + desc.constructionOptions.factoryId + ". Error: " + err + " ");
                return undefined;
            });
        }
        else {
            return Promise.resolve(undefined);
        }
    };
    ShellLayoutRestorer = __decorate([
        inversify_1.injectable(),
        __param(0, inversify_1.inject(widget_manager_1.WidgetManager)),
        __param(1, inversify_1.inject(logger_1.ILogger)),
        __param(2, inversify_1.inject(storage_service_1.StorageService)),
        __metadata("design:paramtypes", [widget_manager_1.WidgetManager, Object, Object])
    ], ShellLayoutRestorer);
    return ShellLayoutRestorer;
}());
exports.ShellLayoutRestorer = ShellLayoutRestorer;
