// @theia/plugin-ext/lib/plugin/env.js
"use strict";
/********************************************************************************
 * Copyright (C) 2018 Red Hat, Inc. and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
Object.defineProperty(exports, "__esModule", { value: true });
var plugin_api_1 = require("../api/plugin-api");
var EnvExtImpl = /** @class */ (function () {
    function EnvExtImpl(rpc) {
        this.proxy = rpc.getProxy(plugin_api_1.PLUGIN_RPC_CONTEXT.ENV_MAIN);
    }
    EnvExtImpl.prototype.getEnvVariable = function (envVarName) {
        return this.proxy.$getEnvVariable(envVarName).then(function (x) {
            if (x === null) {
                return undefined;
            }
            return x;
        });
    };
    EnvExtImpl.prototype.getQueryParameter = function (queryParamName) {
        return this.queryParameters[queryParamName];
    };
    EnvExtImpl.prototype.getQueryParameters = function () {
        return this.queryParameters;
    };
    EnvExtImpl.prototype.setQueryParameters = function (queryParams) {
        this.queryParameters = queryParams;
    };
    return EnvExtImpl;
}());
exports.EnvExtImpl = EnvExtImpl;
