// @theia/plugin-ext/lib/plugin/window-state.js
"use strict";
/********************************************************************************
 * Copyright (C) 2018 Red Hat, Inc. and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
Object.defineProperty(exports, "__esModule", { value: true });
var event_1 = require("@theia/core/lib/common/event");
var WindowStateExtImpl = /** @class */ (function () {
    function WindowStateExtImpl() {
        this.windowStateChangedEmitter = new event_1.Emitter();
        this.onDidChangeWindowState = this.windowStateChangedEmitter.event;
        this.windowStateCached = { focused: true }; // supposed tab is active on start
    }
    WindowStateExtImpl.prototype.getWindowState = function () {
        return this.windowStateCached;
    };
    WindowStateExtImpl.prototype.$onWindowStateChanged = function (focused) {
        var state = { focused: focused };
        if (state === this.windowStateCached) {
            return;
        }
        this.windowStateCached = state;
        this.windowStateChangedEmitter.fire(state);
    };
    return WindowStateExtImpl;
}());
exports.WindowStateExtImpl = WindowStateExtImpl;
