// @theia/typescript/lib/common/index.js
"use strict";
/********************************************************************************
 * Copyright (C) 2017 TypeFox and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
Object.defineProperty(exports, "__esModule", { value: true });
exports.TYPESCRIPT_LANGUAGE_ID = 'typescript';
exports.TYPESCRIPT_LANGUAGE_NAME = 'TypeScript';
exports.TYPESCRIPT_REACT_LANGUAGE_ID = 'typescriptreact';
exports.TYPESCRIPT_REACT_LANGUAGE_NAME = 'TypeScript React';
exports.JAVASCRIPT_LANGUAGE_ID = 'javascript';
exports.JAVASCRIPT_LANGUAGE_NAME = 'JavaScript';
exports.JAVASCRIPT_REACT_LANGUAGE_ID = 'javascriptreact';
exports.JAVASCRIPT_REACT_LANGUAGE_NAME = 'JavaScript React';
exports.TS_JS_LANGUAGES = new Set([
    exports.TYPESCRIPT_LANGUAGE_ID,
    exports.TYPESCRIPT_REACT_LANGUAGE_ID,
    exports.JAVASCRIPT_LANGUAGE_ID,
    exports.JAVASCRIPT_REACT_LANGUAGE_ID
]);
