// yang-sprotty/lib/views.js
"use strict";
/*
 * Copyright (C) 2017 TypeFox and others.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
/** @jsx svg */
var snabbdom_jsx_1 = require("snabbdom-jsx");
var lib_1 = require("sprotty/lib");
var ClassNodeView = /** @class */ (function () {
    function ClassNodeView() {
    }
    ClassNodeView.prototype.render = function (node, context) {
        var vnode = snabbdom_jsx_1.svg("g", { "class-sprotty-node": true },
            snabbdom_jsx_1.svg("rect", { "class-selected": node.selected, "class-mouseover": node.hoverFeedback, x: 0, y: 0, width: Math.max(0, node.bounds.width), height: Math.max(0, node.bounds.height) }),
            context.renderChildren(node));
        lib_1.setAttr(vnode, 'class', node.cssClass);
        return vnode;
    };
    return ClassNodeView;
}());
exports.ClassNodeView = ClassNodeView;
var HeaderCompartmentView = /** @class */ (function () {
    function HeaderCompartmentView() {
    }
    HeaderCompartmentView.prototype.render = function (model, context) {
        var translate = "translate(" + model.bounds.x + ", " + model.bounds.y + ")";
        var parentSize = model.parent.size;
        var width = Math.max(0, parentSize.width);
        var height = Math.max(0, model.size.height);
        var vnode = snabbdom_jsx_1.svg("g", { transform: translate, "class-comp": "{true}" },
            snabbdom_jsx_1.svg("rect", { "class-classHeader": true, x: 0, y: 0, width: width, height: height }),
            context.renderChildren(model));
        return vnode;
    };
    return HeaderCompartmentView;
}());
exports.HeaderCompartmentView = HeaderCompartmentView;
var TagView = /** @class */ (function () {
    function TagView() {
    }
    TagView.prototype.render = function (element, context) {
        var radius = 0.5 * element.size.width;
        return snabbdom_jsx_1.svg("g", null,
            snabbdom_jsx_1.svg("circle", { "class-tag": true, r: radius, cx: radius, cy: radius }),
            context.renderChildren(element));
    };
    return TagView;
}());
exports.TagView = TagView;
var ModuleNodeView = /** @class */ (function () {
    function ModuleNodeView() {
    }
    ModuleNodeView.prototype.render = function (node, context) {
        return snabbdom_jsx_1.svg("g", { "class-sprotty-node": true, "class-module": true, "class-mouseover": node.hoverFeedback },
            snabbdom_jsx_1.svg("rect", { "class-body": true, "class-selected": node.selected, x: 0, y: 0, rx: "5", ry: "5", width: Math.max(0, node.bounds.width), height: Math.max(0, node.bounds.height) }),
            context.renderChildren(node));
    };
    return ModuleNodeView;
}());
exports.ModuleNodeView = ModuleNodeView;
var ChoiceNodeView = /** @class */ (function () {
    function ChoiceNodeView() {
    }
    ChoiceNodeView.prototype.render = function (model, context) {
        var width = Math.max(0, model.size.width * 0.5);
        var height = Math.max(0, model.size.height * 0.5);
        var rhombStr = "M 0," + height + " l " + width + "," + height + " l " + width + ",-" + height + " l -" + width + ",-" + height + "z";
        return snabbdom_jsx_1.svg("g", { "class-sprotty-node": "{true}", "class-choice": true },
            snabbdom_jsx_1.svg("path", { d: rhombStr, "class-choice": true }),
            context.renderChildren(model));
    };
    return ChoiceNodeView;
}());
exports.ChoiceNodeView = ChoiceNodeView;
var CaseNodeView = /** @class */ (function () {
    function CaseNodeView() {
    }
    CaseNodeView.prototype.render = function (node, context) {
        var vnode = snabbdom_jsx_1.svg("g", { "class-sprotty-node": "{true}" },
            snabbdom_jsx_1.svg("rect", { "class-body": true, "class-selected": node.selected, x: 0, y: 0, width: Math.max(0, node.bounds.width), height: Math.max(0, node.bounds.height), rx: Math.max(node.size.width * 0.5, 0), ry: 10 }),
            context.renderChildren(node));
        lib_1.setAttr(vnode, 'class', 'case');
        return vnode;
    };
    return CaseNodeView;
}());
exports.CaseNodeView = CaseNodeView;
var UsesNodeView = /** @class */ (function (_super) {
    __extends(UsesNodeView, _super);
    function UsesNodeView() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    UsesNodeView.prototype.render = function (node, context) {
        var vnode = snabbdom_jsx_1.svg("g", { "class-sprotty-node": true },
            snabbdom_jsx_1.svg("rect", { "class-body": true, "class-selected": node.selected, x: 0, y: 0, width: Math.max(0, node.bounds.width), height: Math.max(0, node.bounds.height), rx: Math.max(node.size.height * 0.5, 0), ry: Math.max(node.size.height * 0.5, 0) }),
            context.renderChildren(node));
        lib_1.setAttr(vnode, 'class', node.cssClass);
        return vnode;
    };
    return UsesNodeView;
}(CaseNodeView));
exports.UsesNodeView = UsesNodeView;
var NoteView = /** @class */ (function () {
    function NoteView() {
    }
    NoteView.prototype.render = function (node, context) {
        return snabbdom_jsx_1.svg("g", { "class-note": true, "class-mouseover": node.hoverFeedback },
            snabbdom_jsx_1.svg("path", { "class-front": true, d: "M 0,0 l 15,0 l 0,10 l 10,0 l 0,25 l -25,0 Z", fill: "#FFEB8A" }),
            snabbdom_jsx_1.svg("path", { "class-noteEdge": true, d: "M 15,0 l 0,10 l 10,0 Z", fill: "#FFCC40" }));
    };
    return NoteView;
}());
exports.NoteView = NoteView;
var CompositionEdgeView = /** @class */ (function (_super) {
    __extends(CompositionEdgeView, _super);
    function CompositionEdgeView() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    CompositionEdgeView.prototype.renderAdditionals = function (edge, segments, context) {
        var p1 = segments[0];
        var p2 = segments[1];
        var r = 6;
        var rhombStr = "M 0,0 l" + r + "," + (r / 2) + " l" + r + ",-" + (r / 2) + " l-" + r + ",-" + (r / 2) + " l-" + r + "," + (r / 2) + " Z";
        return [
            snabbdom_jsx_1.svg("path", { "class-sprotty-edge": true, "class-composition": true, d: rhombStr, transform: "rotate(" + angle(p1, p2) + " " + p1.x + " " + p1.y + ") translate(" + p1.x + " " + p1.y + ")" })
        ];
    };
    CompositionEdgeView.prototype.getSourceAnchorCorrection = function (edge) {
        return CompositionEdgeView.SOURCE_CORRECTION;
    };
    CompositionEdgeView.SOURCE_CORRECTION = Math.sqrt(1 * 1 + 2 * 2);
    return CompositionEdgeView;
}(lib_1.PolylineEdgeView));
exports.CompositionEdgeView = CompositionEdgeView;
var DashedEdgeView = /** @class */ (function (_super) {
    __extends(DashedEdgeView, _super);
    function DashedEdgeView() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    DashedEdgeView.prototype.renderLine = function (edge, segments, context) {
        var firstPoint = segments[0];
        var path = "M " + firstPoint.x + "," + firstPoint.y;
        for (var i = 1; i < segments.length; i++) {
            var p = segments[i];
            path += " L " + p.x + "," + p.y;
        }
        return snabbdom_jsx_1.svg("path", { "class-sprotty-edge": true, "class-dashed": true, d: path });
    };
    return DashedEdgeView;
}(lib_1.PolylineEdgeView));
exports.DashedEdgeView = DashedEdgeView;
var ImportEdgeView = /** @class */ (function (_super) {
    __extends(ImportEdgeView, _super);
    function ImportEdgeView() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ImportEdgeView.prototype.renderAdditionals = function (edge, segments, context) {
        var p1 = segments[0];
        var p2 = segments[1];
        return [
            snabbdom_jsx_1.svg("path", { "class-sprotty-edge": true, d: "M 10,-4 L 0,0 L 10,4", transform: "rotate(" + angle(p1, p2) + " " + p1.x + " " + p1.y + ") translate(" + p1.x + " " + p1.y + ")" })
        ];
    };
    ImportEdgeView.prototype.getSourceAnchorCorrection = function (edge) {
        return CompositionEdgeView.SOURCE_CORRECTION;
    };
    ImportEdgeView.SOURCE_CORRECTION = Math.sqrt(1 * 1 + 2.5 * 2.5);
    return ImportEdgeView;
}(DashedEdgeView));
exports.ImportEdgeView = ImportEdgeView;
var ArrowEdgeView = /** @class */ (function (_super) {
    __extends(ArrowEdgeView, _super);
    function ArrowEdgeView() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ArrowEdgeView.prototype.renderAdditionals = function (edge, segments, context) {
        var p1 = segments[segments.length - 2];
        var p2 = segments[segments.length - 1];
        return [
            snabbdom_jsx_1.svg("path", { "class-sprotty-edge": true, d: "M 10,-4 L 0,0 L 10,4", transform: "rotate(" + angle(p2, p1) + " " + p2.x + " " + p2.y + ") translate(" + p2.x + " " + p2.y + ")" })
        ];
    };
    ArrowEdgeView.prototype.getTargetAnchorCorrection = function (edge) {
        return ArrowEdgeView.TARGET_CORRECTION;
    };
    ArrowEdgeView.TARGET_CORRECTION = Math.sqrt(1 * 1 + 2.5 * 2.5);
    return ArrowEdgeView;
}(lib_1.PolylineEdgeView));
exports.ArrowEdgeView = ArrowEdgeView;
var DashedArrowEdgeView = /** @class */ (function (_super) {
    __extends(DashedArrowEdgeView, _super);
    function DashedArrowEdgeView() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    DashedArrowEdgeView.prototype.renderAdditionals = function (edge, segments, context) {
        var p1 = segments[segments.length - 2];
        var p2 = segments[segments.length - 1];
        return [
            snabbdom_jsx_1.svg("path", { "class-sprotty-edge": true, d: "M 10,-4 L 0,0 L 10,4", transform: "rotate(" + angle(p2, p1) + " " + p2.x + " " + p2.y + ") translate(" + p2.x + " " + p2.y + ")" })
        ];
    };
    DashedArrowEdgeView.prototype.getTargetAnchorCorrection = function (edge) {
        return DashedArrowEdgeView.TARGET_CORRECTION;
    };
    DashedArrowEdgeView.TARGET_CORRECTION = Math.sqrt(1 * 1 + 2.5 * 2.5);
    return DashedArrowEdgeView;
}(DashedEdgeView));
exports.DashedArrowEdgeView = DashedArrowEdgeView;
function angle(x0, x1) {
    return lib_1.toDegrees(Math.atan2(x1.y - x0.y, x1.x - x0.x));
}
exports.angle = angle;
