// @theia/core/lib/browser/keybinding.spec.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spread = (this && this.__spread) || function () {
    for (var ar = [], i = 0; i < arguments.length; i++) ar = ar.concat(__read(arguments[i]));
    return ar;
};
var _this = this;
Object.defineProperty(exports, "__esModule", { value: true });
/********************************************************************************
 * Copyright (C) 2017 Ericsson and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
var jsdom_1 = require("../browser/test/jsdom");
var disableJSDOM = jsdom_1.enableJSDOM();
var inversify_1 = require("inversify");
var contribution_provider_1 = require("../common/contribution-provider");
var logger_1 = require("../common/logger");
var keybinding_1 = require("./keybinding");
var keys_1 = require("./keys");
var command_1 = require("../common/command");
var label_parser_1 = require("./label-parser");
var mock_logger_1 = require("../common/test/mock-logger");
var status_bar_1 = require("./status-bar/status-bar");
var frontend_application_state_1 = require("./frontend-application-state");
var os = require("../common/os");
var chai = require("chai");
var sinon = require("sinon");
disableJSDOM();
/* tslint:disable:no-unused-expression */
var expect = chai.expect;
var keybindingRegistry;
var commandRegistry;
var testContainer;
before(function () { return __awaiter(_this, void 0, void 0, function () {
    var module;
    return __generator(this, function (_a) {
        testContainer = new inversify_1.Container();
        module = new inversify_1.ContainerModule(function (bind, unbind, isBound, rebind) {
            /* Mock logger binding*/
            bind(logger_1.ILogger).to(mock_logger_1.MockLogger);
            contribution_provider_1.bindContributionProvider(bind, keybinding_1.KeybindingContext);
            bind(command_1.CommandRegistry).toSelf().inSingletonScope();
            contribution_provider_1.bindContributionProvider(bind, command_1.CommandContribution);
            bind(keybinding_1.KeybindingRegistry).toSelf();
            contribution_provider_1.bindContributionProvider(bind, keybinding_1.KeybindingContribution);
            bind(TestContribution).toSelf().inSingletonScope();
            [command_1.CommandContribution, keybinding_1.KeybindingContribution].forEach(function (serviceIdentifier) {
                return bind(serviceIdentifier).toService(TestContribution);
            });
            bind(keybinding_1.KeybindingContext).toConstantValue({
                id: 'testContext',
                isEnabled: function (arg) {
                    return true;
                }
            });
            bind(status_bar_1.StatusBarImpl).toSelf().inSingletonScope();
            bind(status_bar_1.StatusBar).toService(status_bar_1.StatusBarImpl);
            bind(command_1.CommandService).toService(command_1.CommandRegistry);
            bind(label_parser_1.LabelParser).toSelf().inSingletonScope();
            bind(frontend_application_state_1.FrontendApplicationStateService).toSelf().inSingletonScope();
        });
        testContainer.load(module);
        commandRegistry = testContainer.get(command_1.CommandRegistry);
        commandRegistry.onStart();
        return [2 /*return*/];
    });
}); });
describe('keybindings', function () {
    var stub;
    before(function () {
        disableJSDOM = jsdom_1.enableJSDOM();
    });
    after(function () {
        disableJSDOM();
    });
    beforeEach(function () {
        stub = sinon.stub(os, 'isOSX').value(false);
        keybindingRegistry = testContainer.get(keybinding_1.KeybindingRegistry);
        keybindingRegistry.onStart();
    });
    afterEach(function () {
        stub.restore();
    });
    it('should register the default keybindings', function () {
        var keybinding = keybindingRegistry.getKeybindingsForCommand(TEST_COMMAND.id);
        expect(keybinding).is.not.undefined;
        var keybinding2 = keybindingRegistry.getKeybindingsForCommand('undefined.command');
        expect(keybinding2.length).is.equal(0);
    });
    it('should set a keymap', function () {
        var keybindings = [{
                command: TEST_COMMAND.id,
                keybinding: 'ctrl+c'
            }];
        keybindingRegistry.setKeymap(keybinding_1.KeybindingScope.USER, keybindings);
        var bindings = keybindingRegistry.getKeybindingsForCommand(TEST_COMMAND.id);
        if (bindings) {
            var keyCode = keys_1.KeyCode.parse(bindings[0].keybinding);
            expect(keyCode.key).to.be.equal(keys_1.Key.KEY_C);
            expect(keyCode.ctrl).to.be.true;
        }
    });
    it('should reset to default in case of invalid keybinding', function () {
        var keybindings = [{
                command: TEST_COMMAND.id,
                keybinding: 'ctrl+invalid'
            }];
        keybindingRegistry.setKeymap(keybinding_1.KeybindingScope.USER, keybindings);
        var bindings = keybindingRegistry.getKeybindingsForCommand(TEST_COMMAND.id);
        if (bindings) {
            var keyCode = keys_1.KeyCode.parse(bindings[0].keybinding);
            expect(keyCode.key).to.be.equal(keys_1.Key.KEY_A);
            expect(keyCode.ctrl).to.be.true;
        }
    });
    it('should remove all keybindings from a command that has multiple keybindings', function () {
        var keybindings = [{
                command: TEST_COMMAND2.id,
                keybinding: 'F3'
            }];
        keybindingRegistry.setKeymap(keybinding_1.KeybindingScope.USER, keybindings);
        var bindings = keybindingRegistry.getKeybindingsForCommand(TEST_COMMAND2.id);
        if (bindings) {
            expect(bindings.length).to.be.equal(2);
            var keyCode = keys_1.KeyCode.parse(bindings[0].keybinding);
            expect(keyCode.key).to.be.equal(keys_1.Key.F1);
            expect(keyCode.ctrl).to.be.true;
        }
    });
    it('should register a correct keybinding, then default back to the original for a wrong one after', function () {
        var keybindings = [{
                command: TEST_COMMAND.id,
                keybinding: 'ctrl+c'
            }];
        // Get default binding
        var keystroke = keybindingRegistry.getKeybindingsForCommand(TEST_COMMAND.id);
        // Set correct new binding
        keybindingRegistry.setKeymap(keybinding_1.KeybindingScope.USER, keybindings);
        var bindings = keybindingRegistry.getKeybindingsForCommand(TEST_COMMAND.id);
        if (bindings) {
            var keyCode = keys_1.KeyCode.parse(bindings[0].keybinding);
            expect(keyCode.key).to.be.equal(keys_1.Key.KEY_C);
            expect(keyCode.ctrl).to.be.true;
        }
        // Set invalid binding
        keybindings = [{
                command: TEST_COMMAND.id,
                keybinding: 'ControlLeft+Invalid'
            }];
        keybindingRegistry.setKeymap(keybinding_1.KeybindingScope.USER, keybindings);
        var defaultBindings = keybindingRegistry.getKeybindingsForCommand(TEST_COMMAND.id);
        if (defaultBindings) {
            if (keystroke) {
                var keyCode = keys_1.KeyCode.parse(defaultBindings[0].keybinding);
                var keyStrokeCode = keys_1.KeyCode.parse(keystroke[0].keybinding);
                expect(keyCode.key).to.be.equal(keyStrokeCode.key);
            }
        }
    });
    it('should only return the more specific keybindings when a keystroke is entered', function () {
        var keybindingsUser = [{
                command: TEST_COMMAND.id,
                keybinding: 'ctrl+b'
            }];
        keybindingRegistry.setKeymap(keybinding_1.KeybindingScope.USER, keybindingsUser);
        var keybindingsSpecific = [{
                command: TEST_COMMAND.id,
                keybinding: 'ctrl+c'
            }];
        var validKeyCode = keys_1.KeyCode.createKeyCode({ first: keys_1.Key.KEY_C, modifiers: [keys_1.KeyModifier.CtrlCmd] });
        keybindingRegistry.setKeymap(keybinding_1.KeybindingScope.WORKSPACE, keybindingsSpecific);
        var bindings = keybindingRegistry.getKeybindingsForKeySequence([keys_1.KeyCode.createKeyCode({ first: keys_1.Key.KEY_A, modifiers: [keys_1.KeyModifier.CtrlCmd] })]).full;
        expect(bindings).to.have.lengthOf(1);
        bindings = keybindingRegistry.getKeybindingsForKeySequence([keys_1.KeyCode.createKeyCode({ first: keys_1.Key.KEY_B, modifiers: [keys_1.KeyModifier.CtrlCmd] })]).full;
        expect(bindings).to.have.lengthOf(1);
        bindings = keybindingRegistry.getKeybindingsForKeySequence([keys_1.KeyCode.createKeyCode({ first: keys_1.Key.KEY_C, modifiers: [keys_1.KeyModifier.CtrlCmd] })]).full;
        var keyCode = keys_1.KeyCode.parse(bindings[0].keybinding);
        expect(keyCode.key).to.be.equal(validKeyCode.key);
    });
    it('should return partial keybinding matches', function () {
        var keybindingsUser = [{
                command: TEST_COMMAND.id,
                keybinding: 'ctrlcmd+x t'
            }];
        keybindingRegistry.setKeymap(keybinding_1.KeybindingScope.USER, keybindingsUser);
        var validKeyCodes = [];
        validKeyCodes.push(keys_1.KeyCode.createKeyCode({ first: keys_1.Key.KEY_C, modifiers: [keys_1.KeyModifier.CtrlCmd] }));
        validKeyCodes.push(keys_1.KeyCode.createKeyCode({ first: keys_1.Key.KEY_T }));
        var bindings = keybindingRegistry.getKeybindingsForKeySequence(keys_1.KeySequence.parse('ctrlcmd+x'));
        expect(bindings.partial.length > 0);
    });
    it('should not register a shadowing keybinding', function () {
        var validKeyBinding = 'ctrlcmd+b a';
        var command = TEST_COMMAND_SHADOW.id;
        var keybindingShadowing = [
            {
                command: command,
                keybinding: validKeyBinding
            },
            {
                command: command,
                keybinding: 'ctrlcmd+b'
            }
        ];
        keybindingRegistry.registerKeybindings.apply(keybindingRegistry, __spread(keybindingShadowing));
        var bindings = keybindingRegistry.getKeybindingsForCommand(command);
        expect(bindings.length).to.be.equal(1);
        expect(bindings[0].keybinding).to.be.equal(validKeyBinding);
    });
    it('shadowed bindings should not be returned', function () {
        var keyCode = keys_1.KeyCode.createKeyCode({ first: keys_1.Key.KEY_A, modifiers: [keys_1.KeyModifier.Shift] });
        var bindings;
        var ignoredDefaultBinding = {
            keybinding: keyCode.toString(),
            command: 'test.ignored-command'
        };
        var defaultBinding = {
            keybinding: keyCode.toString(),
            command: 'test.workspace-command'
        };
        var userBinding = {
            keybinding: keyCode.toString(),
            command: 'test.workspace-command'
        };
        var workspaceBinding = {
            keybinding: keyCode.toString(),
            command: 'test.workspace-command'
        };
        keybindingRegistry.setKeymap(keybinding_1.KeybindingScope.DEFAULT, [defaultBinding, ignoredDefaultBinding]);
        keybindingRegistry.setKeymap(keybinding_1.KeybindingScope.USER, [userBinding]);
        keybindingRegistry.setKeymap(keybinding_1.KeybindingScope.WORKSPACE, [workspaceBinding]);
        // now WORKSPACE bindings are overriding the other scopes
        bindings = keybindingRegistry.getKeybindingsForKeySequence([keyCode]).full;
        expect(bindings).to.have.lengthOf(1);
        expect(bindings[0].command).to.be.equal(workspaceBinding.command);
        keybindingRegistry.resetKeybindingsForScope(keybinding_1.KeybindingScope.WORKSPACE);
        // now it should find USER bindings
        bindings = keybindingRegistry.getKeybindingsForKeySequence([keyCode]).full;
        expect(bindings).to.have.lengthOf(1);
        expect(bindings[0].command).to.be.equal(userBinding.command);
        keybindingRegistry.resetKeybindingsForScope(keybinding_1.KeybindingScope.USER);
        // and finally it should fallback to DEFAULT bindings.
        bindings = keybindingRegistry.getKeybindingsForKeySequence([keyCode]).full;
        expect(bindings).to.have.lengthOf(1);
        expect(bindings[0].command).to.be.equal(defaultBinding.command);
        keybindingRegistry.resetKeybindingsForScope(keybinding_1.KeybindingScope.DEFAULT);
        // now the registry should be empty
        bindings = keybindingRegistry.getKeybindingsForKeySequence([keyCode]).full;
        expect(bindings).to.be.empty;
    });
});
describe('keys api', function () {
    before(function () {
        disableJSDOM = jsdom_1.enableJSDOM();
    });
    after(function () {
        disableJSDOM();
    });
    it('should parse a string to a KeyCode correctly', function () {
        var keycode = keys_1.KeyCode.parse('ctrl+b');
        expect(keycode.ctrl).to.be.true;
        expect(keycode.key).is.equal(keys_1.Key.KEY_B);
        // Invalid keystroke string
        expect(function () { return keys_1.KeyCode.parse('ctl+b'); }).to.throw(Error);
    });
    it('should parse a string containing special modifiers to a KeyCode correctly', function () {
        var stub = sinon.stub(os, 'isOSX').value(false);
        var keycode = keys_1.KeyCode.parse('ctrl+b');
        expect(keycode.ctrl).to.be.true;
        expect(keycode.key).is.equal(keys_1.Key.KEY_B);
        var keycodeOption = keys_1.KeyCode.parse('option+b');
        expect(keycodeOption.alt).to.be.true;
        expect(keycodeOption.key).is.equal(keys_1.Key.KEY_B);
        expect(function () { return keys_1.KeyCode.parse('cmd+b'); }).to.throw(/OSX only/);
        var keycodeCtrlOrCommand = keys_1.KeyCode.parse('ctrlcmd+b');
        expect(keycodeCtrlOrCommand.meta).to.be.false;
        expect(keycodeCtrlOrCommand.ctrl).to.be.true;
        expect(keycodeCtrlOrCommand.key).is.equal(keys_1.Key.KEY_B);
        stub.restore();
    });
    it('should parse a string containing special modifiers to a KeyCode correctly (macOS)', function () {
        keys_1.KeyCode.resetKeyBindings();
        var stub = sinon.stub(os, 'isOSX').value(true);
        var keycode = keys_1.KeyCode.parse('ctrl+b');
        expect(keycode.ctrl).to.be.true;
        expect(keycode.key).is.equal(keys_1.Key.KEY_B);
        var keycodeOption = keys_1.KeyCode.parse('option+b');
        expect(keycodeOption.alt).to.be.true;
        expect(keycodeOption.key).is.equal(keys_1.Key.KEY_B);
        var keycodeCommand = keys_1.KeyCode.parse('cmd+b');
        expect(keycodeCommand.meta).to.be.true;
        expect(keycodeCommand.key).is.equal(keys_1.Key.KEY_B);
        var keycodeCtrlOrCommand = keys_1.KeyCode.parse('ctrlcmd+b');
        expect(keycodeCtrlOrCommand.meta).to.be.true;
        expect(keycodeCtrlOrCommand.ctrl).to.be.false;
        expect(keycodeCtrlOrCommand.key).is.equal(keys_1.Key.KEY_B);
        stub.restore();
    });
    it('it should serialize a keycode properly with BACKQUOTE + M1', function () {
        var stub = sinon.stub(os, 'isOSX').value(true);
        var keyCode = keys_1.KeyCode.createKeyCode({ first: keys_1.Key.BACKQUOTE, modifiers: [keys_1.KeyModifier.CtrlCmd] });
        var keyCodeString = keyCode.toString();
        expect(keyCodeString).to.be.equal('meta+`');
        var parsedKeyCode = keys_1.KeyCode.parse(keyCodeString);
        expect(keys_1.KeyCode.equals(parsedKeyCode, keyCode)).to.be.true;
        sinon.stub(os, 'isOSX').value(false);
        keyCode = keys_1.KeyCode.createKeyCode({ first: keys_1.Key.BACKQUOTE, modifiers: [keys_1.KeyModifier.CtrlCmd] });
        keyCodeString = keyCode.toString();
        expect(keyCodeString).to.be.equal('ctrl+`');
        parsedKeyCode = keys_1.KeyCode.parse(keyCodeString);
        expect(keys_1.KeyCode.equals(parsedKeyCode, keyCode)).to.be.true;
        stub.restore();
    });
    it('it should serialize a keycode properly with a + M2 + M3', function () {
        var keyCode = keys_1.KeyCode.createKeyCode({ first: keys_1.Key.KEY_A, modifiers: [keys_1.KeyModifier.Shift, keys_1.KeyModifier.Alt] });
        var keyCodeString = keyCode.toString();
        expect(keyCodeString).to.be.equal('shift+alt+a');
        var parsedKeyCode = keys_1.KeyCode.parse(keyCodeString);
        expect(keys_1.KeyCode.equals(parsedKeyCode, keyCode)).to.be.true;
    });
    it('the order of the modifiers should not matter when parsing the key code', function () {
        var left = keys_1.KeySequence.parse('shift+alt+a');
        var right = keys_1.KeySequence.parse('alt+shift+a');
        expect(keys_1.KeySequence.compare(left, right)).to.be.equal(keys_1.KeySequence.CompareResult.FULL);
        expect(keys_1.KeySequence.compare([keys_1.KeyCode.createKeyCode({ first: keys_1.Key.KEY_A, modifiers: [keys_1.KeyModifier.Alt, keys_1.KeyModifier.Shift] })], right)).to.be.equal(keys_1.KeySequence.CompareResult.FULL);
        expect(keys_1.KeySequence.compare(left, [keys_1.KeyCode.createKeyCode({ first: keys_1.Key.KEY_A, modifiers: [keys_1.KeyModifier.Alt, keys_1.KeyModifier.Shift] })])).to.be.equal(keys_1.KeySequence.CompareResult.FULL);
        expect(keys_1.KeySequence.compare([keys_1.KeyCode.createKeyCode({ first: keys_1.Key.KEY_A, modifiers: [keys_1.KeyModifier.Shift, keys_1.KeyModifier.Alt] })], right)).to.be.equal(keys_1.KeySequence.CompareResult.FULL);
        expect(keys_1.KeySequence.compare(left, [keys_1.KeyCode.createKeyCode({ first: keys_1.Key.KEY_A, modifiers: [keys_1.KeyModifier.Shift, keys_1.KeyModifier.Alt] })])).to.be.equal(keys_1.KeySequence.CompareResult.FULL);
    });
    it('it should parse ctrl key properly on both OS X and other platforms', function () {
        var event = new KeyboardEvent('keydown', {
            key: keys_1.EasyKey.BACKQUOTE.easyString,
            code: keys_1.Key.BACKQUOTE.code,
            ctrlKey: true,
        });
        var stub = sinon.stub(os, 'isOSX').value(true);
        expect(keys_1.KeyCode.createKeyCode(event).keystroke).to.be.equal('Backquote+M4');
        sinon.stub(os, 'isOSX').value(false);
        expect(keys_1.KeyCode.createKeyCode(event).keystroke).to.be.equal('Backquote+M1');
        stub.restore();
    });
    it('it should serialize a keycode properly with a + M4', function () {
        var stub = sinon.stub(os, 'isOSX').value(true);
        var keyCode = keys_1.KeyCode.createKeyCode({ first: keys_1.Key.KEY_A, modifiers: [keys_1.KeyModifier.MacCtrl] });
        var keyCodeString = keyCode.toString();
        expect(keyCodeString).to.be.equal('ctrl+a');
        var parsedKeyCode = keys_1.KeyCode.parse(keyCodeString);
        expect(keys_1.KeyCode.equals(parsedKeyCode, keyCode)).to.be.true;
        stub.restore();
    });
    it('it should parse a multi keycode keybinding', function () {
        var validKeyCodes = [];
        validKeyCodes.push(keys_1.KeyCode.createKeyCode({ first: keys_1.Key.KEY_A, modifiers: [keys_1.KeyModifier.CtrlCmd] }));
        validKeyCodes.push(keys_1.KeyCode.createKeyCode({ first: keys_1.Key.KEY_C, modifiers: [keys_1.KeyModifier.CtrlCmd, keys_1.KeyModifier.Shift] }));
        var parsedKeyCodes = keys_1.KeySequence.parse('ctrlcmd+a ctrlcmd+shift+c');
        expect(parsedKeyCodes).to.deep.equal(validKeyCodes);
    });
    it('it should parse a multi keycode keybinding with no modifiers', function () {
        var validKeyCodes = [];
        validKeyCodes.push(keys_1.KeyCode.createKeyCode({ first: keys_1.Key.KEY_A, modifiers: [keys_1.KeyModifier.CtrlCmd] }));
        validKeyCodes.push(keys_1.KeyCode.createKeyCode({ first: keys_1.Key.KEY_C }));
        var parsedKeyCodes = keys_1.KeySequence.parse('ctrlcmd+a c');
        expect(parsedKeyCodes).to.deep.equal(validKeyCodes);
    });
    it('it should compare keysequences properly', function () {
        var a = keys_1.KeySequence.parse('ctrlcmd+a');
        var b = keys_1.KeySequence.parse('ctrlcmd+a t');
        expect(keys_1.KeySequence.compare(a, b)).to.be.equal(keys_1.KeySequence.CompareResult.PARTIAL);
        a = keys_1.KeySequence.parse('ctrlcmd+a t');
        b = keys_1.KeySequence.parse('ctrlcmd+a');
        expect(keys_1.KeySequence.compare(a, b)).to.be.equal(keys_1.KeySequence.CompareResult.SHADOW);
        a = keys_1.KeySequence.parse('ctrlcmd+a t');
        b = keys_1.KeySequence.parse('ctrlcmd+a b c');
        expect(keys_1.KeySequence.compare(a, b)).to.be.equal(keys_1.KeySequence.CompareResult.NONE);
        a = keys_1.KeySequence.parse('ctrlcmd+a t');
        b = keys_1.KeySequence.parse('ctrlcmd+a a');
        expect(keys_1.KeySequence.compare(a, b)).to.be.equal(keys_1.KeySequence.CompareResult.NONE);
        a = keys_1.KeySequence.parse('ctrlcmd+a t');
        b = keys_1.KeySequence.parse('ctrlcmd+a t');
        expect(keys_1.KeySequence.compare(a, b)).to.be.equal(keys_1.KeySequence.CompareResult.FULL);
        a = keys_1.KeySequence.parse('ctrlcmd+a t b');
        b = keys_1.KeySequence.parse('ctrlcmd+a t b');
        expect(keys_1.KeySequence.compare(a, b)).to.be.equal(keys_1.KeySequence.CompareResult.FULL);
    });
    it('it should be a modifier only', function () {
        var keyCode = keys_1.KeyCode.createKeyCode({ modifiers: [keys_1.KeyModifier.CtrlCmd] });
        expect(keyCode).to.be.deep.equal(keys_1.KeyCode.createKeyCode({ modifiers: [keys_1.KeyModifier.CtrlCmd] }));
        expect(keyCode.isModifierOnly()).to.be.true;
    });
    it('it should be multiple modifiers only', function () {
        var keyCode = keys_1.KeyCode.createKeyCode({ modifiers: [keys_1.KeyModifier.CtrlCmd, keys_1.KeyModifier.Alt] });
        expect(keyCode).to.be.deep.equal(keys_1.KeyCode.createKeyCode({ modifiers: [keys_1.KeyModifier.CtrlCmd, keys_1.KeyModifier.Alt] }));
        expect(keyCode.isModifierOnly()).to.be.true;
    });
    it('it should translate non US layout chords properly', function () {
        // mimic a german layout, i.e. the '/' is on the 'Shift+7'.
        var keyCode = new keys_1.KeyCode(keys_1.KeyCode.parse('ctrlcmd+shift+7').keystroke, '/');
        var normalized = keyCode.normalizeToUsLayout();
        expect(normalized).to.be.deep.equal(keys_1.KeyCode.parse('ctrlcmd+/'));
    });
});
var TEST_COMMAND = {
    id: 'test.command'
};
var TEST_COMMAND2 = {
    id: 'test.command2'
};
var TEST_COMMAND_SHADOW = {
    id: 'test.command-shadow'
};
var TestContribution = /** @class */ (function () {
    function TestContribution() {
    }
    TestContribution.prototype.registerCommands = function (commands) {
        commands.registerCommand(TEST_COMMAND);
        commands.registerCommand(TEST_COMMAND2);
        commands.registerCommand(TEST_COMMAND_SHADOW);
    };
    TestContribution.prototype.registerKeybindings = function (keybindings) {
        [{
                command: TEST_COMMAND.id,
                context: 'testContext',
                keybinding: 'ctrl+a'
            },
            {
                command: TEST_COMMAND2.id,
                context: 'testContext',
                keybinding: 'ctrl+f1'
            },
            {
                command: TEST_COMMAND2.id,
                context: 'testContext',
                keybinding: 'ctrl+f2'
            },
        ].forEach(function (binding) {
            keybindings.registerKeybinding(binding);
        });
    };
    TestContribution = __decorate([
        inversify_1.injectable()
    ], TestContribution);
    return TestContribution;
}());
exports.TestContribution = TestContribution;
