// @theia/core/lib/browser/keys.js
"use strict";
/********************************************************************************
 * Copyright (C) 2017 TypeFox and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
var __values = (this && this.__values) || function (o) {
    var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
    if (m) return m.call(o);
    return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spread = (this && this.__spread) || function () {
    for (var ar = [], i = 0; i < arguments.length; i++) ar = ar.concat(__read(arguments[i]));
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
var os_1 = require("../common/os");
var KeySequence;
(function (KeySequence) {
    function equals(a, b) {
        if (a.length !== b.length) {
            return false;
        }
        for (var i = 0; i < a.length; i++) {
            if (a[i].equals(b[i]) === false) {
                return false;
            }
        }
        return true;
    }
    KeySequence.equals = equals;
    var CompareResult;
    (function (CompareResult) {
        CompareResult[CompareResult["NONE"] = 0] = "NONE";
        CompareResult[CompareResult["PARTIAL"] = 1] = "PARTIAL";
        CompareResult[CompareResult["SHADOW"] = 2] = "SHADOW";
        CompareResult[CompareResult["FULL"] = 3] = "FULL";
    })(CompareResult = KeySequence.CompareResult || (KeySequence.CompareResult = {}));
    /* Compares two KeySequences, returns:
     * FULL if the KeySequences are the same.
     * PARTIAL if the KeySequence a part of b.
     * SHADOW if the KeySequence b part of a.
     * NONE if the KeySequences are not the same at all.
     */
    function compare(a, b) {
        var first = a;
        var second = b;
        var shadow = false;
        if (b.length < a.length) {
            first = b;
            second = a;
            shadow = true;
        }
        for (var i = 0; i < first.length; i++) {
            if (first[i].equals(second[i]) === false) {
                return KeySequence.CompareResult.NONE;
            }
        }
        if (first.length < second.length) {
            if (shadow === false) {
                return KeySequence.CompareResult.PARTIAL;
            }
            else {
                return KeySequence.CompareResult.SHADOW;
            }
        }
        return KeySequence.CompareResult.FULL;
    }
    KeySequence.compare = compare;
    function parse(keybinding) {
        var e_1, _a;
        var keyCodes = [];
        var rawKeyCodes = keybinding.split(' ');
        try {
            for (var rawKeyCodes_1 = __values(rawKeyCodes), rawKeyCodes_1_1 = rawKeyCodes_1.next(); !rawKeyCodes_1_1.done; rawKeyCodes_1_1 = rawKeyCodes_1.next()) {
                var rawKeyCode = rawKeyCodes_1_1.value;
                var keyCode = KeyCode.parse(rawKeyCode);
                if (keyCode !== undefined) {
                    keyCodes.push(keyCode);
                }
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (rawKeyCodes_1_1 && !rawKeyCodes_1_1.done && (_a = rawKeyCodes_1.return)) _a.call(rawKeyCodes_1);
            }
            finally { if (e_1) throw e_1.error; }
        }
        return keyCodes;
    }
    KeySequence.parse = parse;
    /**
     * Return an array of strings representing the keys in this keysequence.
     * For example ['ctrlcmd p', 'ctrlcmd k'].  The character used between
     * keys can be overriden using `separator`.
     */
    function acceleratorFor(keySequence, separator) {
        if (separator === void 0) { separator = ' '; }
        var e_2, _a;
        var result = [];
        try {
            for (var keySequence_1 = __values(keySequence), keySequence_1_1 = keySequence_1.next(); !keySequence_1_1.done; keySequence_1_1 = keySequence_1.next()) {
                var keyCode = keySequence_1_1.value;
                var keyCodeResult = '';
                var previous = false;
                if (keyCode.meta && os_1.isOSX) {
                    if (os_1.isOSX) {
                        keyCodeResult += 'Cmd';
                        previous = true;
                    }
                }
                if (keyCode.ctrl) {
                    if (previous) {
                        keyCodeResult += separator;
                    }
                    keyCodeResult += 'Ctrl';
                    previous = true;
                }
                if (keyCode.alt) {
                    if (previous) {
                        keyCodeResult += separator;
                    }
                    keyCodeResult += 'Alt';
                    previous = true;
                }
                if (keyCode.shift) {
                    if (previous) {
                        keyCodeResult += separator;
                    }
                    keyCodeResult += 'Shift';
                    previous = true;
                }
                if (keyCode.key) {
                    if (previous) {
                        keyCodeResult += separator;
                    }
                    var keyString = Key.getEasyKey(keyCode.key).easyString;
                    if (keyCode.key.keyCode >= Key.KEY_A.keyCode && keyCode.key.keyCode <= Key.KEY_Z.keyCode ||
                        keyCode.key.keyCode >= Key.F1.keyCode && keyCode.key.keyCode <= Key.F24.keyCode) {
                        /* We want to capitalize single letters and function keys */
                        keyCodeResult += keyString.toUpperCase();
                    }
                    else if (Key.getEasyKey(keyCode.key).easyString.length > 1) {
                        /* We want to only capitalize the first letter */
                        keyCodeResult += keyString.charAt(0).toUpperCase() + keyString.slice(1);
                    }
                    else {
                        /* Return the symbol as is */
                        keyCodeResult += keyString;
                    }
                }
                result.push(keyCodeResult);
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (keySequence_1_1 && !keySequence_1_1.done && (_a = keySequence_1.return)) _a.call(keySequence_1);
            }
            finally { if (e_2) throw e_2.error; }
        }
        return result;
    }
    KeySequence.acceleratorFor = acceleratorFor;
})(KeySequence = exports.KeySequence || (exports.KeySequence = {}));
/**
 * Representation of a platform independent key code.
 */
var KeyCode = /** @class */ (function () {
    function KeyCode(keystroke, character) {
        this.keystroke = keystroke;
        this.character = character;
        this.key = undefined;
        var parts = keystroke.split('+');
        if (parts.every(KeyCode.isModifierString) === false) {
            this.key = Key.getKey(parts[0]);
        }
        if (os_1.isOSX) {
            this.meta = parts.some(function (part) { return part === KeyModifier.CtrlCmd; });
            this.shift = parts.some(function (part) { return part === KeyModifier.Shift; });
            this.alt = parts.some(function (part) { return part === KeyModifier.Alt; });
            this.ctrl = parts.some(function (part) { return part === KeyModifier.MacCtrl; });
        }
        else {
            this.meta = false;
            this.ctrl = parts.some(function (part) { return part === KeyModifier.CtrlCmd; });
            this.shift = parts.some(function (part) { return part === KeyModifier.Shift; });
            this.alt = parts.some(function (part) { return part === KeyModifier.Alt; });
        }
    }
    /**
     * Returns a normalized version of this keycode, if
     * - shift or alt was pressed
     * - meta or ctrl was pressed
     * - a character exists
     * - the character corresponds to a key on the US keyboard layout
     *
     * The resulting KeyCode will remove the shift/alt keys and use the character as the key.
     * For instance on a german kb layout the sequence `ctrlCmd+shift+7` would be translated to `ctrlCmd+/`.
     * @returns a normalized keycode or undefined
     */
    KeyCode.prototype.normalizeToUsLayout = function () {
        if ((this.shift || this.alt) && (this.meta || this.ctrl) && this.character && EASY_TO_KEY[this.character]) {
            var modifiers = [];
            if (os_1.isOSX) {
                if (this.meta) {
                    modifiers.push(KeyModifier.CtrlCmd);
                }
                if (this.ctrl) {
                    modifiers.push(KeyModifier.MacCtrl);
                }
            }
            else {
                if (this.ctrl) {
                    modifiers.push(KeyModifier.CtrlCmd);
                }
            }
            return KeyCode.createKeyCode({ first: EASY_TO_KEY[this.character], modifiers: modifiers });
        }
        return undefined;
    };
    /* Return true of string is a modifier M1 to M4 */
    KeyCode.isModifierString = function (key) {
        if (key === KeyModifier.CtrlCmd
            || key === KeyModifier.Shift
            || key === KeyModifier.Alt
            || key === KeyModifier.MacCtrl) {
            return true;
        }
        return false;
    };
    /**
     * Returns true KeyCode only contains modifiers.
     */
    KeyCode.prototype.isModifierOnly = function () {
        if (this.key === undefined) {
            return true;
        }
        else {
            return false;
        }
    };
    /**
     * Parses a string and returns a KeyCode object.
     * @param keybinding String representation of a keybinding
     */
    KeyCode.parse = function (keybinding) {
        var e_3, _a;
        if (KeyCode.keybindings[keybinding]) {
            return KeyCode.keybindings[keybinding];
        }
        var sequence = [];
        var keys = keybinding.split('+');
        /* If duplicates i.e ctrl+ctrl+a or alt+alt+b or b+alt+b it is invalid */
        if (keys.length !== new Set(keys).size) {
            throw new Error("Can't parse keybinding " + keybinding + " Duplicate modifiers");
        }
        try {
            for (var keys_1 = __values(keys), keys_1_1 = keys_1.next(); !keys_1_1.done; keys_1_1 = keys_1.next()) {
                var keyString = keys_1_1.value;
                if (SPECIAL_ALIASES[keyString] !== undefined) {
                    keyString = SPECIAL_ALIASES[keyString];
                }
                var key = EASY_TO_KEY[keyString];
                /* meta only works on macOS */
                if (keyString === SpecialCases.META) {
                    if (os_1.isOSX) {
                        sequence.push("" + KeyModifier.CtrlCmd);
                    }
                    else {
                        throw new Error("Can't parse keybinding " + keybinding + " meta is for OSX only");
                    }
                    /* ctrlcmd for M1 keybindings that work on both macOS and other platforms */
                }
                else if (keyString === SpecialCases.CTRLCMD) {
                    sequence.push("" + KeyModifier.CtrlCmd);
                }
                else if (Key.isKey(key)) {
                    if (Key.isModifier(key.code)) {
                        if (key.keyCode === EasyKey.CONTROL.keyCode) {
                            // CTRL on MacOS X (M4)
                            if (os_1.isOSX) {
                                sequence.push("" + KeyModifier.MacCtrl);
                            }
                            else {
                                sequence.push("" + KeyModifier.CtrlCmd);
                            }
                        }
                        else if (key.keyCode === EasyKey.SHIFT.keyCode) {
                            sequence.push("" + KeyModifier.Shift);
                        }
                        else if (key.keyCode === EasyKey.ALT.keyCode) {
                            sequence.push("" + KeyModifier.Alt);
                        }
                    }
                    else {
                        sequence.unshift(key.code);
                    }
                }
                else {
                    throw new Error("Unrecognized key '" + keyString + "' in '" + keybinding + "'");
                }
            }
        }
        catch (e_3_1) { e_3 = { error: e_3_1 }; }
        finally {
            try {
                if (keys_1_1 && !keys_1_1.done && (_a = keys_1.return)) _a.call(keys_1);
            }
            finally { if (e_3) throw e_3.error; }
        }
        // We need to sort the modifier keys, but on the modifiers, so it always keeps the M1 less than M2, M2 less than M3 and so on order.
        // We intentionally ignore other cases.
        sequence.sort(function (left, right) { return KeyModifier.isModifier(left) && KeyModifier.isModifier(right) ? left.localeCompare(right) : 0; });
        KeyCode.keybindings[keybinding] = new KeyCode(sequence.join('+'));
        return KeyCode.keybindings[keybinding];
    };
    KeyCode.createKeyCode = function (event) {
        if (KeyCode.isKeyboardEvent(event)) {
            var code = KeyCode.toCode(event);
            var sequence = [];
            if (!Key.isModifier(code)) {
                sequence.push(code);
            }
            // CTRL + COMMAND (M1)
            if ((os_1.isOSX && event.metaKey) || (!os_1.isOSX && event.ctrlKey)) {
                sequence.push("" + KeyModifier.CtrlCmd);
            }
            // SHIFT (M2)
            if (event.shiftKey) {
                sequence.push("" + KeyModifier.Shift);
            }
            // ALT (M3)
            if (event.altKey) {
                sequence.push("" + KeyModifier.Alt);
            }
            // CTRL on MacOS X (M4)
            if (os_1.isOSX && !event.metaKey && event.ctrlKey) {
                sequence.push("" + KeyModifier.MacCtrl);
            }
            return new KeyCode(sequence.join('+'), KeyCode.toCharacter(event));
        }
        else {
            var key = event.first ? [event.first.code] : [];
            return new KeyCode(key
                .concat((event.modifiers || []).sort().map(function (modifier) { return "" + modifier; }))
                .join('+'));
        }
    };
    /**
     * Determine a code that can be turned into a Key using `Key.getKey()`.
     *
     * This function should consider the actual keyboard layout as much as possible, which is not supported
     * by the `code` property of `KeyboardEvent`. Some keyboard layouts contain keys that cannot be represented
     * with such a code (e.g. the &Ouml; key on German keyboards). In these cases this function often returns
     * the key code according to the US keyboard layout; `toCharacter()` might deliver a more appropriate hint
     * on which key was actually pressed.
     *
     * `keyIdentifier` is used to access this deprecated field:
     * https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/keyIdentifier
     */
    KeyCode.toCode = function (event) {
        // A system and implementation dependent numerical code identifying the unmodified value of the pressed key.
        if (event.keyCode) {
            var key = Key.getKey(event.keyCode);
            if (key) {
                return key.code;
            }
        }
        // A "key identifier" string that can be used to determine what key was pressed.
        if (event.keyIdentifier) {
            var key = Key.getKey(event.keyIdentifier);
            if (key) {
                return key.code;
            }
        }
        // The numeric keyCode of the key pressed, or the character code (charCode) for an alphanumeric key pressed.
        if (event.which) {
            var key = Key.getKey(event.which);
            if (key) {
                return key.code;
            }
        }
        // A physical key on the keyboard, as opposed to the character generated by pressing the key.
        // This property returns a value which isn't altered by keyboard layout or the state of the modifier keys.
        if (event.code) {
            var key = Key.getKey(event.code);
            if (key) {
                return key.code;
            }
        }
        throw new Error("Cannot get key code from the keyboard event: " + event + ".");
    };
    /**
     * Determine the actual printable character that is generated from a pressed key.
     * If the key does not correspond to a printable character, `undefined` is returned.
     * The result may be altered by modifier keys.
     */
    KeyCode.toCharacter = function (event) {
        var key = event.key;
        // Use the key property if it contains exactly one unicode character
        if (key && Array.from(key).length === 1) {
            return key;
        }
        var charCode = event.charCode;
        // Use the charCode property if it does not correspond to a unicode control character
        if (charCode && charCode > 0x1f && !(charCode >= 0x80 && charCode <= 0x9f)) {
            return String.fromCharCode(charCode);
        }
        return undefined;
    };
    KeyCode.equals = function (keyCode1, keyCode2) {
        return JSON.stringify(keyCode1) === JSON.stringify(keyCode2);
    };
    KeyCode.prototype.equals = function (event) {
        return (event instanceof KeyCode ? event : KeyCode.createKeyCode(event)).keystroke === this.keystroke;
    };
    /* Reset the key hashmap, this is for testing purposes.  */
    KeyCode.resetKeyBindings = function () {
        KeyCode.keybindings = {};
    };
    /* Return a keybinding string compatible with the Keybinding.keybinding property.  */
    KeyCode.prototype.toString = function () {
        var result = '';
        var previous = false;
        if (this.meta) {
            result += SpecialCases.META;
            previous = true;
        }
        if (this.shift) {
            if (previous) {
                result += '+';
            }
            result += EasyKey.SHIFT.easyString;
            previous = true;
        }
        if (this.alt) {
            if (previous) {
                result += '+';
            }
            result += EasyKey.ALT.easyString;
            previous = true;
        }
        if (this.ctrl) {
            if (previous) {
                result += '+';
            }
            result += EasyKey.CONTROL.easyString;
            previous = true;
        }
        if (this.key) {
            if (previous) {
                result += '+';
            }
            result += KEY_CODE_TO_EASY[this.key.keyCode].easyString;
        }
        return result;
    };
    KeyCode.keybindings = {};
    return KeyCode;
}());
exports.KeyCode = KeyCode;
(function (KeyCode) {
    /**
     * Different scopes have different execution environments. This means that they have different built-ins
     * (different global object, different constructors, etc.). This may result in unexpected results. For instance,
     * `[] instanceof window.frames[0].Array` will return `false`, because `Array.prototype !== window.frames[0].Array`
     * and arrays inherit from the former.
     * See: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/instanceof
     *
     * Note: just add another check if the current `event.type` checking is insufficient.
     */
    function isKeyboardEvent(event) {
        if (event instanceof KeyboardEvent) {
            return true;
        }
        var type = event.type;
        if (type) {
            return type === 'keypress' || type === 'keydown' || type === 'keyup';
        }
        return false;
    }
    KeyCode.isKeyboardEvent = isKeyboardEvent;
})(KeyCode = exports.KeyCode || (exports.KeyCode = {}));
exports.KeyCode = KeyCode;
var KeyModifier;
(function (KeyModifier) {
    /**
     * M1 is the COMMAND key on MacOS X, and the CTRL key on most other platforms.
     */
    KeyModifier["CtrlCmd"] = "M1";
    /**
     * M2 is the SHIFT key.
     */
    KeyModifier["Shift"] = "M2";
    /**
     * M3 is the Option key on MacOS X, and the ALT key on most other platforms.
     */
    KeyModifier["Alt"] = "M3";
    /**
     * M4 is the CTRL key on MacOS X, and is undefined on other platforms.
     */
    KeyModifier["MacCtrl"] = "M4";
})(KeyModifier = exports.KeyModifier || (exports.KeyModifier = {}));
(function (KeyModifier) {
    /**
     * The CTRL key, independently of the platform.
     * _Note:_ In general `KeyModifier.CtrlCmd` should be preferred over this constant.
     */
    KeyModifier.CTRL = os_1.isOSX ? KeyModifier.MacCtrl : KeyModifier.CtrlCmd;
    /**
     * An alias for the SHIFT key (`KeyModifier.Shift`).
     */
    KeyModifier.SHIFT = KeyModifier.Shift;
    /**
     * `true` if the argument represents a modifier. Otherwise, `false`.
     */
    function isModifier(key) {
        if (key) {
            switch (key) {
                case 'M1': // Fall through.
                case 'M2': // Fall through.
                case 'M3': // Fall through.
                case 'M4': return true;
                default: return false;
            }
        }
        return false;
    }
    KeyModifier.isModifier = isModifier;
})(KeyModifier = exports.KeyModifier || (exports.KeyModifier = {}));
var CODE_TO_KEY = {};
var KEY_CODE_TO_KEY = {};
var KEY_CODE_TO_EASY = {};
var EASY_TO_KEY = {}; // From 'ctrl' to Key structure
var MODIFIERS = [];
var SPECIAL_ALIASES = {
    'option': 'alt',
    'command': 'meta',
    'cmd': 'meta',
    'return': 'enter',
    'esc': 'escape',
    'mod': 'ctrl',
    'ins': 'insert',
    'del': 'delete',
    'control': 'ctrl',
};
var SpecialCases;
(function (SpecialCases) {
    SpecialCases.META = 'meta';
    SpecialCases.CTRLCMD = 'ctrlcmd';
})(SpecialCases = exports.SpecialCases || (exports.SpecialCases = {}));
var EasyKey;
(function (EasyKey) {
    EasyKey.ENTER = { keyCode: 13, easyString: 'enter' };
    EasyKey.SPACE = { keyCode: 32, easyString: 'space' };
    EasyKey.BACKSPACE = { keyCode: 8, easyString: 'backspace' };
    EasyKey.TAB = { keyCode: 9, easyString: 'tab' };
    EasyKey.DELETE = { keyCode: 46, easyString: 'delete' };
    EasyKey.END = { keyCode: 35, easyString: 'end' };
    EasyKey.HOME = { keyCode: 36, easyString: 'home' };
    EasyKey.INSERT = { keyCode: 45, easyString: 'insert' };
    EasyKey.PAGE_DOWN = { keyCode: 34, easyString: 'pagedown' };
    EasyKey.PAGE_UP = { keyCode: 33, easyString: 'pageup' };
    EasyKey.ARROW_DOWN = { keyCode: 40, easyString: 'down' };
    EasyKey.ARROW_LEFT = { keyCode: 37, easyString: 'left' };
    EasyKey.ARROW_RIGHT = { keyCode: 39, easyString: 'right' };
    EasyKey.ARROW_UP = { keyCode: 38, easyString: 'up' };
    EasyKey.ESCAPE = { keyCode: 27, easyString: 'escape' };
    EasyKey.ALT = { keyCode: 18, easyString: 'alt' };
    EasyKey.CAPS_LOCK = { keyCode: 20, easyString: 'capslock' };
    EasyKey.CONTROL = { keyCode: 17, easyString: 'ctrl' };
    EasyKey.OS = { keyCode: 91, easyString: 'super' };
    EasyKey.SHIFT = { keyCode: 16, easyString: 'shift' };
    EasyKey.DIGIT1 = { keyCode: 49, easyString: '1' };
    EasyKey.DIGIT2 = { keyCode: 50, easyString: '2' };
    EasyKey.DIGIT3 = { keyCode: 51, easyString: '3' };
    EasyKey.DIGIT4 = { keyCode: 52, easyString: '4' };
    EasyKey.DIGIT5 = { keyCode: 53, easyString: '5' };
    EasyKey.DIGIT6 = { keyCode: 54, easyString: '6' };
    EasyKey.DIGIT7 = { keyCode: 55, easyString: '7' };
    EasyKey.DIGIT8 = { keyCode: 56, easyString: '8' };
    EasyKey.DIGIT9 = { keyCode: 57, easyString: '9' };
    EasyKey.DIGIT0 = { keyCode: 48, easyString: '0' };
    EasyKey.KEY_A = { keyCode: 65, easyString: 'a' };
    EasyKey.KEY_B = { keyCode: 66, easyString: 'b' };
    EasyKey.KEY_C = { keyCode: 67, easyString: 'c' };
    EasyKey.KEY_D = { keyCode: 68, easyString: 'd' };
    EasyKey.KEY_E = { keyCode: 69, easyString: 'e' };
    EasyKey.KEY_F = { keyCode: 70, easyString: 'f' };
    EasyKey.KEY_G = { keyCode: 71, easyString: 'g' };
    EasyKey.KEY_H = { keyCode: 72, easyString: 'h' };
    EasyKey.KEY_I = { keyCode: 73, easyString: 'i' };
    EasyKey.KEY_J = { keyCode: 74, easyString: 'j' };
    EasyKey.KEY_K = { keyCode: 75, easyString: 'k' };
    EasyKey.KEY_L = { keyCode: 76, easyString: 'l' };
    EasyKey.KEY_M = { keyCode: 77, easyString: 'm' };
    EasyKey.KEY_N = { keyCode: 78, easyString: 'n' };
    EasyKey.KEY_O = { keyCode: 79, easyString: 'o' };
    EasyKey.KEY_P = { keyCode: 80, easyString: 'p' };
    EasyKey.KEY_Q = { keyCode: 81, easyString: 'q' };
    EasyKey.KEY_R = { keyCode: 82, easyString: 'r' };
    EasyKey.KEY_S = { keyCode: 83, easyString: 's' };
    EasyKey.KEY_T = { keyCode: 84, easyString: 't' };
    EasyKey.KEY_U = { keyCode: 85, easyString: 'u' };
    EasyKey.KEY_V = { keyCode: 86, easyString: 'v' };
    EasyKey.KEY_W = { keyCode: 87, easyString: 'w' };
    EasyKey.KEY_X = { keyCode: 88, easyString: 'x' };
    EasyKey.KEY_Y = { keyCode: 89, easyString: 'y' };
    EasyKey.KEY_Z = { keyCode: 90, easyString: 'z' };
    EasyKey.MULTIPLY = { keyCode: 106, easyString: 'multiply' };
    EasyKey.ADD = { keyCode: 107, easyString: 'add' };
    EasyKey.DECIMAL = { keyCode: 108, easyString: 'decimal' };
    EasyKey.SUBTRACT = { keyCode: 109, easyString: 'subtract' };
    EasyKey.DIVIDE = { keyCode: 111, easyString: 'divide' };
    EasyKey.F1 = { keyCode: 112, easyString: 'f1' };
    EasyKey.F2 = { keyCode: 113, easyString: 'f2' };
    EasyKey.F3 = { keyCode: 114, easyString: 'f3' };
    EasyKey.F4 = { keyCode: 115, easyString: 'f4' };
    EasyKey.F5 = { keyCode: 116, easyString: 'f5' };
    EasyKey.F6 = { keyCode: 117, easyString: 'f6' };
    EasyKey.F7 = { keyCode: 118, easyString: 'f7' };
    EasyKey.F8 = { keyCode: 119, easyString: 'f8' };
    EasyKey.F9 = { keyCode: 120, easyString: 'f9' };
    EasyKey.F10 = { keyCode: 121, easyString: 'f10' };
    EasyKey.F11 = { keyCode: 122, easyString: 'f11' };
    EasyKey.F12 = { keyCode: 123, easyString: 'f12' };
    EasyKey.F13 = { keyCode: 124, easyString: 'f13' };
    EasyKey.F14 = { keyCode: 125, easyString: 'f14' };
    EasyKey.F15 = { keyCode: 126, easyString: 'f15' };
    EasyKey.F16 = { keyCode: 127, easyString: 'f16' };
    EasyKey.F17 = { keyCode: 128, easyString: 'f17' };
    EasyKey.F18 = { keyCode: 129, easyString: 'f18' };
    EasyKey.F19 = { keyCode: 130, easyString: 'f19' };
    EasyKey.F20 = { keyCode: 131, easyString: 'f20' };
    EasyKey.F21 = { keyCode: 132, easyString: 'f21' };
    EasyKey.F22 = { keyCode: 133, easyString: 'f22' };
    EasyKey.F23 = { keyCode: 134, easyString: 'f23' };
    EasyKey.F24 = { keyCode: 135, easyString: 'f24' };
    EasyKey.NUM_LOCK = { keyCode: 144, easyString: 'numlock' };
    EasyKey.COMMA = { keyCode: 188, easyString: ',' };
    EasyKey.PERIOD = { keyCode: 190, easyString: '.' };
    EasyKey.SLASH = { keyCode: 191, easyString: '/' };
    EasyKey.SEMICOLON = { keyCode: 186, easyString: ';' };
    EasyKey.QUOTE = { keyCode: 222, easyString: '\'' };
    EasyKey.BRACKET_LEFT = { keyCode: 219, easyString: '[' };
    EasyKey.BRACKET_RIGHT = { keyCode: 221, easyString: ']' };
    EasyKey.BACKQUOTE = { keyCode: 192, easyString: '`' };
    EasyKey.BACKSLASH = { keyCode: 220, easyString: '\\' };
    EasyKey.MINUS = { keyCode: 189, easyString: '-' };
    EasyKey.EQUAL = { keyCode: 187, easyString: '=' };
    EasyKey.INTL_RO = { keyCode: 193, easyString: 'intlro' };
    EasyKey.INTL_YEN = { keyCode: 255, easyString: 'intlyen' };
})(EasyKey = exports.EasyKey || (exports.EasyKey = {}));
var Key;
(function (Key) {
    // tslint:disable-next-line:no-any
    function isKey(arg) {
        return !!arg && ('code' in arg) && ('keyCode' in arg);
    }
    Key.isKey = isKey;
    function getKey(arg) {
        if (typeof arg === 'number') {
            return KEY_CODE_TO_KEY[arg];
        }
        else {
            return CODE_TO_KEY[arg];
        }
    }
    Key.getKey = getKey;
    function getEasyKey(key) {
        return KEY_CODE_TO_EASY[key.keyCode];
    }
    Key.getEasyKey = getEasyKey;
    function isModifier(arg) {
        if (typeof arg === 'number') {
            return MODIFIERS.map(function (key) { return key.keyCode; }).indexOf(arg) > 0;
        }
        return MODIFIERS.map(function (key) { return key.code; }).indexOf(arg) > 0;
    }
    Key.isModifier = isModifier;
    function equals(key, keyCode) {
        return !!keyCode.key && key.keyCode === keyCode.key.keyCode;
    }
    Key.equals = equals;
    Key.BACKSPACE = { code: 'Backspace', keyCode: 8 };
    Key.TAB = { code: 'Tab', keyCode: 9 };
    Key.ENTER = { code: 'Enter', keyCode: 13 };
    Key.ESCAPE = { code: 'Escape', keyCode: 27 };
    Key.SPACE = { code: 'Space', keyCode: 32 };
    Key.PAGE_UP = { code: 'PageUp', keyCode: 33 };
    Key.PAGE_DOWN = { code: 'PageDown', keyCode: 34 };
    Key.END = { code: 'End', keyCode: 35 };
    Key.HOME = { code: 'Home', keyCode: 36 };
    Key.ARROW_LEFT = { code: 'ArrowLeft', keyCode: 37 };
    Key.ARROW_UP = { code: 'ArrowUp', keyCode: 38 };
    Key.ARROW_RIGHT = { code: 'ArrowRight', keyCode: 39 };
    Key.ARROW_DOWN = { code: 'ArrowDown', keyCode: 40 };
    Key.INSERT = { code: 'Insert', keyCode: 45 };
    Key.DELETE = { code: 'Delete', keyCode: 46 };
    Key.SHIFT_LEFT = { code: 'ShiftLeft', keyCode: 16 };
    Key.SHIFT_RIGHT = { code: 'ShiftRight', keyCode: 16 };
    Key.CONTROL_LEFT = { code: 'ControlLeft', keyCode: 17 };
    Key.CONTROL_RIGHT = { code: 'ControlRight', keyCode: 17 };
    Key.ALT_LEFT = { code: 'AltLeft', keyCode: 18 };
    Key.ALT_RIGHT = { code: 'AltRight', keyCode: 18 };
    Key.CAPS_LOCK = { code: 'CapsLock', keyCode: 20 };
    Key.OS_LEFT = { code: 'OSLeft', keyCode: 91 };
    Key.OS_RIGHT = { code: 'OSRight', keyCode: 92 };
    Key.DIGIT0 = { code: 'Digit0', keyCode: 48 };
    Key.DIGIT1 = { code: 'Digit1', keyCode: 49 };
    Key.DIGIT2 = { code: 'Digit2', keyCode: 50 };
    Key.DIGIT3 = { code: 'Digit3', keyCode: 51 };
    Key.DIGIT4 = { code: 'Digit4', keyCode: 52 };
    Key.DIGIT5 = { code: 'Digit5', keyCode: 53 };
    Key.DIGIT6 = { code: 'Digit6', keyCode: 54 };
    Key.DIGIT7 = { code: 'Digit7', keyCode: 55 };
    Key.DIGIT8 = { code: 'Digit8', keyCode: 56 };
    Key.DIGIT9 = { code: 'Digit9', keyCode: 57 };
    Key.KEY_A = { code: 'KeyA', keyCode: 65 };
    Key.KEY_B = { code: 'KeyB', keyCode: 66 };
    Key.KEY_C = { code: 'KeyC', keyCode: 67 };
    Key.KEY_D = { code: 'KeyD', keyCode: 68 };
    Key.KEY_E = { code: 'KeyE', keyCode: 69 };
    Key.KEY_F = { code: 'KeyF', keyCode: 70 };
    Key.KEY_G = { code: 'KeyG', keyCode: 71 };
    Key.KEY_H = { code: 'KeyH', keyCode: 72 };
    Key.KEY_I = { code: 'KeyI', keyCode: 73 };
    Key.KEY_J = { code: 'KeyJ', keyCode: 74 };
    Key.KEY_K = { code: 'KeyK', keyCode: 75 };
    Key.KEY_L = { code: 'KeyL', keyCode: 76 };
    Key.KEY_M = { code: 'KeyM', keyCode: 77 };
    Key.KEY_N = { code: 'KeyN', keyCode: 78 };
    Key.KEY_O = { code: 'KeyO', keyCode: 79 };
    Key.KEY_P = { code: 'KeyP', keyCode: 80 };
    Key.KEY_Q = { code: 'KeyQ', keyCode: 81 };
    Key.KEY_R = { code: 'KeyR', keyCode: 82 };
    Key.KEY_S = { code: 'KeyS', keyCode: 83 };
    Key.KEY_T = { code: 'KeyT', keyCode: 84 };
    Key.KEY_U = { code: 'KeyU', keyCode: 85 };
    Key.KEY_V = { code: 'KeyV', keyCode: 86 };
    Key.KEY_W = { code: 'KeyW', keyCode: 87 };
    Key.KEY_X = { code: 'KeyX', keyCode: 88 };
    Key.KEY_Y = { code: 'KeyY', keyCode: 89 };
    Key.KEY_Z = { code: 'KeyZ', keyCode: 90 };
    Key.MULTIPLY = { code: 'NumpadMultiply', keyCode: 106 };
    Key.ADD = { code: 'NumpadAdd', keyCode: 107 };
    Key.DECIMAL = { code: 'NumpadDecimal', keyCode: 108 };
    Key.SUBTRACT = { code: 'NumpadSubtract', keyCode: 109 };
    Key.DIVIDE = { code: 'NumpadDivide', keyCode: 111 };
    Key.F1 = { code: 'F1', keyCode: 112 };
    Key.F2 = { code: 'F2', keyCode: 113 };
    Key.F3 = { code: 'F3', keyCode: 114 };
    Key.F4 = { code: 'F4', keyCode: 115 };
    Key.F5 = { code: 'F5', keyCode: 116 };
    Key.F6 = { code: 'F6', keyCode: 117 };
    Key.F7 = { code: 'F7', keyCode: 118 };
    Key.F8 = { code: 'F8', keyCode: 119 };
    Key.F9 = { code: 'F9', keyCode: 120 };
    Key.F10 = { code: 'F10', keyCode: 121 };
    Key.F11 = { code: 'F11', keyCode: 122 };
    Key.F12 = { code: 'F12', keyCode: 123 };
    Key.F13 = { code: 'F13', keyCode: 124 };
    Key.F14 = { code: 'F14', keyCode: 125 };
    Key.F15 = { code: 'F15', keyCode: 126 };
    Key.F16 = { code: 'F16', keyCode: 127 };
    Key.F17 = { code: 'F17', keyCode: 128 };
    Key.F18 = { code: 'F18', keyCode: 129 };
    Key.F19 = { code: 'F19', keyCode: 130 };
    Key.F20 = { code: 'F20', keyCode: 131 };
    Key.F21 = { code: 'F21', keyCode: 132 };
    Key.F22 = { code: 'F22', keyCode: 133 };
    Key.F23 = { code: 'F23', keyCode: 134 };
    Key.F24 = { code: 'F24', keyCode: 135 };
    Key.NUM_LOCK = { code: 'NumLock', keyCode: 144 };
    Key.SEMICOLON = { code: 'Semicolon', keyCode: 186 };
    Key.EQUAL = { code: 'Equal', keyCode: 187 };
    Key.COMMA = { code: 'Comma', keyCode: 188 };
    Key.MINUS = { code: 'Minus', keyCode: 189 };
    Key.PERIOD = { code: 'Period', keyCode: 190 };
    Key.SLASH = { code: 'Slash', keyCode: 191 };
    Key.BACKQUOTE = { code: 'Backquote', keyCode: 192 };
    Key.INTL_RO = { code: 'IntlRo', keyCode: 193 };
    Key.BRACKET_LEFT = { code: 'BracketLeft', keyCode: 219 };
    Key.BACKSLASH = { code: 'Backslash', keyCode: 220 };
    Key.BRACKET_RIGHT = { code: 'BracketRight', keyCode: 221 };
    Key.QUOTE = { code: 'Quote', keyCode: 222 };
    Key.INTL_YEN = { code: 'IntlYen', keyCode: 255 };
})(Key = exports.Key || (exports.Key = {}));
/*-------------------- Initialize the static key mappings --------------------*/
(function () {
    // Set the default key mappings from the constants in the Key namespace
    Object.keys(Key).map(function (prop) { return Reflect.get(Key, prop); }).filter(function (key) { return Key.isKey(key); }).forEach(function (key) {
        CODE_TO_KEY[key.code] = key;
        KEY_CODE_TO_KEY[key.keyCode] = key;
    });
    // Set additional key mappings
    CODE_TO_KEY['Numpad0'] = Key.DIGIT0;
    KEY_CODE_TO_KEY[96] = Key.DIGIT0;
    CODE_TO_KEY['Numpad1'] = Key.DIGIT1;
    KEY_CODE_TO_KEY[97] = Key.DIGIT1;
    CODE_TO_KEY['Numpad2'] = Key.DIGIT2;
    KEY_CODE_TO_KEY[98] = Key.DIGIT2;
    CODE_TO_KEY['Numpad3'] = Key.DIGIT3;
    KEY_CODE_TO_KEY[99] = Key.DIGIT3;
    CODE_TO_KEY['Numpad4'] = Key.DIGIT4;
    KEY_CODE_TO_KEY[100] = Key.DIGIT4;
    CODE_TO_KEY['Numpad5'] = Key.DIGIT5;
    KEY_CODE_TO_KEY[101] = Key.DIGIT5;
    CODE_TO_KEY['Numpad6'] = Key.DIGIT6;
    KEY_CODE_TO_KEY[102] = Key.DIGIT6;
    CODE_TO_KEY['Numpad7'] = Key.DIGIT7;
    KEY_CODE_TO_KEY[103] = Key.DIGIT7;
    CODE_TO_KEY['Numpad8'] = Key.DIGIT8;
    KEY_CODE_TO_KEY[104] = Key.DIGIT8;
    CODE_TO_KEY['Numpad9'] = Key.DIGIT9;
    KEY_CODE_TO_KEY[105] = Key.DIGIT9;
    CODE_TO_KEY['NumpadEnter'] = Key.ENTER;
    CODE_TO_KEY['NumpadEqual'] = Key.EQUAL;
    CODE_TO_KEY['IntlBackslash'] = Key.BACKSLASH;
    CODE_TO_KEY['MetaLeft'] = Key.OS_LEFT; // Chrome, Safari
    KEY_CODE_TO_KEY[224] = Key.OS_LEFT; // Firefox on Mac
    CODE_TO_KEY['MetaRight'] = Key.OS_RIGHT; // Chrome, Safari
    KEY_CODE_TO_KEY[93] = Key.OS_RIGHT; // Chrome, Safari, Edge
    KEY_CODE_TO_KEY[225] = Key.ALT_RIGHT; // Linux
    KEY_CODE_TO_KEY[110] = Key.DECIMAL; // Mac, Windows
    KEY_CODE_TO_KEY[59] = Key.SEMICOLON; // Firefox
    KEY_CODE_TO_KEY[61] = Key.EQUAL; // Firefox
    KEY_CODE_TO_KEY[173] = Key.MINUS; // Firefox
    KEY_CODE_TO_KEY[226] = Key.BACKSLASH; // Chrome, Edge on Windows
    KEY_CODE_TO_KEY[60] = Key.BACKSLASH; // Firefox on Linux
    // Set the default easy key mappings from the constants in the EasyKey namespace
    Object.keys(EasyKey).map(function (prop) { return Reflect.get(EasyKey, prop); }).forEach(function (easyKey) {
        EASY_TO_KEY[easyKey.easyString] = KEY_CODE_TO_KEY[easyKey.keyCode];
        KEY_CODE_TO_EASY[easyKey.keyCode] = easyKey;
    });
    // Set additional easy key mappings
    KEY_CODE_TO_EASY[91] = EasyKey.OS;
    // Set the modifier keys
    MODIFIERS.push.apply(MODIFIERS, __spread([Key.ALT_LEFT, Key.ALT_RIGHT, Key.CONTROL_LEFT, Key.CONTROL_RIGHT, Key.OS_LEFT, Key.OS_RIGHT, Key.SHIFT_LEFT, Key.SHIFT_RIGHT]));
})();
var KeysOrKeyCodes;
(function (KeysOrKeyCodes) {
    KeysOrKeyCodes.toKeyCode = function (keyOrKeyCode) {
        return keyOrKeyCode instanceof KeyCode ? keyOrKeyCode : KeyCode.createKeyCode({ first: keyOrKeyCode });
    };
    KeysOrKeyCodes.toKeyCodes = function (keysOrKeyCodes) {
        if (keysOrKeyCodes instanceof KeyCode) {
            return [keysOrKeyCodes];
        }
        else if (Array.isArray(keysOrKeyCodes)) {
            return keysOrKeyCodes.slice().map(KeysOrKeyCodes.toKeyCode);
        }
        return [KeysOrKeyCodes.toKeyCode(keysOrKeyCodes)];
    };
})(KeysOrKeyCodes = exports.KeysOrKeyCodes || (exports.KeysOrKeyCodes = {}));
