// @theia/core/lib/browser/view-container.js
"use strict";
/********************************************************************************
 * Copyright (C) 2018 TypeFox and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var widgets_1 = require("./widgets");
var disposable_1 = require("../common/disposable");
var ViewContainer = /** @class */ (function (_super) {
    __extends(ViewContainer, _super);
    function ViewContainer() {
        var _this = _super.call(this) || this;
        _this.widgets = new Set();
        _this.addClass('theia-view-container');
        return _this;
    }
    ViewContainer.prototype.render = function () {
        var _this = this;
        var parts = [];
        this.widgets.forEach(function (widget) { return parts.push(_this.renderPart(widget)); });
        return React.createElement(React.Fragment, null, parts);
    };
    ViewContainer.prototype.renderPart = function (widget) {
        return React.createElement(ViewContainerPart, { key: widget.id, widget: widget });
    };
    ViewContainer.prototype.addWidget = function (widget) {
        var _this = this;
        if (this.widgets.has(widget)) {
            return disposable_1.Disposable.NULL;
        }
        this.widgets.add(widget);
        this.update();
        return disposable_1.Disposable.create(function () { return _this.removeWidget(widget); });
    };
    ViewContainer.prototype.removeWidget = function (widget) {
        if (!this.widgets.delete(widget)) {
            return false;
        }
        this.update();
        return true;
    };
    ViewContainer.prototype.onResize = function (msg) {
        _super.prototype.onResize.call(this, msg);
        this.widgets.forEach(function (widget) { return widgets_1.MessageLoop.sendMessage(widget, widgets_1.Widget.ResizeMessage.UnknownSize); });
    };
    ViewContainer.prototype.onUpdateRequest = function (msg) {
        this.widgets.forEach(function (widget) { return widget.update(); });
        _super.prototype.onUpdateRequest.call(this, msg);
    };
    ViewContainer.prototype.onActivateRequest = function (msg) {
        _super.prototype.onActivateRequest.call(this, msg);
        var widget = this.widgets.values().next().value;
        if (widget) {
            widget.activate();
        }
    };
    return ViewContainer;
}(widgets_1.ReactWidget));
exports.ViewContainer = ViewContainer;
var ViewContainerPart = /** @class */ (function (_super) {
    __extends(ViewContainerPart, _super);
    function ViewContainerPart(props) {
        var _this = _super.call(this, props) || this;
        _this.detaching = false;
        _this.toggle = function () {
            if (_this.state.expanded) {
                widgets_1.Widget.detach(_this.props.widget);
            }
            _this.setState({
                expanded: !_this.state.expanded
            });
        };
        _this.setRef = function (ref) {
            var widget = _this.props.widget;
            if (ref) {
                widgets_1.MessageLoop.sendMessage(widget, widgets_1.Widget.Msg.BeforeAttach);
                // tslint:disable:no-null-keyword
                ref.insertBefore(widget.node, null);
                widgets_1.MessageLoop.sendMessage(widget, widgets_1.Widget.Msg.AfterAttach);
                widget.update();
            }
            else if (_this.detaching) {
                _this.detaching = false;
                widgets_1.MessageLoop.sendMessage(widget, widgets_1.Widget.Msg.AfterDetach);
            }
        };
        _this.state = {
            expanded: true
        };
        return _this;
    }
    ViewContainerPart.prototype.componentWillUnmount = function () {
        var widget = this.props.widget;
        if (widget.isAttached) {
            this.detaching = true;
            widgets_1.MessageLoop.sendMessage(widget, widgets_1.Widget.Msg.BeforeDetach);
        }
    };
    ViewContainerPart.prototype.render = function () {
        var widget = this.props.widget;
        var toggleClassNames = [widgets_1.EXPANSION_TOGGLE_CLASS];
        if (!this.state.expanded) {
            toggleClassNames.push(widgets_1.COLLAPSED_CLASS);
        }
        var toggleClassName = toggleClassNames.join(' ');
        return React.createElement("div", { className: 'theia-view-container-part' },
            React.createElement("div", { className: 'theia-header theia-view-container-part-head', title: widget.title.caption, onClick: this.toggle },
                React.createElement("span", { className: toggleClassName }),
                React.createElement("span", { className: 'label' }, widget.title.label),
                this.state.expanded && this.renderToolbar()),
            this.state.expanded && React.createElement("div", { className: 'theia-view-container-part-body', ref: this.setRef }));
    };
    ViewContainerPart.prototype.renderToolbar = function () {
        var _this = this;
        var widget = this.props.widget;
        if (!ViewContainerPartWidget.is(widget)) {
            return undefined;
        }
        return React.createElement(React.Fragment, null, widget.toolbarElements.map(function (element, key) { return _this.renderToolbarElement(key, element); }));
    };
    ViewContainerPart.prototype.renderToolbarElement = function (key, element) {
        var _this = this;
        if (element.enabled === false) {
            return undefined;
        }
        var className = element.className, tooltip = element.tooltip, execute = element.execute;
        var classNames = ['element'];
        if (className) {
            classNames.push(className);
        }
        return React.createElement("span", { key: key, title: tooltip, className: classNames.join(' '), onClick: function (e) { return __awaiter(_this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            e.stopPropagation();
                            e.preventDefault();
                            return [4 /*yield*/, execute()];
                        case 1:
                            _a.sent();
                            this.forceUpdate();
                            return [2 /*return*/];
                    }
                });
            }); } });
    };
    return ViewContainerPart;
}(React.Component));
exports.ViewContainerPart = ViewContainerPart;
var ViewContainerPartWidget;
(function (ViewContainerPartWidget) {
    function is(widget) {
        return !!widget && ('toolbarElements' in widget);
    }
    ViewContainerPartWidget.is = is;
})(ViewContainerPartWidget = exports.ViewContainerPartWidget || (exports.ViewContainerPartWidget = {}));
