// @theia/core/lib/browser/widget-open-handler.js
"use strict";
/********************************************************************************
 * Copyright (C) 2018 TypeFox and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var inversify_1 = require("inversify");
var common_1 = require("../common");
var shell_1 = require("./shell");
var widget_manager_1 = require("./widget-manager");
var WidgetOpenHandler = /** @class */ (function () {
    function WidgetOpenHandler() {
        this.onCreatedEmitter = new common_1.Emitter();
        /**
         * Emit when a new widget is created.
         */
        this.onCreated = this.onCreatedEmitter.event;
    }
    WidgetOpenHandler.prototype.init = function () {
        var _this = this;
        this.widgetManager.onDidCreateWidget(function (_a) {
            var factoryId = _a.factoryId, widget = _a.widget;
            if (factoryId === _this.id) {
                _this.onCreatedEmitter.fire(widget);
            }
        });
    };
    /**
     * Open a widget for the given uri and options.
     * Reject if the given options is not an widget options or a widget cannot be opened.
     */
    WidgetOpenHandler.prototype.open = function (uri, options) {
        return __awaiter(this, void 0, void 0, function () {
            var widget;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.getOrCreateWidget(uri, options)];
                    case 1:
                        widget = _a.sent();
                        return [4 /*yield*/, this.doOpen(widget, options)];
                    case 2:
                        _a.sent();
                        return [2 /*return*/, widget];
                }
            });
        });
    };
    WidgetOpenHandler.prototype.doOpen = function (widget, options) {
        return __awaiter(this, void 0, void 0, function () {
            var op, promises;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        op = __assign({ mode: 'activate' }, options);
                        if (!widget.isAttached) {
                            this.shell.addWidget(widget, op.widgetOptions || { area: 'main' });
                        }
                        promises = [];
                        if (op.mode === 'activate') {
                            promises.push(this.onActive(widget));
                            promises.push(this.onReveal(widget));
                            this.shell.activateWidget(widget.id);
                        }
                        else if (op.mode === 'reveal') {
                            promises.push(this.onReveal(widget));
                            this.shell.revealWidget(widget.id);
                        }
                        return [4 /*yield*/, Promise.all(promises)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    WidgetOpenHandler.prototype.onActive = function (widget) {
        var _this = this;
        if (this.shell.activeWidget === widget) {
            return Promise.resolve();
        }
        return new Promise(function (resolve) {
            var listener = function (shell, args) {
                if (args.newValue === widget) {
                    _this.shell.activeChanged.disconnect(listener);
                    resolve();
                }
            };
            _this.shell.activeChanged.connect(listener);
        });
    };
    WidgetOpenHandler.prototype.onReveal = function (widget) {
        if (widget.isVisible) {
            return new Promise(function (resolve) { return window.requestAnimationFrame(function () { return resolve(); }); });
        }
        return new Promise(function (resolve) {
            var waitForVisible = function () { return window.requestAnimationFrame(function () {
                if (widget.isVisible) {
                    window.requestAnimationFrame(function () { return resolve(); });
                }
                else {
                    waitForVisible();
                }
            }); };
            waitForVisible();
        });
    };
    /**
     * Return an existing widget for the given uri.
     */
    WidgetOpenHandler.prototype.getByUri = function (uri) {
        return this.getWidget(uri);
    };
    /**
     * Return an existing widget for the given uri or creates a new one.
     *
     * It does not open a widget, use `open` instead.
     */
    WidgetOpenHandler.prototype.getOrCreateByUri = function (uri) {
        return this.getOrCreateWidget(uri);
    };
    Object.defineProperty(WidgetOpenHandler.prototype, "all", {
        /**
         * All opened widgets.
         */
        get: function () {
            return this.widgetManager.getWidgets(this.id);
        },
        enumerable: true,
        configurable: true
    });
    WidgetOpenHandler.prototype.getWidget = function (uri, options) {
        var widgetOptions = this.createWidgetOptions(uri, options);
        return this.widgetManager.getWidget(this.id, widgetOptions);
    };
    WidgetOpenHandler.prototype.getOrCreateWidget = function (uri, options) {
        var widgetOptions = this.createWidgetOptions(uri, options);
        return this.widgetManager.getOrCreateWidget(this.id, widgetOptions);
    };
    __decorate([
        inversify_1.inject(shell_1.ApplicationShell),
        __metadata("design:type", shell_1.ApplicationShell)
    ], WidgetOpenHandler.prototype, "shell", void 0);
    __decorate([
        inversify_1.inject(widget_manager_1.WidgetManager),
        __metadata("design:type", widget_manager_1.WidgetManager)
    ], WidgetOpenHandler.prototype, "widgetManager", void 0);
    __decorate([
        inversify_1.postConstruct(),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", void 0)
    ], WidgetOpenHandler.prototype, "init", null);
    WidgetOpenHandler = __decorate([
        inversify_1.injectable()
    ], WidgetOpenHandler);
    return WidgetOpenHandler;
}());
exports.WidgetOpenHandler = WidgetOpenHandler;
