// monaco-languageclient/dist/amd/console-window.js
define(["require", "exports", "./services"], function (require, exports, services_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var ConsoleWindow = /** @class */ (function () {
        function ConsoleWindow() {
            this.channels = new Map();
        }
        ConsoleWindow.prototype.showMessage = function (type, message) {
            var actions = [];
            for (var _i = 2; _i < arguments.length; _i++) {
                actions[_i - 2] = arguments[_i];
            }
            if (type === services_1.MessageType.Error) {
                console.error(message);
            }
            if (type === services_1.MessageType.Warning) {
                console.warn(message);
            }
            if (type === services_1.MessageType.Info) {
                console.info(message);
            }
            if (type === services_1.MessageType.Log) {
                console.log(message);
            }
            return Promise.resolve(undefined);
        };
        ConsoleWindow.prototype.createOutputChannel = function (name) {
            var existing = this.channels.get(name);
            if (existing) {
                return existing;
            }
            var channel = {
                append: function (value) {
                    console.log(name + ': ' + value);
                },
                appendLine: function (line) {
                    console.log(name + ': ' + line);
                },
                show: function () {
                    // no-op
                },
                dispose: function () {
                    // no-op
                }
            };
            this.channels.set(name, channel);
            return channel;
        };
        return ConsoleWindow;
    }());
    exports.ConsoleWindow = ConsoleWindow;
});
