// vscode-ws-jsonrpc/lib/server/launch.js
"use strict";
/* --------------------------------------------------------------------------------------------
 * Copyright (c) 2018 TypeFox GmbH (http://www.typefox.io). All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for license information.
 * ------------------------------------------------------------------------------------------ */
Object.defineProperty(exports, "__esModule", { value: true });
var cp = require("child_process");
var vscode_jsonrpc_1 = require("vscode-jsonrpc");
var connection_1 = require("./connection");
var socket_1 = require("../socket");
function createServerProcess(serverName, command, args, options) {
    var serverProcess = cp.spawn(command, args, options);
    serverProcess.on('error', function (error) {
        return console.error("Launching " + serverName + " Server failed: " + error);
    });
    serverProcess.stderr.on('data', function (data) {
        return console.error(serverName + " Server: " + data);
    });
    return createProcessStreamConnection(serverProcess);
}
exports.createServerProcess = createServerProcess;
function createWebSocketConnection(socket) {
    var reader = new socket_1.WebSocketMessageReader(socket);
    var writer = new socket_1.WebSocketMessageWriter(socket);
    return connection_1.createConnection(reader, writer, function () { return socket.dispose(); });
}
exports.createWebSocketConnection = createWebSocketConnection;
function createProcessSocketConnection(process, outSocket, inSocket) {
    if (inSocket === void 0) { inSocket = outSocket; }
    return createSocketConnection(outSocket, inSocket, function () { return process.kill(); });
}
exports.createProcessSocketConnection = createProcessSocketConnection;
function createSocketConnection(outSocket, inSocket, onDispose) {
    var reader = new vscode_jsonrpc_1.SocketMessageReader(outSocket);
    var writer = new vscode_jsonrpc_1.SocketMessageWriter(inSocket);
    return connection_1.createConnection(reader, writer, onDispose);
}
exports.createSocketConnection = createSocketConnection;
function createProcessStreamConnection(process) {
    return createStreamConnection(process.stdout, process.stdin, function () { return process.kill(); });
}
exports.createProcessStreamConnection = createProcessStreamConnection;
function createStreamConnection(outStream, inStream, onDispose) {
    var reader = new vscode_jsonrpc_1.StreamMessageReader(outStream);
    var writer = new vscode_jsonrpc_1.StreamMessageWriter(inStream);
    return connection_1.createConnection(reader, writer, onDispose);
}
exports.createStreamConnection = createStreamConnection;
