// @theia/core/lib/browser/tree/tree-widget.js
"use strict";
/********************************************************************************
 * Copyright (C) 2018 TypeFox and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __values = (this && this.__values) || function (o) {
    var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
    if (m) return m.call(o);
    return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spread = (this && this.__spread) || function () {
    for (var ar = [], i = 0; i < arguments.length; i++) ar = ar.concat(__read(arguments[i]));
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
var inversify_1 = require("inversify");
var common_1 = require("../../common");
var keys_1 = require("../keys");
var context_menu_renderer_1 = require("../context-menu-renderer");
var widgets_1 = require("../widgets");
var tree_1 = require("./tree");
var tree_model_1 = require("./tree-model");
var tree_expansion_1 = require("./tree-expansion");
var tree_selection_1 = require("./tree-selection");
var tree_decorator_1 = require("./tree-decorator");
var objects_1 = require("../../common/objects");
var os_1 = require("../../common/os");
var react_widget_1 = require("../widgets/react-widget");
var React = require("react");
var react_virtualized_1 = require("react-virtualized");
var tree_iterator_1 = require("./tree-iterator");
var search_box_1 = require("./search-box");
var tree_search_1 = require("./tree-search");
var domutils_1 = require("@phosphor/domutils");
var debounce = require('lodash.debounce');
exports.TREE_CLASS = 'theia-Tree';
exports.TREE_CONTAINER_CLASS = 'theia-TreeContainer';
exports.TREE_NODE_CLASS = 'theia-TreeNode';
exports.TREE_NODE_CONTENT_CLASS = 'theia-TreeNodeContent';
exports.TREE_NODE_TAIL_CLASS = 'theia-TreeNodeTail';
exports.TREE_NODE_SEGMENT_CLASS = 'theia-TreeNodeSegment';
exports.TREE_NODE_SEGMENT_GROW_CLASS = 'theia-TreeNodeSegmentGrow';
exports.EXPANDABLE_TREE_NODE_CLASS = 'theia-ExpandableTreeNode';
exports.COMPOSITE_TREE_NODE_CLASS = 'theia-CompositeTreeNode';
exports.TREE_NODE_CAPTION_CLASS = 'theia-TreeNodeCaption';
exports.TreeProps = Symbol('TreeProps');
exports.defaultTreeProps = {
    leftPadding: 16
};
var TreeWidget = /** @class */ (function (_super) {
    __extends(TreeWidget, _super);
    function TreeWidget(props, model, contextMenuRenderer) {
        var _this = _super.call(this) || this;
        _this.props = props;
        _this.model = model;
        _this.contextMenuRenderer = contextMenuRenderer;
        _this.decorations = new Map();
        _this.rows = new Map();
        _this.updateRows = debounce(function () { return _this.doUpdateRows(); }, 10);
        _this.updateDecorations = debounce(function () { return _this.doUpdateDecorations(); }, 150);
        _this.scrollArea = _this.node;
        _this.handleScroll = function (info) {
            _this.node.scrollTop = info.scrollTop;
        };
        _this.renderNodeRow = function (row) { return _this.doRenderNodeRow(row); };
        _this.toggle = function (event) { return _this.doToggle(event); };
        _this.scrollOptions = {
            suppressScrollX: true
        };
        _this.addClass(exports.TREE_CLASS);
        _this.node.tabIndex = 0;
        return _this;
    }
    TreeWidget_1 = TreeWidget;
    TreeWidget.prototype.init = function () {
        var _this = this;
        if (this.props.search) {
            this.searchBox = this.searchBoxFactory(search_box_1.SearchBoxProps.DEFAULT);
            this.toDispose.pushAll([
                this.searchBox,
                this.searchBox.onTextChange(function (data) { return __awaiter(_this, void 0, void 0, function () {
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0: return [4 /*yield*/, this.treeSearch.filter(data)];
                            case 1:
                                _a.sent();
                                this.searchHighlights = this.treeSearch.getHighlights();
                                this.update();
                                return [2 /*return*/];
                        }
                    });
                }); }),
                this.searchBox.onClose(function (data) { return _this.treeSearch.filter(undefined); }),
                this.searchBox.onNext(function () { return _this.model.selectNextNode(); }),
                this.searchBox.onPrevious(function () { return _this.model.selectPrevNode(); }),
                this.treeSearch,
                this.treeSearch.onFilteredNodesChanged(function (nodes) {
                    var node = nodes.find(tree_selection_1.SelectableTreeNode.is);
                    if (node) {
                        _this.model.selectNode(node);
                    }
                }),
                this.model.onExpansionChanged(function () {
                    _this.searchBox.hide();
                })
            ]);
        }
        this.toDispose.pushAll([
            this.model,
            this.model.onChanged(function () { return _this.updateRows(); }),
            this.model.onSelectionChanged(function () { return _this.updateScrollToRow({ resize: false }); }),
            this.model.onNodeRefreshed(function () { return _this.updateDecorations(); }),
            this.model.onExpansionChanged(function () { return _this.updateDecorations(); }),
            this.decoratorService,
            this.decoratorService.onDidChangeDecorations(function () { return _this.updateDecorations(); })
        ]);
        setTimeout(function () {
            _this.updateRows();
            _this.updateDecorations();
        });
    };
    TreeWidget.prototype.doUpdateRows = function () {
        var e_1, _a;
        var root = this.model.root;
        var rowsToUpdate = [];
        if (root) {
            var depths = new Map();
            var index = 0;
            try {
                for (var _b = __values(new tree_iterator_1.TopDownTreeIterator(root, {
                    pruneCollapsed: true,
                    pruneSiblings: true
                })), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var node = _c.value;
                    if (tree_1.TreeNode.isVisible(node)) {
                        var parentDepth = depths.get(node.parent);
                        var depth = parentDepth === undefined ? 0 : tree_1.TreeNode.isVisible(node.parent) ? parentDepth + 1 : parentDepth;
                        if (tree_1.CompositeTreeNode.is(node)) {
                            depths.set(node, depth);
                        }
                        rowsToUpdate.push([node.id, {
                                index: index++,
                                node: node,
                                depth: depth
                            }]);
                    }
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_1) throw e_1.error; }
            }
        }
        this.rows = new Map(rowsToUpdate);
        this.updateScrollToRow();
    };
    TreeWidget.prototype.updateScrollToRow = function (updateOptions) {
        var selected = this.model.selectedNodes;
        var node = selected.find(tree_selection_1.SelectableTreeNode.hasFocus) || selected[0];
        var row = node && this.rows.get(node.id);
        this.scrollToRow = row && row.index;
        this.forceUpdate(updateOptions);
    };
    TreeWidget.prototype.doUpdateDecorations = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _a = this;
                        return [4 /*yield*/, this.decoratorService.getDecorations(this.model)];
                    case 1:
                        _a.decorations = _b.sent();
                        this.forceUpdate();
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Force deep resizing and rendering of rows.
     * https://github.com/bvaughn/react-virtualized/blob/master/docs/List.md#recomputerowheights-index-number
     */
    TreeWidget.prototype.forceUpdate = function (_a) {
        var resize = (_a === void 0 ? { resize: true } : _a).resize;
        if (this.view && this.view.list) {
            if (resize && this.isVisible) {
                this.view.cache.clearAll();
                this.view.list.recomputeRowHeights();
            }
            else {
                this.view.list.forceUpdateGrid();
            }
        }
        this.update();
    };
    TreeWidget.prototype.onActivateRequest = function (msg) {
        _super.prototype.onActivateRequest.call(this, msg);
        this.node.focus();
        if (this.model.selectedNodes.length === 0) {
            var root = this.model.root;
            if (tree_selection_1.SelectableTreeNode.is(root)) {
                this.model.selectNode(root);
            }
            else if (tree_1.CompositeTreeNode.is(root) && root.children.length >= 1) {
                var firstChild = root.children[0];
                if (tree_selection_1.SelectableTreeNode.is(firstChild)) {
                    this.model.selectNode(firstChild);
                }
            }
        }
        this.forceUpdate();
    };
    TreeWidget.prototype.onUpdateRequest = function (msg) {
        if (!this.isAttached || !this.isVisible) {
            return;
        }
        _super.prototype.onUpdateRequest.call(this, msg);
    };
    TreeWidget.prototype.onResize = function (msg) {
        _super.prototype.onResize.call(this, msg);
        this.forceUpdate();
    };
    TreeWidget.prototype.render = function () {
        return React.createElement('div', this.createContainerAttributes(), this.renderTree(this.model));
    };
    TreeWidget.prototype.createContainerAttributes = function () {
        var _this = this;
        return {
            className: exports.TREE_CONTAINER_CLASS,
            onContextMenu: function (event) { return _this.handleContextMenuEvent(_this.getContainerTreeNode(), event); }
        };
    };
    TreeWidget.prototype.getContainerTreeNode = function () {
        return this.model.root;
    };
    TreeWidget.prototype.renderTree = function (model) {
        var _this = this;
        if (model.root) {
            var rows = Array.from(this.rows.values());
            if (this.props.virtualized === false) {
                this.onRender.push(common_1.Disposable.create(function () { return _this.scrollToSelected(); }));
                return rows.map(function (row) { return React.createElement("div", { key: row.index }, _this.renderNodeRow(row)); });
            }
            return React.createElement(TreeWidget_1.View, { ref: function (view) { return _this.view = (view || undefined); }, width: this.node.offsetWidth, height: this.node.offsetHeight, rows: rows, renderNodeRow: this.renderNodeRow, scrollToRow: this.scrollToRow, handleScroll: this.handleScroll });
        }
        // tslint:disable-next-line:no-null-keyword
        return null;
    };
    TreeWidget.prototype.scrollToSelected = function () {
        if (this.props.scrollIfActive === true && !this.node.contains(document.activeElement)) {
            return;
        }
        var focus = this.node.getElementsByClassName(widgets_1.FOCUS_CLASS)[0];
        if (focus) {
            domutils_1.ElementExt.scrollIntoViewIfNeeded(this.scrollArea, focus);
        }
        else {
            var selected = this.node.getElementsByClassName(widgets_1.SELECTED_CLASS)[0];
            if (selected) {
                domutils_1.ElementExt.scrollIntoViewIfNeeded(this.scrollArea, selected);
            }
        }
    };
    TreeWidget.prototype.doRenderNodeRow = function (_a) {
        var index = _a.index, node = _a.node, depth = _a.depth;
        return this.renderNode(node, { depth: depth });
    };
    TreeWidget.prototype.renderIcon = function (node, props) {
        // tslint:disable-next-line:no-null-keyword
        return null;
    };
    TreeWidget.prototype.doToggle = function (event) {
        var nodeId = event.currentTarget.getAttribute('data-node-id');
        if (nodeId) {
            var node = this.model.getNode(nodeId);
            this.handleClickEvent(node, event);
        }
        event.stopPropagation();
    };
    TreeWidget.prototype.renderExpansionToggle = function (node, props) {
        if (!this.isExpandable(node)) {
            // tslint:disable-next-line:no-null-keyword
            return null;
        }
        var classes = [exports.TREE_NODE_SEGMENT_CLASS, widgets_1.EXPANSION_TOGGLE_CLASS];
        if (!node.expanded) {
            classes.push(widgets_1.COLLAPSED_CLASS);
        }
        var className = classes.join(' ');
        return React.createElement("div", { "data-node-id": node.id, className: className, style: { paddingLeft: '4px' }, onClick: this.toggle });
    };
    TreeWidget.prototype.renderCaption = function (node, props) {
        var tooltip = this.getDecorationData(node, 'tooltip').filter(objects_1.notEmpty).join(' • ');
        var classes = [exports.TREE_NODE_SEGMENT_CLASS];
        if (!this.hasTrailingSuffixes(node)) {
            classes.push(exports.TREE_NODE_SEGMENT_GROW_CLASS);
        }
        var className = classes.join(' ');
        var attrs = this.decorateCaption(node, {
            className: className, id: node.id
        });
        if (tooltip.length > 0) {
            attrs = __assign({}, attrs, { title: tooltip });
        }
        var children = [];
        var caption = node.name;
        var highlight = this.getDecorationData(node, 'highlight')[0];
        if (highlight) {
            children.push(this.toReactNode(caption, highlight));
        }
        var searchHighlight = this.searchHighlights ? this.searchHighlights.get(node.id) : undefined;
        if (searchHighlight) {
            children.push.apply(children, __spread(this.toReactNode(caption, searchHighlight)));
        }
        else if (!highlight) {
            children.push(caption);
        }
        return React.createElement.apply(React, __spread(['div', attrs], children));
    };
    TreeWidget.prototype.toReactNode = function (caption, highlight) {
        var style = {};
        if (highlight.color) {
            style = __assign({}, style, { color: highlight.color });
        }
        if (highlight.backgroundColor) {
            style = __assign({}, style, { backgroundColor: highlight.backgroundColor });
        }
        var createChildren = function (fragment) {
            var data = fragment.data;
            if (fragment.highligh) {
                return React.createElement("mark", { className: tree_decorator_1.TreeDecoration.Styles.CAPTION_HIGHLIGHT_CLASS, style: style }, data);
            }
            else {
                return data;
            }
        };
        return tree_decorator_1.TreeDecoration.CaptionHighlight.split(caption, highlight).map(createChildren);
    };
    TreeWidget.prototype.decorateCaption = function (node, attrs) {
        var _this = this;
        var style = this.getDecorationData(node, 'fontData').filter(objects_1.notEmpty).reverse().map(function (fontData) { return _this.applyFontStyles({}, fontData); }).reduce(function (acc, current) {
            return (__assign({}, acc, current));
        }, {});
        return __assign({}, attrs, { style: style });
    };
    TreeWidget.prototype.hasTrailingSuffixes = function (node) {
        return this.getDecorationData(node, 'captionSuffixes').filter(objects_1.notEmpty).reduce(function (acc, current) { return acc.concat(current); }, []).length > 0;
    };
    TreeWidget.prototype.applyFontStyles = function (original, fontData) {
        if (fontData === undefined) {
            return original;
        }
        var modified = original;
        var color = fontData.color, style = fontData.style;
        if (color) {
            modified = __assign({}, modified, { color: color });
        }
        if (style) {
            (Array.isArray(style) ? style : [style]).forEach(function (s) {
                switch (style) {
                    case 'bold':
                        modified = __assign({}, modified, { fontWeight: style });
                        break;
                    case 'normal': // Fall through.
                    case 'oblique': // Fall through.
                    case 'italic':
                        modified = __assign({}, modified, { fontStyle: style });
                        break;
                    case 'underline': // Fall through.
                    case 'line-through':
                        modified = __assign({}, modified, { textDecoration: style });
                        break;
                    default:
                        throw new Error("Unexpected font style: " + style + ".");
                }
            });
        }
        return modified;
    };
    TreeWidget.prototype.renderCaptionAffixes = function (node, props, affixKey) {
        var suffix = affixKey === 'captionSuffixes';
        var affixClass = suffix ? tree_decorator_1.TreeDecoration.Styles.CAPTION_SUFFIX_CLASS : tree_decorator_1.TreeDecoration.Styles.CAPTION_PREFIX_CLASS;
        var classes = [exports.TREE_NODE_SEGMENT_CLASS, affixClass];
        var affixes = this.getDecorationData(node, affixKey).filter(objects_1.notEmpty).reduce(function (acc, current) { return acc.concat(current); }, []);
        var children = [];
        for (var i = 0; i < affixes.length; i++) {
            var affix = affixes[i];
            if (suffix && i === affixes.length - 1) {
                classes.push(exports.TREE_NODE_SEGMENT_GROW_CLASS);
            }
            var style = this.applyFontStyles({}, affix.fontData);
            var className = classes.join(' ');
            var key = node.id + '_' + i;
            var attrs = {
                className: className,
                style: style,
                key: key
            };
            children.push(React.createElement('div', attrs, affix.data));
        }
        return React.createElement(React.Fragment, null, children);
    };
    TreeWidget.prototype.decorateIcon = function (node, icon) {
        if (icon === null) {
            // tslint:disable-next-line:no-null-keyword
            return null;
        }
        var overlayIcons = [];
        new Map(this.getDecorationData(node, 'iconOverlay').reverse().filter(objects_1.notEmpty)
            .map(function (overlay) { return [overlay.position, overlay]; }))
            .forEach(function (overlay, position) {
            var overlayClass = function (iconName) {
                return ['a', 'fa', "fa-" + iconName, tree_decorator_1.TreeDecoration.Styles.DECORATOR_SIZE_CLASS, tree_decorator_1.TreeDecoration.IconOverlayPosition.getStyle(position)].join(' ');
            };
            var style = function (color) { return color === undefined ? {} : { color: color }; };
            if (overlay.background) {
                overlayIcons.push(React.createElement("span", { key: node.id + 'bg', className: overlayClass(overlay.background.shape), style: style(overlay.background.color) }));
            }
            overlayIcons.push(React.createElement("span", { key: node.id, className: overlayClass(overlay.icon), style: style(overlay.color) }));
        });
        if (overlayIcons.length > 0) {
            return React.createElement("div", { className: tree_decorator_1.TreeDecoration.Styles.ICON_WRAPPER_CLASS },
                icon,
                overlayIcons);
        }
        return icon;
    };
    TreeWidget.prototype.renderTailDecorations = function (node, props) {
        var _this = this;
        var style = function (fontData) { return _this.applyFontStyles({}, fontData); };
        return React.createElement(React.Fragment, null, this.getDecorationData(node, 'tailDecorations').filter(objects_1.notEmpty).reduce(function (acc, current) { return acc.concat(current); }, []).map(function (decoration) {
            var fontData = decoration.fontData, data = decoration.data, tooltip = decoration.tooltip;
            var className = [exports.TREE_NODE_SEGMENT_CLASS, exports.TREE_NODE_TAIL_CLASS].join(' ');
            return React.createElement("div", { key: node.id + className, className: className, style: style(fontData), title: tooltip }, data);
        }));
    };
    TreeWidget.prototype.renderNode = function (node, props) {
        if (!tree_1.TreeNode.isVisible(node)) {
            return undefined;
        }
        var attributes = this.createNodeAttributes(node, props);
        var content = React.createElement("div", { className: exports.TREE_NODE_CONTENT_CLASS },
            this.renderExpansionToggle(node, props),
            this.decorateIcon(node, this.renderIcon(node, props)),
            this.renderCaptionAffixes(node, props, 'captionPrefixes'),
            this.renderCaption(node, props),
            this.renderCaptionAffixes(node, props, 'captionSuffixes'),
            this.renderTailDecorations(node, props));
        return React.createElement('div', attributes, content);
    };
    TreeWidget.prototype.createNodeAttributes = function (node, props) {
        var _this = this;
        var className = this.createNodeClassNames(node, props).join(' ');
        var style = this.createNodeStyle(node, props);
        return {
            className: className,
            style: style,
            onClick: function (event) { return _this.handleClickEvent(node, event); },
            onDoubleClick: function (event) { return _this.handleDblClickEvent(node, event); },
            onContextMenu: function (event) { return _this.handleContextMenuEvent(node, event); }
        };
    };
    TreeWidget.prototype.createNodeClassNames = function (node, props) {
        var classNames = [exports.TREE_NODE_CLASS];
        if (tree_1.CompositeTreeNode.is(node)) {
            classNames.push(exports.COMPOSITE_TREE_NODE_CLASS);
        }
        if (this.isExpandable(node)) {
            classNames.push(exports.EXPANDABLE_TREE_NODE_CLASS);
        }
        if (tree_selection_1.SelectableTreeNode.isSelected(node)) {
            classNames.push(widgets_1.SELECTED_CLASS);
        }
        if (tree_selection_1.SelectableTreeNode.hasFocus(node)) {
            classNames.push(widgets_1.FOCUS_CLASS);
        }
        return classNames;
    };
    TreeWidget.prototype.getDefaultNodeStyle = function (node, props) {
        // If the node is a composite, a toggle will be rendered. Otherwise we need to add the width and the left, right padding => 18px
        var paddingLeft = props.depth * this.props.leftPadding + (this.isExpandable(node) ? 0 : 18) + "px";
        return {
            paddingLeft: paddingLeft
        };
    };
    TreeWidget.prototype.createNodeStyle = function (node, props) {
        return this.decorateNodeStyle(node, this.getDefaultNodeStyle(node, props));
    };
    TreeWidget.prototype.decorateNodeStyle = function (node, style) {
        var backgroundColor = this.getDecorationData(node, 'backgroundColor').filter(objects_1.notEmpty).shift();
        if (backgroundColor) {
            style = __assign({}, (style || {}), { backgroundColor: backgroundColor });
        }
        return style;
    };
    TreeWidget.prototype.isExpandable = function (node) {
        return tree_expansion_1.ExpandableTreeNode.is(node);
    };
    TreeWidget.prototype.getDecorations = function (node) {
        var decorations = this.decorations.get(node.id);
        if (decorations) {
            return decorations.sort(tree_decorator_1.TreeDecoration.Data.comparePriority);
        }
        return [];
    };
    TreeWidget.prototype.getDecorationData = function (node, key) {
        return this.getDecorations(node).filter(function (data) { return data[key] !== undefined; }).map(function (data) { return data[key]; }).filter(objects_1.notEmpty);
    };
    TreeWidget.prototype.onAfterAttach = function (msg) {
        var _this = this;
        var up = [
            keys_1.Key.ARROW_UP,
            keys_1.KeyCode.createKeyCode({ first: keys_1.Key.ARROW_UP, modifiers: [keys_1.KeyModifier.Shift] })
        ];
        var down = [
            keys_1.Key.ARROW_DOWN,
            keys_1.KeyCode.createKeyCode({ first: keys_1.Key.ARROW_DOWN, modifiers: [keys_1.KeyModifier.Shift] })
        ];
        if (this.props.search) {
            if (this.searchBox.isAttached) {
                widgets_1.Widget.detach(this.searchBox);
            }
            widgets_1.Widget.attach(this.searchBox, this.node.parentElement);
            this.addKeyListener(this.node, this.searchBox.keyCodePredicate.bind(this.searchBox), this.searchBox.handle.bind(this.searchBox));
            this.toDisposeOnDetach.push(common_1.Disposable.create(function () {
                widgets_1.Widget.detach(_this.searchBox);
            }));
        }
        _super.prototype.onAfterAttach.call(this, msg);
        this.addKeyListener(this.node, keys_1.Key.ARROW_LEFT, function (event) { return _this.handleLeft(event); });
        this.addKeyListener(this.node, keys_1.Key.ARROW_RIGHT, function (event) { return _this.handleRight(event); });
        this.addKeyListener(this.node, up, function (event) { return _this.handleUp(event); });
        this.addKeyListener(this.node, down, function (event) { return _this.handleDown(event); });
        this.addKeyListener(this.node, keys_1.Key.ENTER, function (event) { return _this.handleEnter(event); });
        // tslint:disable-next-line:no-any
        this.addEventListener(this.node, 'ps-scroll-y', function (e) {
            if (_this.view && _this.view.list && _this.view.list.Grid) {
                var scrollTop = e.target.scrollTop;
                _this.view.list.Grid.handleScrollEvent({ scrollTop: scrollTop });
            }
        });
    };
    TreeWidget.prototype.handleLeft = function (event) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!!this.props.multiSelect && (this.hasCtrlCmdMask(event) || this.hasShiftMask(event))) {
                            return [2 /*return*/];
                        }
                        return [4 /*yield*/, this.model.collapseNode()];
                    case 1:
                        if (!(_a.sent())) {
                            this.model.selectParent();
                        }
                        return [2 /*return*/];
                }
            });
        });
    };
    TreeWidget.prototype.handleRight = function (event) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!!this.props.multiSelect && (this.hasCtrlCmdMask(event) || this.hasShiftMask(event))) {
                            return [2 /*return*/];
                        }
                        return [4 /*yield*/, this.model.expandNode()];
                    case 1:
                        if (!(_a.sent())) {
                            this.model.selectNextNode();
                        }
                        return [2 /*return*/];
                }
            });
        });
    };
    TreeWidget.prototype.handleUp = function (event) {
        if (!!this.props.multiSelect && this.hasShiftMask(event)) {
            this.model.selectPrevNode(tree_selection_1.TreeSelection.SelectionType.RANGE);
        }
        else {
            this.model.selectPrevNode();
        }
    };
    TreeWidget.prototype.handleDown = function (event) {
        if (!!this.props.multiSelect && this.hasShiftMask(event)) {
            this.model.selectNextNode(tree_selection_1.TreeSelection.SelectionType.RANGE);
        }
        else {
            this.model.selectNextNode();
        }
    };
    TreeWidget.prototype.handleEnter = function (event) {
        this.model.openNode();
    };
    TreeWidget.prototype.handleClickEvent = function (node, event) {
        if (node) {
            if (!!this.props.multiSelect) {
                var shiftMask = this.hasShiftMask(event);
                var ctrlCmdMask = this.hasCtrlCmdMask(event);
                if (tree_selection_1.SelectableTreeNode.is(node)) {
                    if (shiftMask) {
                        this.model.selectRange(node);
                    }
                    else if (ctrlCmdMask) {
                        this.model.toggleNode(node);
                    }
                    else {
                        this.model.selectNode(node);
                    }
                }
                if (this.isExpandable(node) && !shiftMask && !ctrlCmdMask) {
                    this.model.toggleNodeExpansion(node);
                }
            }
            else {
                if (tree_selection_1.SelectableTreeNode.is(node)) {
                    this.model.selectNode(node);
                }
                if (this.isExpandable(node) && !this.hasCtrlCmdMask(event) && !this.hasShiftMask(event)) {
                    this.model.toggleNodeExpansion(node);
                }
            }
            event.stopPropagation();
        }
    };
    TreeWidget.prototype.handleDblClickEvent = function (node, event) {
        this.model.openNode(node);
        event.stopPropagation();
    };
    TreeWidget.prototype.handleContextMenuEvent = function (node, event) {
        var _this = this;
        if (tree_selection_1.SelectableTreeNode.is(node)) {
            // Keep the selection for the context menu, if the widget support multi-selection and the right click happens on an already selected node.
            if (!this.props.multiSelect || !node.selected) {
                var type = !!this.props.multiSelect && this.hasCtrlCmdMask(event) ? tree_selection_1.TreeSelection.SelectionType.TOGGLE : tree_selection_1.TreeSelection.SelectionType.DEFAULT;
                this.model.addSelection({ node: node, type: type });
            }
            var contextMenuPath_1 = this.props.contextMenuPath;
            if (contextMenuPath_1) {
                var _a = event.nativeEvent, x_1 = _a.x, y_1 = _a.y;
                this.onRender.push(common_1.Disposable.create(function () {
                    return setTimeout(function () {
                        return _this.contextMenuRenderer.render(contextMenuPath_1, { x: x_1, y: y_1 });
                    });
                }));
            }
            this.update();
        }
        event.stopPropagation();
        event.preventDefault();
    };
    TreeWidget.prototype.hasCtrlCmdMask = function (event) {
        var metaKey = event.metaKey, ctrlKey = event.ctrlKey;
        return (os_1.isOSX && metaKey) || ctrlKey;
    };
    TreeWidget.prototype.hasShiftMask = function (event) {
        // Ctrl/Cmd mask overrules the Shift mask.
        if (this.hasCtrlCmdMask(event)) {
            return false;
        }
        return event.shiftKey;
    };
    TreeWidget.prototype.deflateForStorage = function (node) {
        var e_2, _a;
        // tslint:disable-next-line:no-any
        var copy = Object.assign({}, node);
        if (copy.parent) {
            delete copy.parent;
        }
        if ('previousSibling' in copy) {
            delete copy.previousSibling;
        }
        if ('nextSibling' in copy) {
            delete copy.nextSibling;
        }
        if (tree_1.CompositeTreeNode.is(node)) {
            copy.children = [];
            try {
                for (var _b = __values(node.children), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var child = _c.value;
                    copy.children.push(this.deflateForStorage(child));
                }
            }
            catch (e_2_1) { e_2 = { error: e_2_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_2) throw e_2.error; }
            }
        }
        return copy;
    };
    // tslint:disable-next-line:no-any
    TreeWidget.prototype.inflateFromStorage = function (node, parent) {
        var e_3, _a;
        if (node.selected) {
            node.selected = false;
        }
        if (parent) {
            node.parent = parent;
        }
        if (Array.isArray(node.children)) {
            try {
                for (var _b = __values(node.children), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var child = _c.value;
                    this.inflateFromStorage(child, node);
                }
            }
            catch (e_3_1) { e_3 = { error: e_3_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_3) throw e_3.error; }
            }
        }
        return node;
    };
    TreeWidget.prototype.storeState = function () {
        var decorations = this.decoratorService.deflateDecorators(this.decorations);
        var state = {
            decorations: decorations
        };
        if (this.model.root) {
            state = __assign({}, state, { root: this.deflateForStorage(this.model.root) });
        }
        return state;
    };
    TreeWidget.prototype.restoreState = function (oldState) {
        // tslint:disable-next-line:no-any
        var _a = oldState, root = _a.root, decorations = _a.decorations;
        if (root) {
            this.model.root = this.inflateFromStorage(root);
        }
        if (decorations) {
            this.decorations = this.decoratorService.inflateDecorators(decorations);
        }
    };
    var TreeWidget_1;
    __decorate([
        inversify_1.inject(tree_decorator_1.TreeDecoratorService),
        __metadata("design:type", Object)
    ], TreeWidget.prototype, "decoratorService", void 0);
    __decorate([
        inversify_1.inject(tree_search_1.TreeSearch),
        __metadata("design:type", tree_search_1.TreeSearch)
    ], TreeWidget.prototype, "treeSearch", void 0);
    __decorate([
        inversify_1.inject(search_box_1.SearchBoxFactory),
        __metadata("design:type", Function)
    ], TreeWidget.prototype, "searchBoxFactory", void 0);
    __decorate([
        inversify_1.postConstruct(),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", void 0)
    ], TreeWidget.prototype, "init", null);
    TreeWidget = TreeWidget_1 = __decorate([
        inversify_1.injectable(),
        __param(0, inversify_1.inject(exports.TreeProps)),
        __param(1, inversify_1.inject(tree_model_1.TreeModel)),
        __param(2, inversify_1.inject(context_menu_renderer_1.ContextMenuRenderer)),
        __metadata("design:paramtypes", [Object, Object, Object])
    ], TreeWidget);
    return TreeWidget;
}(react_widget_1.ReactWidget));
exports.TreeWidget = TreeWidget;
(function (TreeWidget) {
    var View = /** @class */ (function (_super) {
        __extends(View, _super);
        function View() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.cache = new react_virtualized_1.CellMeasurerCache({
                fixedWidth: true
            });
            _this.renderTreeRow = function (_a) {
                var key = _a.key, index = _a.index, style = _a.style, parent = _a.parent;
                var row = _this.props.rows[index];
                return React.createElement(react_virtualized_1.CellMeasurer, { cache: _this.cache, columnIndex: 0, key: key, parent: parent, rowIndex: index },
                    React.createElement("div", { key: key, style: style }, _this.props.renderNodeRow(row)));
            };
            return _this;
        }
        View.prototype.render = function () {
            var _this = this;
            var _a = this.props, rows = _a.rows, width = _a.width, height = _a.height, scrollToRow = _a.scrollToRow, handleScroll = _a.handleScroll;
            return React.createElement(react_virtualized_1.List, { ref: function (list) { return _this.list = (list || undefined); }, width: width, height: height, rowCount: rows.length, rowHeight: this.cache.rowHeight, rowRenderer: this.renderTreeRow, scrollToIndex: scrollToRow, onScroll: handleScroll, tabIndex: -1, style: {
                    overflowY: 'visible',
                    overflowX: 'visible'
                } });
        };
        return View;
    }(React.Component));
    TreeWidget.View = View;
})(TreeWidget = exports.TreeWidget || (exports.TreeWidget = {}));
exports.TreeWidget = TreeWidget;
