// @theia/core/lib/browser/tree/tree-selection-impl.js
"use strict";
/********************************************************************************
 * Copyright (C) 2018 TypeFox and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var inversify_1 = require("inversify");
var tree_1 = require("./tree");
var common_1 = require("../../common");
var tree_selection_state_1 = require("./tree-selection-state");
var tree_selection_1 = require("./tree-selection");
var TreeSelectionServiceImpl = /** @class */ (function () {
    function TreeSelectionServiceImpl() {
        this.onSelectionChangedEmitter = new common_1.Emitter();
    }
    TreeSelectionServiceImpl.prototype.init = function () {
        this.state = new tree_selection_state_1.TreeSelectionState(this.tree);
    };
    TreeSelectionServiceImpl.prototype.dispose = function () {
        this.onSelectionChangedEmitter.dispose();
    };
    Object.defineProperty(TreeSelectionServiceImpl.prototype, "selectedNodes", {
        get: function () {
            return this.state.selection();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TreeSelectionServiceImpl.prototype, "onSelectionChanged", {
        get: function () {
            return this.onSelectionChangedEmitter.event;
        },
        enumerable: true,
        configurable: true
    });
    TreeSelectionServiceImpl.prototype.fireSelectionChanged = function () {
        this.onSelectionChangedEmitter.fire(this.state.selection());
    };
    TreeSelectionServiceImpl.prototype.addSelection = function (selectionOrTreeNode) {
        var selection = (function (arg) {
            var type = tree_selection_1.TreeSelection.SelectionType.DEFAULT;
            if (tree_selection_1.TreeSelection.is(arg)) {
                return __assign({ type: type }, arg);
            }
            var node = arg;
            return {
                type: type,
                node: node
            };
        })(selectionOrTreeNode);
        if (this.validateNode(selection.node) === undefined) {
            return;
        }
        var oldState = this.state;
        var newState = this.state.nextState(selection);
        var oldNodes = oldState.selection();
        var newNodes = newState.selection();
        var toUnselect = this.difference(oldNodes, newNodes);
        var toSelect = this.difference(newNodes, oldNodes);
        this.unselect(toUnselect);
        this.select(toSelect);
        this.removeFocus(oldNodes, newNodes);
        this.addFocus(newState.focus);
        this.state = newState;
        this.fireSelectionChanged();
    };
    TreeSelectionServiceImpl.prototype.unselect = function (nodes) {
        nodes.forEach(function (node) { return node.selected = false; });
    };
    TreeSelectionServiceImpl.prototype.select = function (nodes) {
        nodes.forEach(function (node) { return node.selected = true; });
    };
    TreeSelectionServiceImpl.prototype.removeFocus = function () {
        var nodes = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            nodes[_i] = arguments[_i];
        }
        nodes.forEach(function (node) { return node.forEach(function (n) { return n.focus = false; }); });
    };
    TreeSelectionServiceImpl.prototype.addFocus = function (node) {
        if (node) {
            node.focus = true;
        }
    };
    /**
     * Returns an array of the difference of two arrays. The returned array contains all elements that are contained by
     * `left` and not contained by `right`. `right` may also contain elements not present in `left`: these are simply ignored.
     */
    TreeSelectionServiceImpl.prototype.difference = function (left, right) {
        return left.filter(function (item) { return right.indexOf(item) === -1; });
    };
    /**
     * Returns a reference to the argument if the node exists in the tree. Otherwise, `undefined`.
     */
    TreeSelectionServiceImpl.prototype.validateNode = function (node) {
        return this.tree.validateNode(node);
    };
    __decorate([
        inversify_1.inject(tree_1.Tree),
        __metadata("design:type", Object)
    ], TreeSelectionServiceImpl.prototype, "tree", void 0);
    __decorate([
        inversify_1.postConstruct(),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", void 0)
    ], TreeSelectionServiceImpl.prototype, "init", null);
    TreeSelectionServiceImpl = __decorate([
        inversify_1.injectable()
    ], TreeSelectionServiceImpl);
    return TreeSelectionServiceImpl;
}());
exports.TreeSelectionServiceImpl = TreeSelectionServiceImpl;
