// @theia/core/lib/browser/tree/tree-decorator.js
"use strict";
/********************************************************************************
 * Copyright (C) 2018 TypeFox and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __values = (this && this.__values) || function (o) {
    var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
    if (m) return m.call(o);
    return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
var inversify_1 = require("inversify");
var common_1 = require("../../common");
/**
 * Decorator service which emits events from all known tree decorators.
 * Keys are the unique tree node IDs and the values
 * are the decoration data collected from all the decorators known by this service.
 */
exports.TreeDecoratorService = Symbol('TreeDecoratorService');
/**
 * The default tree decorator service. Does nothing at all. One has to rebind to a concrete implementation
 * if decorators have to be supported in the tree widget.
 */
var NoopTreeDecoratorService = /** @class */ (function () {
    function NoopTreeDecoratorService() {
        this.emitter = new common_1.Emitter();
        this.onDidChangeDecorations = this.emitter.event;
    }
    NoopTreeDecoratorService.prototype.dispose = function () {
        this.emitter.dispose();
    };
    NoopTreeDecoratorService.prototype.getDecorations = function () {
        return new Map();
    };
    NoopTreeDecoratorService.prototype.deflateDecorators = function () {
        return {};
    };
    NoopTreeDecoratorService.prototype.inflateDecorators = function () {
        return new Map();
    };
    NoopTreeDecoratorService = __decorate([
        inversify_1.injectable()
    ], NoopTreeDecoratorService);
    return NoopTreeDecoratorService;
}());
exports.NoopTreeDecoratorService = NoopTreeDecoratorService;
/**
 * Abstract decorator service implementation which emits events from all known tree decorators and caches the current state.
 */
var AbstractTreeDecoratorService = /** @class */ (function () {
    function AbstractTreeDecoratorService(decorators) {
        var _this = this;
        this.decorators = decorators;
        this.onDidChangeDecorationsEmitter = new common_1.Emitter();
        this.onDidChangeDecorations = this.onDidChangeDecorationsEmitter.event;
        this.toDispose = new common_1.DisposableCollection();
        this.toDispose.push(this.onDidChangeDecorationsEmitter);
        this.toDispose.pushAll(this.decorators.map(function (decorator) {
            return decorator.onDidChangeDecorations(function (data) {
                return _this.onDidChangeDecorationsEmitter.fire(undefined);
            });
        }));
    }
    AbstractTreeDecoratorService.prototype.dispose = function () {
        this.toDispose.dispose();
    };
    AbstractTreeDecoratorService.prototype.getDecorations = function (tree) {
        return __awaiter(this, void 0, void 0, function () {
            var e_1, _a, e_2, _b, changes, _c, _d, decorator, _e, _f, _g, id, data, e_2_1, e_1_1;
            return __generator(this, function (_h) {
                switch (_h.label) {
                    case 0:
                        changes = new Map();
                        _h.label = 1;
                    case 1:
                        _h.trys.push([1, 12, 13, 14]);
                        _c = __values(this.decorators), _d = _c.next();
                        _h.label = 2;
                    case 2:
                        if (!!_d.done) return [3 /*break*/, 11];
                        decorator = _d.value;
                        _h.label = 3;
                    case 3:
                        _h.trys.push([3, 8, 9, 10]);
                        return [4 /*yield*/, decorator.decorations(tree)];
                    case 4:
                        _e = __values.apply(void 0, [(_h.sent()).entries()]), _f = _e.next();
                        _h.label = 5;
                    case 5:
                        if (!!_f.done) return [3 /*break*/, 7];
                        _g = __read(_f.value, 2), id = _g[0], data = _g[1];
                        if (changes.has(id)) {
                            changes.get(id).push(data);
                        }
                        else {
                            changes.set(id, [data]);
                        }
                        _h.label = 6;
                    case 6:
                        _f = _e.next();
                        return [3 /*break*/, 5];
                    case 7: return [3 /*break*/, 10];
                    case 8:
                        e_2_1 = _h.sent();
                        e_2 = { error: e_2_1 };
                        return [3 /*break*/, 10];
                    case 9:
                        try {
                            if (_f && !_f.done && (_b = _e.return)) _b.call(_e);
                        }
                        finally { if (e_2) throw e_2.error; }
                        return [7 /*endfinally*/];
                    case 10:
                        _d = _c.next();
                        return [3 /*break*/, 2];
                    case 11: return [3 /*break*/, 14];
                    case 12:
                        e_1_1 = _h.sent();
                        e_1 = { error: e_1_1 };
                        return [3 /*break*/, 14];
                    case 13:
                        try {
                            if (_d && !_d.done && (_a = _c.return)) _a.call(_c);
                        }
                        finally { if (e_1) throw e_1.error; }
                        return [7 /*endfinally*/];
                    case 14: return [2 /*return*/, changes];
                }
            });
        });
    };
    AbstractTreeDecoratorService.prototype.deflateDecorators = function (decorations) {
        var e_3, _a;
        // tslint:disable-next-line:no-null-keyword
        var state = Object.create(null);
        try {
            for (var decorations_1 = __values(decorations), decorations_1_1 = decorations_1.next(); !decorations_1_1.done; decorations_1_1 = decorations_1.next()) {
                var _b = __read(decorations_1_1.value, 2), id = _b[0], data = _b[1];
                state[id] = data;
            }
        }
        catch (e_3_1) { e_3 = { error: e_3_1 }; }
        finally {
            try {
                if (decorations_1_1 && !decorations_1_1.done && (_a = decorations_1.return)) _a.call(decorations_1);
            }
            finally { if (e_3) throw e_3.error; }
        }
        return state;
    };
    // tslint:disable-next-line:no-any
    AbstractTreeDecoratorService.prototype.inflateDecorators = function (state) {
        var e_4, _a;
        var decorators = new Map();
        try {
            for (var _b = __values(Object.keys(state)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var id = _c.value;
                decorators.set(id, state[id]);
            }
        }
        catch (e_4_1) { e_4 = { error: e_4_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_4) throw e_4.error; }
        }
        return decorators;
    };
    AbstractTreeDecoratorService = __decorate([
        inversify_1.injectable(),
        __metadata("design:paramtypes", [Object])
    ], AbstractTreeDecoratorService);
    return AbstractTreeDecoratorService;
}());
exports.AbstractTreeDecoratorService = AbstractTreeDecoratorService;
/**
 * Namespace for the decoration data and the styling refinements for the decorated tree nodes.
 */
var TreeDecoration;
(function (TreeDecoration) {
    /**
     * CSS styles for the tree decorators.
     */
    var Styles;
    (function (Styles) {
        Styles.CAPTION_HIGHLIGHT_CLASS = 'theia-caption-highlight';
        Styles.CAPTION_PREFIX_CLASS = 'theia-caption-prefix';
        Styles.CAPTION_SUFFIX_CLASS = 'theia-caption-suffix';
        Styles.ICON_WRAPPER_CLASS = 'theia-icon-wrapper';
        Styles.DECORATOR_SIZE_CLASS = 'theia-decorator-size';
        Styles.TOP_RIGHT_CLASS = 'theia-top-right';
        Styles.BOTTOM_RIGHT_CLASS = 'theia-bottom-right';
        Styles.BOTTOM_LEFT_CLASS = 'theia-bottom-left';
        Styles.TOP_LEFT_CLASS = 'theia-top-left';
    })(Styles = TreeDecoration.Styles || (TreeDecoration.Styles = {}));
    /**
     * Enumeration for the quadrant to overlay the image on.
     */
    var IconOverlayPosition;
    (function (IconOverlayPosition) {
        /**
         * Overlays the top right quarter of the original image.
         */
        IconOverlayPosition[IconOverlayPosition["TOP_RIGHT"] = 0] = "TOP_RIGHT";
        /**
         * Overlays the bottom right of the original image.
         */
        IconOverlayPosition[IconOverlayPosition["BOTTOM_RIGHT"] = 1] = "BOTTOM_RIGHT";
        /**
         * Overlays the bottom left segment of the original image.
         */
        IconOverlayPosition[IconOverlayPosition["BOTTOM_LEFT"] = 2] = "BOTTOM_LEFT";
        /**
         * Occupies the top left quarter of the original icon.
         */
        IconOverlayPosition[IconOverlayPosition["TOP_LEFT"] = 3] = "TOP_LEFT";
    })(IconOverlayPosition = TreeDecoration.IconOverlayPosition || (TreeDecoration.IconOverlayPosition = {}));
    (function (IconOverlayPosition) {
        /**
         * Returns with the CSS class style for the enum.
         */
        function getStyle(position) {
            switch (position) {
                case IconOverlayPosition.TOP_RIGHT: return TreeDecoration.Styles.TOP_RIGHT_CLASS;
                case IconOverlayPosition.BOTTOM_RIGHT: return TreeDecoration.Styles.BOTTOM_RIGHT_CLASS;
                case IconOverlayPosition.BOTTOM_LEFT: return TreeDecoration.Styles.BOTTOM_LEFT_CLASS;
                case IconOverlayPosition.TOP_LEFT: return TreeDecoration.Styles.TOP_LEFT_CLASS;
            }
        }
        IconOverlayPosition.getStyle = getStyle;
    })(IconOverlayPosition = TreeDecoration.IconOverlayPosition || (TreeDecoration.IconOverlayPosition = {}));
    var CaptionHighlight;
    (function (CaptionHighlight) {
        var Range;
        (function (Range) {
            /**
             * `true` if the `arg` is contained in the range. The ranges are closed ranges, hence the check is inclusive.
             */
            function contains(arg, range) {
                return arg >= range.offset && arg <= (range.offset + range.length);
            }
            Range.contains = contains;
        })(Range = CaptionHighlight.Range || (CaptionHighlight.Range = {}));
        /**
         * Splits the `caption` argument based on the ranges from the `highlight` argument.
         */
        function split(caption, highlight) {
            var result = [];
            var ranges = highlight.ranges.slice();
            var containerOf = function (index) { return ranges.findIndex(function (range) { return Range.contains(index, range); }); };
            var data = '';
            for (var i = 0; i < caption.length; i++) {
                var containerIndex = containerOf(i);
                if (containerIndex === -1) {
                    data += caption[i];
                }
                else {
                    if (data.length > 0) {
                        result.push({ data: data });
                    }
                    var length_1 = ranges.splice(containerIndex, 1).shift().length;
                    result.push({ data: caption.substr(i, length_1), highligh: true });
                    data = '';
                    i = i + length_1 - 1;
                }
            }
            if (data.length > 0) {
                result.push({ data: data });
            }
            if (ranges.length !== 0) {
                throw new Error('Error occurred when splitting the caption. There was a mismatch between the caption and the corresponding highlighting ranges.');
            }
            return result;
        }
        CaptionHighlight.split = split;
    })(CaptionHighlight = TreeDecoration.CaptionHighlight || (TreeDecoration.CaptionHighlight = {}));
    var Data;
    (function (Data) {
        /**
         * Compares the decoration data based on the priority. Lowest priorities come first.
         */
        Data.comparePriority = function (left, right) { return (left.priority || 0) - (right.priority || 0); };
    })(Data = TreeDecoration.Data || (TreeDecoration.Data = {}));
})(TreeDecoration = exports.TreeDecoration || (exports.TreeDecoration = {}));
