// @theia/debug/lib/browser/debug-session-manager.js
"use strict";
/********************************************************************************
 * Copyright (C) 2018 Red Hat, Inc. and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __values = (this && this.__values) || function (o) {
    var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
    if (m) return m.call(o);
    return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
};
Object.defineProperty(exports, "__esModule", { value: true });
// tslint:disable:no-any
var inversify_1 = require("inversify");
var core_1 = require("@theia/core");
var browser_1 = require("@theia/core/lib/browser");
var browser_2 = require("@theia/editor/lib/browser");
var debug_service_1 = require("../common/debug-service");
var debug_session_1 = require("./debug-session");
var debug_session_contribution_1 = require("./debug-session-contribution");
var debug_breakpoint_1 = require("./model/debug-breakpoint");
var breakpoint_manager_1 = require("./breakpoint/breakpoint-manager");
var uri_1 = require("@theia/core/lib/common/uri");
var browser_3 = require("@theia/variable-resolver/lib/browser");
var debug_session_options_1 = require("./debug-session-options");
var DebugSessionManager = /** @class */ (function () {
    function DebugSessionManager() {
        this._sessions = new Map();
        this.onDidCreateDebugSessionEmitter = new core_1.Emitter();
        this.onDidCreateDebugSession = this.onDidCreateDebugSessionEmitter.event;
        this.onDidStartDebugSessionEmitter = new core_1.Emitter();
        this.onDidStartDebugSession = this.onDidStartDebugSessionEmitter.event;
        this.onDidStopDebugSessionEmitter = new core_1.Emitter();
        this.onDidStopDebugSession = this.onDidStopDebugSessionEmitter.event;
        this.onDidChangeActiveDebugSessionEmitter = new core_1.Emitter();
        this.onDidChangeActiveDebugSession = this.onDidChangeActiveDebugSessionEmitter.event;
        this.onDidDestroyDebugSessionEmitter = new core_1.Emitter();
        this.onDidDestroyDebugSession = this.onDidDestroyDebugSessionEmitter.event;
        this.onDidReceiveDebugSessionCustomEventEmitter = new core_1.Emitter();
        this.onDidReceiveDebugSessionCustomEvent = this.onDidReceiveDebugSessionCustomEventEmitter.event;
        this.onDidChangeBreakpointsEmitter = new core_1.Emitter();
        this.onDidChangeBreakpoints = this.onDidChangeBreakpointsEmitter.event;
        this.onDidChangeEmitter = new core_1.Emitter();
        this.onDidChange = this.onDidChangeEmitter.event;
        this.configurationIds = new Map();
        this.toDisposeOnCurrentSession = new core_1.DisposableCollection();
    }
    DebugSessionManager.prototype.fireDidChangeBreakpoints = function (event) {
        this.onDidChangeBreakpointsEmitter.fire(event);
    };
    DebugSessionManager.prototype.fireDidChange = function (current) {
        this.onDidChangeEmitter.fire(current);
    };
    DebugSessionManager.prototype.init = function () {
        var _this = this;
        this.breakpoints.onDidChangeMarkers(function (uri) { return _this.fireDidChangeBreakpoints({ uri: uri }); });
    };
    DebugSessionManager.prototype.start = function (options) {
        return __awaiter(this, void 0, void 0, function () {
            var resolved, sessionId, e_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 3, , 4]);
                        return [4 /*yield*/, this.resolveConfiguration(options)];
                    case 1:
                        resolved = _a.sent();
                        return [4 /*yield*/, this.debug.createDebugSession(resolved.configuration)];
                    case 2:
                        sessionId = _a.sent();
                        return [2 /*return*/, this.doStart(sessionId, resolved)];
                    case 3:
                        e_1 = _a.sent();
                        if (debug_service_1.DebugError.NotFound.is(e_1)) {
                            this.messageService.error("The debug session type \"" + e_1.data.type + "\" is not supported.");
                            return [2 /*return*/, undefined];
                        }
                        this.messageService.error('There was an error starting the debug session, check the logs for more details.');
                        console.error('Error starting the debug session', e_1);
                        throw e_1;
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    DebugSessionManager.prototype.resolveConfiguration = function (options) {
        return __awaiter(this, void 0, void 0, function () {
            var workspaceFolderUri, resolvedConfiguration, configuration, key, id;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (debug_session_options_1.InternalDebugSessionOptions.is(options)) {
                            return [2 /*return*/, options];
                        }
                        workspaceFolderUri = options.workspaceFolderUri;
                        return [4 /*yield*/, this.debug.resolveDebugConfiguration(options.configuration, workspaceFolderUri)];
                    case 1:
                        resolvedConfiguration = _a.sent();
                        return [4 /*yield*/, this.variableResolver.resolve(resolvedConfiguration)];
                    case 2:
                        configuration = _a.sent();
                        key = configuration.name + workspaceFolderUri;
                        id = this.configurationIds.has(key) ? this.configurationIds.get(key) + 1 : 0;
                        this.configurationIds.set(key, id);
                        return [2 /*return*/, {
                                id: id,
                                configuration: configuration,
                                workspaceFolderUri: workspaceFolderUri
                            }];
                }
            });
        });
    };
    DebugSessionManager.prototype.doStart = function (sessionId, options) {
        return __awaiter(this, void 0, void 0, function () {
            var contrib, sessionFactory, session, state;
            var _this = this;
            return __generator(this, function (_a) {
                contrib = this.sessionContributionRegistry.get(options.configuration.type);
                sessionFactory = contrib ? contrib.debugSessionFactory() : this.debugSessionFactory;
                session = sessionFactory.get(sessionId, options);
                this._sessions.set(sessionId, session);
                this.onDidCreateDebugSessionEmitter.fire(session);
                state = debug_session_1.DebugState.Inactive;
                session.onDidChange(function () {
                    if (state !== session.state) {
                        state = session.state;
                        if (state === debug_session_1.DebugState.Stopped) {
                            _this.onDidStopDebugSessionEmitter.fire(session);
                        }
                    }
                    _this.updateCurrentSession(session);
                });
                session.onDidChangeBreakpoints(function (uri) { return _this.fireDidChangeBreakpoints({ session: session, uri: uri }); });
                session.on('terminated', function (event) {
                    var restart = event.body && event.body.restart;
                    if (restart) {
                        _this.doRestart(session, restart);
                    }
                    else {
                        session.terminate();
                    }
                });
                session.on('exited', function () { return _this.destroy(session.id); });
                session.start().then(function () { return _this.onDidStartDebugSessionEmitter.fire(session); });
                session.onDidCustomEvent(function (_a) {
                    var event = _a.event, body = _a.body;
                    return _this.onDidReceiveDebugSessionCustomEventEmitter.fire({ event: event, body: body, session: session });
                });
                return [2 /*return*/, session];
            });
        });
    };
    DebugSessionManager.prototype.restart = function (session) {
        if (session === void 0) { session = this.currentSession; }
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, session && this.doRestart(session)];
            });
        });
    };
    DebugSessionManager.prototype.doRestart = function (session, restart) {
        return __awaiter(this, void 0, void 0, function () {
            var options, configuration;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, session.restart()];
                    case 1:
                        if (_a.sent()) {
                            return [2 /*return*/, session];
                        }
                        return [4 /*yield*/, session.terminate(!!restart)];
                    case 2:
                        _a.sent();
                        options = session.options, configuration = session.configuration;
                        configuration.__restart = restart;
                        return [2 /*return*/, this.start(options)];
                }
            });
        });
    };
    DebugSessionManager.prototype.remove = function (sessionId) {
        this._sessions.delete(sessionId);
        var currentSession = this.currentSession;
        if (currentSession && currentSession.id === sessionId) {
            this.updateCurrentSession(undefined);
        }
    };
    DebugSessionManager.prototype.getSession = function (sessionId) {
        return this._sessions.get(sessionId);
    };
    Object.defineProperty(DebugSessionManager.prototype, "sessions", {
        get: function () {
            return Array.from(this._sessions.values()).filter(function (session) { return session.state > debug_session_1.DebugState.Inactive; });
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DebugSessionManager.prototype, "currentSession", {
        get: function () {
            return this._currentSession;
        },
        set: function (current) {
            var _this = this;
            if (this._currentSession === current) {
                return;
            }
            this.toDisposeOnCurrentSession.dispose();
            var previous = this.currentSession;
            this._currentSession = current;
            this.onDidChangeActiveDebugSessionEmitter.fire({ previous: previous, current: current });
            if (current) {
                this.toDisposeOnCurrentSession.push(current.onDidChange(function () {
                    if (_this.currentFrame === _this.topFrame) {
                        _this.open();
                    }
                    _this.fireDidChange(current);
                }));
            }
            this.updateBreakpoints(previous, current);
            this.open();
            this.fireDidChange(current);
        },
        enumerable: true,
        configurable: true
    });
    DebugSessionManager.prototype.open = function () {
        var currentFrame = this.currentFrame;
        if (currentFrame) {
            currentFrame.open();
        }
    };
    DebugSessionManager.prototype.updateBreakpoints = function (previous, current) {
        var e_2, _a, e_3, _b;
        var affectedUri = new Set();
        try {
            for (var _c = __values([previous, current]), _d = _c.next(); !_d.done; _d = _c.next()) {
                var session = _d.value;
                if (session) {
                    try {
                        for (var _e = __values(session.breakpointUris), _f = _e.next(); !_f.done; _f = _e.next()) {
                            var uriString = _f.value;
                            if (!affectedUri.has(uriString)) {
                                affectedUri.add(uriString);
                                this.fireDidChangeBreakpoints({
                                    session: current,
                                    uri: new uri_1.default(uriString)
                                });
                            }
                        }
                    }
                    catch (e_3_1) { e_3 = { error: e_3_1 }; }
                    finally {
                        try {
                            if (_f && !_f.done && (_b = _e.return)) _b.call(_e);
                        }
                        finally { if (e_3) throw e_3.error; }
                    }
                }
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (_d && !_d.done && (_a = _c.return)) _a.call(_c);
            }
            finally { if (e_2) throw e_2.error; }
        }
    };
    DebugSessionManager.prototype.updateCurrentSession = function (session) {
        this.currentSession = session || this.sessions[0];
    };
    Object.defineProperty(DebugSessionManager.prototype, "currentThread", {
        get: function () {
            var session = this.currentSession;
            return session && session.currentThread;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DebugSessionManager.prototype, "state", {
        get: function () {
            var session = this.currentSession;
            return session ? session.state : debug_session_1.DebugState.Inactive;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DebugSessionManager.prototype, "currentFrame", {
        get: function () {
            var currentThread = this.currentThread;
            return currentThread && currentThread.currentFrame;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DebugSessionManager.prototype, "topFrame", {
        get: function () {
            var currentThread = this.currentThread;
            return currentThread && currentThread.topFrame;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Destroy the debug session. If session identifier isn't provided then
     * all active debug session will be destroyed.
     * @param sessionId The session identifier
     */
    DebugSessionManager.prototype.destroy = function (sessionId) {
        var _this = this;
        if (sessionId) {
            var session = this._sessions.get(sessionId);
            if (session) {
                this.doDestroy(session);
            }
        }
        else {
            this._sessions.forEach(function (session) { return _this.doDestroy(session); });
        }
    };
    DebugSessionManager.prototype.doDestroy = function (session) {
        this.debug.terminateDebugSession(session.id);
        session.dispose();
        this.remove(session.id);
        this.onDidDestroyDebugSessionEmitter.fire(session);
    };
    DebugSessionManager.prototype.getBreakpoints = function (arg, arg2) {
        var _this = this;
        var uri = arg instanceof uri_1.default ? arg : undefined;
        var session = arg instanceof debug_session_1.DebugSession ? arg : arg2 instanceof debug_session_1.DebugSession ? arg2 : this.currentSession;
        if (session && session.state > debug_session_1.DebugState.Initializing) {
            return session.getBreakpoints(uri);
        }
        return this.breakpoints.findMarkers({ uri: uri }).map(function (_a) {
            var data = _a.data;
            return new debug_breakpoint_1.DebugBreakpoint(data, _this.labelProvider, _this.breakpoints, _this.editorManager);
        });
    };
    DebugSessionManager.prototype.getBreakpoint = function (uri, line) {
        var session = this.currentSession;
        if (session && session.state > debug_session_1.DebugState.Initializing) {
            return session.getBreakpoints(uri).filter(function (breakpoint) { return breakpoint.line === line; })[0];
        }
        var origin = this.breakpoints.getBreakpoint(uri, line);
        return origin && new debug_breakpoint_1.DebugBreakpoint(origin, this.labelProvider, this.breakpoints, this.editorManager);
    };
    DebugSessionManager.prototype.addBreakpoints = function (breakpoints) {
        var _this = this;
        breakpoints.forEach(function (breakpoint) {
            _this.breakpoints.addBreakpoint(breakpoint.uri, breakpoint.line, breakpoint.column);
            _this.fireDidChangeBreakpoints({ uri: breakpoint.uri });
        });
    };
    DebugSessionManager.prototype.deleteBreakpoints = function (breakpoints) {
        var _this = this;
        breakpoints.forEach(function (breakpoint) {
            _this.breakpoints.deleteBreakpoint(breakpoint.uri, breakpoint.line, breakpoint.column);
            _this.fireDidChangeBreakpoints({ uri: breakpoint.uri });
        });
    };
    __decorate([
        inversify_1.inject(debug_session_contribution_1.DebugSessionFactory),
        __metadata("design:type", Object)
    ], DebugSessionManager.prototype, "debugSessionFactory", void 0);
    __decorate([
        inversify_1.inject(debug_service_1.DebugService),
        __metadata("design:type", Object)
    ], DebugSessionManager.prototype, "debug", void 0);
    __decorate([
        inversify_1.inject(browser_1.LabelProvider),
        __metadata("design:type", browser_1.LabelProvider)
    ], DebugSessionManager.prototype, "labelProvider", void 0);
    __decorate([
        inversify_1.inject(browser_2.EditorManager),
        __metadata("design:type", browser_2.EditorManager)
    ], DebugSessionManager.prototype, "editorManager", void 0);
    __decorate([
        inversify_1.inject(breakpoint_manager_1.BreakpointManager),
        __metadata("design:type", breakpoint_manager_1.BreakpointManager)
    ], DebugSessionManager.prototype, "breakpoints", void 0);
    __decorate([
        inversify_1.inject(browser_3.VariableResolverService),
        __metadata("design:type", browser_3.VariableResolverService)
    ], DebugSessionManager.prototype, "variableResolver", void 0);
    __decorate([
        inversify_1.inject(debug_session_contribution_1.DebugSessionContributionRegistry),
        __metadata("design:type", Object)
    ], DebugSessionManager.prototype, "sessionContributionRegistry", void 0);
    __decorate([
        inversify_1.inject(core_1.MessageService),
        __metadata("design:type", core_1.MessageService)
    ], DebugSessionManager.prototype, "messageService", void 0);
    __decorate([
        inversify_1.postConstruct(),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", void 0)
    ], DebugSessionManager.prototype, "init", null);
    DebugSessionManager = __decorate([
        inversify_1.injectable()
    ], DebugSessionManager);
    return DebugSessionManager;
}());
exports.DebugSessionManager = DebugSessionManager;
