// @theia/plugin-ext/lib/main/browser/status-bar-message-registry-main.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var types = require("../../plugin/types-impl");
var status_bar_1 = require("@theia/core/lib/browser/status-bar/status-bar");
var STATUS_BAR_MESSAGE_PRE = 'status-bar-entry';
var StatusBarMessageRegistryMainImpl = /** @class */ (function () {
    function StatusBarMessageRegistryMainImpl(container) {
        this.entries = new Map();
        this.delegate = container.get(status_bar_1.StatusBar);
    }
    StatusBarMessageRegistryMainImpl.prototype.$setMessage = function (text, priority, alignment, color, tooltip, command) {
        var id = this.uniqueId;
        var entry = {
            text: text,
            priority: priority,
            alignment: alignment === types.StatusBarAlignment.Left ? status_bar_1.StatusBarAlignment.LEFT : status_bar_1.StatusBarAlignment.RIGHT,
            color: color,
            tooltip: tooltip,
            command: command
        };
        this.entries.set(id, entry);
        return this.delegate.setElement(id, entry).then(function () { return Promise.resolve(id); });
    };
    StatusBarMessageRegistryMainImpl.prototype.$update = function (id, message) {
        var entry = this.entries.get(id);
        if (entry) {
            entry.text = message;
            this.delegate.setElement(id, entry);
        }
    };
    StatusBarMessageRegistryMainImpl.prototype.$dispose = function (id) {
        var _this = this;
        this.delegate.removeElement(id).then(function () {
            _this.entries.delete(id);
        });
    };
    Object.defineProperty(StatusBarMessageRegistryMainImpl.prototype, "uniqueId", {
        get: function () {
            var extensionId = STATUS_BAR_MESSAGE_PRE;
            for (var counter = 0; counter < 100; counter++) {
                extensionId = STATUS_BAR_MESSAGE_PRE + "_id_" + ('0000' + (Math.random() * Math.pow(36, 4) << 0).toString(36)).slice(-4);
                if (!this.entries.get(extensionId)) {
                    break;
                }
            }
            return extensionId;
        },
        enumerable: true,
        configurable: true
    });
    return StatusBarMessageRegistryMainImpl;
}());
exports.StatusBarMessageRegistryMainImpl = StatusBarMessageRegistryMainImpl;
