// @theia/task/lib/browser/quick-open-task.js
"use strict";
/********************************************************************************
 * Copyright (C) 2017 Ericsson and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __values = (this && this.__values) || function (o) {
    var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
    if (m) return m.call(o);
    return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
};
Object.defineProperty(exports, "__esModule", { value: true });
var inversify_1 = require("inversify");
var quick_open_1 = require("@theia/core/lib/browser/quick-open/");
var task_service_1 = require("./task-service");
var task_configurations_1 = require("./task-configurations");
var QuickOpenTask = /** @class */ (function () {
    function QuickOpenTask() {
        this.prefix = 'task ';
        this.description = 'Run Task';
    }
    /** Initialize this quick open model with the tasks. */
    QuickOpenTask.prototype.init = function () {
        return __awaiter(this, void 0, void 0, function () {
            var e_1, _a, e_2, _b, configuredTasks, configuredTasks_1, configuredTasks_1_1, task, providedTasks, providedTasks_1, providedTasks_1_1, task;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        this.items = [];
                        return [4 /*yield*/, this.taskConfigurations.getTasks()];
                    case 1:
                        configuredTasks = _c.sent();
                        if (!configuredTasks.length) {
                            this.items.push(new quick_open_1.QuickOpenItem({
                                label: 'No tasks found',
                                run: function (mode) { return false; }
                            }));
                        }
                        try {
                            for (configuredTasks_1 = __values(configuredTasks), configuredTasks_1_1 = configuredTasks_1.next(); !configuredTasks_1_1.done; configuredTasks_1_1 = configuredTasks_1.next()) {
                                task = configuredTasks_1_1.value;
                                this.items.push(new TaskRunQuickOpenItem(task, this.taskService, false));
                            }
                        }
                        catch (e_1_1) { e_1 = { error: e_1_1 }; }
                        finally {
                            try {
                                if (configuredTasks_1_1 && !configuredTasks_1_1.done && (_a = configuredTasks_1.return)) _a.call(configuredTasks_1);
                            }
                            finally { if (e_1) throw e_1.error; }
                        }
                        return [4 /*yield*/, this.taskService.getProvidedTasks()];
                    case 2:
                        providedTasks = _c.sent();
                        try {
                            for (providedTasks_1 = __values(providedTasks), providedTasks_1_1 = providedTasks_1.next(); !providedTasks_1_1.done; providedTasks_1_1 = providedTasks_1.next()) {
                                task = providedTasks_1_1.value;
                                this.items.push(new TaskRunQuickOpenItem(task, this.taskService, true));
                            }
                        }
                        catch (e_2_1) { e_2 = { error: e_2_1 }; }
                        finally {
                            try {
                                if (providedTasks_1_1 && !providedTasks_1_1.done && (_b = providedTasks_1.return)) _b.call(providedTasks_1);
                            }
                            finally { if (e_2) throw e_2.error; }
                        }
                        return [2 /*return*/];
                }
            });
        });
    };
    QuickOpenTask.prototype.open = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.init()];
                    case 1:
                        _a.sent();
                        this.quickOpenService.open(this, {
                            placeholder: 'Type the name of a task you want to execute',
                            fuzzyMatchLabel: true,
                            fuzzySort: true
                        });
                        return [2 /*return*/];
                }
            });
        });
    };
    QuickOpenTask.prototype.getModel = function () {
        return this;
    };
    QuickOpenTask.prototype.getOptions = function () {
        return {
            fuzzyMatchLabel: true,
            fuzzySort: true
        };
    };
    QuickOpenTask.prototype.attach = function () {
        var _this = this;
        this.items = [];
        this.taskService.getRunningTasks().then(function (tasks) {
            var e_3, _a;
            if (!tasks.length) {
                _this.items.push(new quick_open_1.QuickOpenItem({
                    label: 'No tasks found',
                    run: function (_mode) { return false; }
                }));
            }
            try {
                for (var tasks_1 = __values(tasks), tasks_1_1 = tasks_1.next(); !tasks_1_1.done; tasks_1_1 = tasks_1.next()) {
                    var task = tasks_1_1.value;
                    // can only attach to terminal processes, so only list those
                    if (task.terminalId) {
                        _this.items.push(new TaskAttachQuickOpenItem(task, _this.getRunningTaskLabel(task), _this.taskService));
                    }
                }
            }
            catch (e_3_1) { e_3 = { error: e_3_1 }; }
            finally {
                try {
                    if (tasks_1_1 && !tasks_1_1.done && (_a = tasks_1.return)) _a.call(tasks_1);
                }
                finally { if (e_3) throw e_3.error; }
            }
            _this.quickOpenService.open(_this, {
                placeholder: 'Choose task to open',
                fuzzyMatchLabel: true,
                fuzzySort: true
            });
        });
    };
    QuickOpenTask.prototype.onType = function (lookFor, acceptor) {
        acceptor(this.items);
    };
    QuickOpenTask.prototype.getRunningTaskLabel = function (task) {
        return "Task id: " + task.taskId + ", label: " + task.config.label;
    };
    __decorate([
        inversify_1.inject(task_service_1.TaskService),
        __metadata("design:type", task_service_1.TaskService)
    ], QuickOpenTask.prototype, "taskService", void 0);
    __decorate([
        inversify_1.inject(task_configurations_1.TaskConfigurations),
        __metadata("design:type", task_configurations_1.TaskConfigurations)
    ], QuickOpenTask.prototype, "taskConfigurations", void 0);
    __decorate([
        inversify_1.inject(quick_open_1.QuickOpenService),
        __metadata("design:type", quick_open_1.QuickOpenService)
    ], QuickOpenTask.prototype, "quickOpenService", void 0);
    QuickOpenTask = __decorate([
        inversify_1.injectable()
    ], QuickOpenTask);
    return QuickOpenTask;
}());
exports.QuickOpenTask = QuickOpenTask;
var TaskRunQuickOpenItem = /** @class */ (function (_super) {
    __extends(TaskRunQuickOpenItem, _super);
    function TaskRunQuickOpenItem(task, taskService, provided) {
        var _this = _super.call(this) || this;
        _this.task = task;
        _this.taskService = taskService;
        _this.provided = provided;
        return _this;
    }
    TaskRunQuickOpenItem.prototype.getLabel = function () {
        return this.task.type + ": " + this.task.label;
    };
    TaskRunQuickOpenItem.prototype.getDescription = function () {
        return this.provided ? 'provided' : '';
    };
    TaskRunQuickOpenItem.prototype.run = function (mode) {
        if (mode !== quick_open_1.QuickOpenMode.OPEN) {
            return false;
        }
        this.taskService.run(this.task.type, this.task.label);
        return true;
    };
    return TaskRunQuickOpenItem;
}(quick_open_1.QuickOpenItem));
exports.TaskRunQuickOpenItem = TaskRunQuickOpenItem;
var TaskAttachQuickOpenItem = /** @class */ (function (_super) {
    __extends(TaskAttachQuickOpenItem, _super);
    function TaskAttachQuickOpenItem(task, taskLabel, taskService) {
        var _this = _super.call(this) || this;
        _this.task = task;
        _this.taskLabel = taskLabel;
        _this.taskService = taskService;
        return _this;
    }
    TaskAttachQuickOpenItem.prototype.getLabel = function () {
        return this.taskLabel;
    };
    TaskAttachQuickOpenItem.prototype.run = function (mode) {
        if (mode !== quick_open_1.QuickOpenMode.OPEN) {
            return false;
        }
        if (this.task.terminalId) {
            this.taskService.attach(this.task.terminalId, this.task.taskId);
        }
        return true;
    };
    return TaskAttachQuickOpenItem;
}(quick_open_1.QuickOpenItem));
exports.TaskAttachQuickOpenItem = TaskAttachQuickOpenItem;
