// @theia/debug/lib/browser/view/debug-view-model.js
"use strict";
/********************************************************************************
 * Copyright (C) 2018 TypeFox and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var inversify_1 = require("inversify");
var common_1 = require("@theia/core/lib/common");
var debug_session_1 = require("../debug-session");
var debug_session_manager_1 = require("../debug-session-manager");
exports.DebugViewOptions = Symbol('DebugViewOptions');
var DebugViewModel = /** @class */ (function () {
    function DebugViewModel() {
        this.onDidChangeEmitter = new common_1.Emitter();
        this.onDidChange = this.onDidChangeEmitter.event;
        this.onDidChangeBreakpointsEmitter = new common_1.Emitter();
        this.onDidChangeBreakpoints = this.onDidChangeBreakpointsEmitter.event;
        this.toDispose = new common_1.DisposableCollection(this.onDidChangeEmitter, this.onDidChangeBreakpointsEmitter);
        this._sessions = new Set();
    }
    DebugViewModel.prototype.fireDidChange = function () {
        this.onDidChangeEmitter.fire(undefined);
    };
    DebugViewModel.prototype.fireDidChangeBreakpoints = function (uri) {
        this.onDidChangeBreakpointsEmitter.fire(uri);
    };
    Object.defineProperty(DebugViewModel.prototype, "sessions", {
        get: function () {
            return this._sessions.values();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DebugViewModel.prototype, "sessionCount", {
        get: function () {
            return this._sessions.size;
        },
        enumerable: true,
        configurable: true
    });
    DebugViewModel.prototype.push = function (session) {
        if (this._sessions.has(session)) {
            return;
        }
        this._sessions.add(session);
        this.fireDidChange();
    };
    DebugViewModel.prototype.delete = function (session) {
        if (this._sessions.delete(session)) {
            this.fireDidChange();
            return true;
        }
        return false;
    };
    Object.defineProperty(DebugViewModel.prototype, "session", {
        get: function () {
            return this.sessions.next().value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DebugViewModel.prototype, "id", {
        get: function () {
            return this.session && this.session.id || '-1';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DebugViewModel.prototype, "label", {
        get: function () {
            return this.session && this.session.label || 'Unknown Session';
        },
        enumerable: true,
        configurable: true
    });
    DebugViewModel.prototype.has = function (session) {
        return !!session && this._sessions.has(session);
    };
    DebugViewModel.prototype.init = function () {
        var _this = this;
        if (this.options.session) {
            this.push(this.options.session);
        }
        this.toDispose.push(this.manager.onDidChangeActiveDebugSession(function (_a) {
            var previous = _a.previous, current = _a.current;
            if (_this.has(previous) && !_this.has(current)) {
                _this.fireDidChange();
            }
        }));
        this.toDispose.push(this.manager.onDidChange(function (current) {
            if (_this.has(current)) {
                _this.fireDidChange();
            }
        }));
        this.toDispose.push(this.manager.onDidChangeBreakpoints(function (_a) {
            var session = _a.session, uri = _a.uri;
            if (!session || session === _this.currentSession) {
                _this.fireDidChangeBreakpoints(uri);
            }
        }));
    };
    DebugViewModel.prototype.dispose = function () {
        this.toDispose.dispose();
    };
    Object.defineProperty(DebugViewModel.prototype, "currentSession", {
        get: function () {
            var currentSession = this.manager.currentSession;
            return this.has(currentSession) && currentSession || this.session;
        },
        set: function (currentSession) {
            this.manager.currentSession = currentSession;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DebugViewModel.prototype, "state", {
        get: function () {
            var currentSession = this.currentSession;
            return currentSession && currentSession.state || debug_session_1.DebugState.Inactive;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DebugViewModel.prototype, "currentThread", {
        get: function () {
            var currentSession = this.currentSession;
            return currentSession && currentSession.currentThread;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DebugViewModel.prototype, "currentFrame", {
        get: function () {
            var currentThread = this.currentThread;
            return currentThread && currentThread.currentFrame;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DebugViewModel.prototype, "breakpoints", {
        get: function () {
            return this.manager.getBreakpoints(this.currentSession);
        },
        enumerable: true,
        configurable: true
    });
    DebugViewModel.prototype.start = function () {
        return __awaiter(this, void 0, void 0, function () {
            var session, newSession;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        session = this.session;
                        if (!session) {
                            return [2 /*return*/];
                        }
                        return [4 /*yield*/, this.manager.start(session.options)];
                    case 1:
                        newSession = _a.sent();
                        if (newSession) {
                            this._sessions.delete(session);
                            this._sessions.add(newSession);
                            this.fireDidChange();
                        }
                        return [2 /*return*/];
                }
            });
        });
    };
    DebugViewModel.prototype.restart = function () {
        return __awaiter(this, void 0, void 0, function () {
            var session, newSession;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        session = this.session;
                        if (!session) {
                            return [2 /*return*/];
                        }
                        return [4 /*yield*/, this.manager.restart(session)];
                    case 1:
                        newSession = _a.sent();
                        if (newSession !== session) {
                            this._sessions.delete(session);
                            this._sessions.add(newSession);
                        }
                        this.fireDidChange();
                        return [2 /*return*/];
                }
            });
        });
    };
    __decorate([
        inversify_1.inject(exports.DebugViewOptions),
        __metadata("design:type", Object)
    ], DebugViewModel.prototype, "options", void 0);
    __decorate([
        inversify_1.inject(debug_session_manager_1.DebugSessionManager),
        __metadata("design:type", debug_session_manager_1.DebugSessionManager)
    ], DebugViewModel.prototype, "manager", void 0);
    __decorate([
        inversify_1.postConstruct(),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", void 0)
    ], DebugViewModel.prototype, "init", null);
    DebugViewModel = __decorate([
        inversify_1.injectable()
    ], DebugViewModel);
    return DebugViewModel;
}());
exports.DebugViewModel = DebugViewModel;
