// go-language-server/out/src-vscode-mock/vscode.js
"use strict";
/*
 * Copyright (C) 2017 TypeFox and others.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 */
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Mocks VSCode API or replaces it by LSP
 */
const vscode_uri_1 = require("vscode-uri");
exports.Uri = vscode_uri_1.default;
__export(require("./commands"));
__export(require("./debug"));
__export(require("./diagnostics"));
__export(require("./languages"));
__export(require("./text-document"));
__export(require("./text-editor"));
__export(require("./types"));
__export(require("./window"));
__export(require("./workspace"));
var protocol_1 = require("vscode-languageclient/lib/protocol");
exports.MessageType = protocol_1.MessageType;
var vscode_languageserver_1 = require("vscode-languageserver");
exports.Disposable = vscode_languageserver_1.Disposable;
exports.Event = vscode_languageserver_1.Event;
exports.EventEmitter = vscode_languageserver_1.Emitter;
exports.CancellationToken = vscode_languageserver_1.CancellationToken;
exports.CancellationTokenSource = vscode_languageserver_1.CancellationTokenSource;
