// @theia/plugin-ext/lib/main/browser/debug/debug-main.js
"use strict";
/********************************************************************************
 * Copyright (C) 2018 Red Hat, Inc. and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __values = (this && this.__values) || function (o) {
    var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
    if (m) return m.call(o);
    return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
};
Object.defineProperty(exports, "__esModule", { value: true });
var plugin_api_1 = require("../../../api/plugin-api");
var debug_session_manager_1 = require("@theia/debug/lib/browser/debug-session-manager");
var browser_1 = require("@theia/core/lib/browser");
var browser_2 = require("@theia/editor/lib/browser");
var breakpoint_manager_1 = require("@theia/debug/lib/browser/breakpoint/breakpoint-manager");
var debug_breakpoint_1 = require("@theia/debug/lib/browser/model/debug-breakpoint");
var vscode_uri_1 = require("vscode-uri");
var debug_console_session_1 = require("@theia/debug/lib/browser/console/debug-console-session");
var debug_configuration_manager_1 = require("@theia/debug/lib/browser/debug-configuration-manager");
var terminal_service_1 = require("@theia/terminal/lib/browser/base/terminal-service");
var message_service_protocol_1 = require("@theia/core/lib/common/message-service-protocol");
var output_channel_1 = require("@theia/output/lib/common/output-channel");
var debug_preferences_1 = require("@theia/debug/lib/browser/debug-preferences");
var plugin_debug_adapter_contribution_1 = require("./plugin-debug-adapter-contribution");
var plugin_debug_session_contribution_registry_1 = require("./plugin-debug-session-contribution-registry");
var disposable_1 = require("@theia/core/lib/common/disposable");
var plugin_debug_session_factory_1 = require("./plugin-debug-session-factory");
var connection_1 = require("../../../common/connection");
var plugin_debug_service_1 = require("./plugin-debug-service");
var debug_schema_updater_1 = require("@theia/debug/lib/browser/debug-schema-updater");
var DebugMainImpl = /** @class */ (function () {
    function DebugMainImpl(rpc, connectionMain, container) {
        var _this = this;
        this.connectionMain = connectionMain;
        this.toDispose = new Map();
        this.debugExt = rpc.getProxy(plugin_api_1.MAIN_RPC_CONTEXT.DEBUG_EXT);
        this.sessionManager = container.get(debug_session_manager_1.DebugSessionManager);
        this.labelProvider = container.get(browser_1.LabelProvider);
        this.editorManager = container.get(browser_2.EditorManager);
        this.breakpointsManager = container.get(breakpoint_manager_1.BreakpointManager);
        this.debugConsoleSession = container.get(debug_console_session_1.DebugConsoleSession);
        this.configurationManager = container.get(debug_configuration_manager_1.DebugConfigurationManager);
        this.terminalService = container.get(terminal_service_1.TerminalService);
        this.messages = container.get(message_service_protocol_1.MessageClient);
        this.outputChannelManager = container.get(output_channel_1.OutputChannelManager);
        this.debugPreferences = container.get(debug_preferences_1.DebugPreferences);
        this.adapterContributionRegistrator = container.get(plugin_debug_service_1.PluginDebugService);
        this.sessionContributionRegistrator = container.get(plugin_debug_session_contribution_registry_1.PluginDebugSessionContributionRegistry);
        this.debugSchemaUpdater = container.get(debug_schema_updater_1.DebugSchemaUpdater);
        // TODO: distinguish added/deleted breakpoints
        this.breakpointsManager.onDidChangeMarkers(function (uri) {
            var all = _this.breakpointsManager.getBreakpoints();
            var affected = _this.breakpointsManager.getBreakpoints(uri);
            _this.debugExt.$breakpointsDidChange(_this.toTheiaPluginApiBreakpoints(all), [], [], _this.toTheiaPluginApiBreakpoints(affected));
        });
        this.sessionManager.onDidCreateDebugSession(function (debugSession) { return _this.debugExt.$sessionDidCreate(debugSession.id); });
        this.sessionManager.onDidDestroyDebugSession(function (debugSession) { return _this.debugExt.$sessionDidDestroy(debugSession.id); });
        this.sessionManager.onDidChangeActiveDebugSession(function (event) { return _this.debugExt.$sessionDidChange(event.current && event.current.id); });
        this.sessionManager.onDidReceiveDebugSessionCustomEvent(function (event) { return _this.debugExt.$onSessionCustomEvent(event.session.id, event.event, event.body); });
    }
    DebugMainImpl.prototype.$appendToDebugConsole = function (value) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                this.debugConsoleSession.append(value);
                return [2 /*return*/];
            });
        });
    };
    DebugMainImpl.prototype.$appendLineToDebugConsole = function (value) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                this.debugConsoleSession.appendLine(value);
                return [2 /*return*/];
            });
        });
    };
    DebugMainImpl.prototype.$registerDebuggerContribution = function (description) {
        return __awaiter(this, void 0, void 0, function () {
            var disposable, debugSessionFactory;
            var _this = this;
            return __generator(this, function (_a) {
                disposable = new disposable_1.DisposableCollection();
                this.toDispose.set(description.type, disposable);
                debugSessionFactory = new plugin_debug_session_factory_1.PluginDebugSessionFactory(this.terminalService, this.editorManager, this.breakpointsManager, this.labelProvider, this.messages, this.outputChannelManager, this.debugPreferences, function (sessionId) { return __awaiter(_this, void 0, void 0, function () {
                    var connection;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0: return [4 /*yield*/, this.connectionMain.ensureConnection(sessionId)];
                            case 1:
                                connection = _a.sent();
                                return [2 /*return*/, new connection_1.PluginWebSocketChannel(connection)];
                        }
                    });
                }); });
                disposable.pushAll([
                    this.adapterContributionRegistrator.registerDebugAdapterContribution(new plugin_debug_adapter_contribution_1.PluginDebugAdapterContribution(description, this.debugExt)),
                    this.sessionContributionRegistrator.registerDebugSessionContribution({
                        debugType: description.type,
                        debugSessionFactory: function () { return debugSessionFactory; }
                    })
                ]);
                this.debugSchemaUpdater.update();
                return [2 /*return*/];
            });
        });
    };
    DebugMainImpl.prototype.$unregisterDebuggerConfiguration = function (debugType) {
        return __awaiter(this, void 0, void 0, function () {
            var disposable;
            return __generator(this, function (_a) {
                disposable = this.toDispose.get(debugType);
                if (disposable) {
                    disposable.dispose();
                    this.toDispose.delete(debugType);
                    this.debugSchemaUpdater.update();
                }
                return [2 /*return*/];
            });
        });
    };
    DebugMainImpl.prototype.$addBreakpoints = function (breakpoints) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                this.sessionManager.addBreakpoints(this.toInternalBreakpoints(breakpoints));
                return [2 /*return*/];
            });
        });
    };
    DebugMainImpl.prototype.$removeBreakpoints = function (breakpoints) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                this.sessionManager.deleteBreakpoints(this.toInternalBreakpoints(breakpoints));
                return [2 /*return*/];
            });
        });
    };
    DebugMainImpl.prototype.$customRequest = function (sessionId, command, args) {
        return __awaiter(this, void 0, void 0, function () {
            var session;
            return __generator(this, function (_a) {
                session = this.sessionManager.getSession(sessionId);
                if (session) {
                    return [2 /*return*/, session.sendCustomRequest(command, args)];
                }
                throw new Error("Debug session '" + sessionId + "' not found");
            });
        });
    };
    DebugMainImpl.prototype.$startDebugging = function (folder, nameOrConfiguration) {
        return __awaiter(this, void 0, void 0, function () {
            var e_1, _a, configuration, _b, _c, options, session;
            return __generator(this, function (_d) {
                switch (_d.label) {
                    case 0:
                        if (typeof nameOrConfiguration === 'string') {
                            try {
                                for (_b = __values(this.configurationManager.all), _c = _b.next(); !_c.done; _c = _b.next()) {
                                    options = _c.value;
                                    if (options.configuration.name === nameOrConfiguration) {
                                        configuration = options.configuration;
                                    }
                                }
                            }
                            catch (e_1_1) { e_1 = { error: e_1_1 }; }
                            finally {
                                try {
                                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                                }
                                finally { if (e_1) throw e_1.error; }
                            }
                        }
                        else {
                            configuration = nameOrConfiguration;
                        }
                        if (!configuration) {
                            console.error("There is no debug configuration for " + nameOrConfiguration);
                            return [2 /*return*/, false];
                        }
                        return [4 /*yield*/, this.sessionManager.start({
                                configuration: configuration,
                                workspaceFolderUri: folder && vscode_uri_1.default.revive(folder.uri).toString()
                            })];
                    case 1:
                        session = _d.sent();
                        return [2 /*return*/, !!session];
                }
            });
        });
    };
    DebugMainImpl.prototype.toInternalBreakpoints = function (breakpoints) {
        var _this = this;
        return breakpoints
            .filter(function (breakpoint) { return !!breakpoint.location; })
            .map(function (breakpoint) {
            var location = breakpoint.location;
            var uri = vscode_uri_1.default.revive(location.uri);
            var uriString = uri.toString();
            var origin = {
                uri: uriString,
                enabled: true,
                raw: {
                    line: location.range.startLineNumber,
                    column: location.range.startColumn,
                    condition: breakpoint.condition,
                    hitCondition: breakpoint.hitCondition,
                    logMessage: breakpoint.logMessage
                }
            };
            return new debug_breakpoint_1.DebugBreakpoint(origin, _this.labelProvider, _this.breakpointsManager, _this.editorManager, _this.sessionManager.currentSession);
        });
    };
    DebugMainImpl.prototype.toTheiaPluginApiBreakpoints = function (sourceBreakpoints) {
        return sourceBreakpoints.map(function (b) {
            var breakpoint = {
                enabled: b.enabled,
                condition: b.raw.condition,
                hitCondition: b.raw.hitCondition,
                logMessage: b.raw.logMessage,
                location: {
                    uri: vscode_uri_1.default.parse(b.uri),
                    range: {
                        startLineNumber: b.raw.line,
                        startColumn: b.raw.column || 0,
                        endLineNumber: b.raw.line,
                        endColumn: b.raw.column || 0
                    }
                }
            };
            return breakpoint;
        });
    };
    return DebugMainImpl;
}());
exports.DebugMainImpl = DebugMainImpl;
