// @theia/debug/lib/browser/console/debug-console-contribution.js
"use strict";
/********************************************************************************
 * Copyright (C) 2018 TypeFox and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var inversify_1 = require("inversify");
var browser_1 = require("@theia/core/lib/browser");
var console_widget_1 = require("@theia/console/lib/browser/console-widget");
var debug_console_session_1 = require("./debug-console-session");
var DebugConsoleContribution = /** @class */ (function (_super) {
    __extends(DebugConsoleContribution, _super);
    function DebugConsoleContribution() {
        return _super.call(this, {
            widgetId: DebugConsoleContribution_1.options.id,
            widgetName: DebugConsoleContribution_1.options.title.label,
            defaultWidgetOptions: {
                area: 'bottom'
            },
            toggleCommandId: 'debug:console:toggle',
            toggleKeybinding: 'ctrlcmd+shift+y'
        }) || this;
    }
    DebugConsoleContribution_1 = DebugConsoleContribution;
    DebugConsoleContribution.create = function (parent) {
        var child = console_widget_1.ConsoleWidget.createContainer(parent, DebugConsoleContribution_1.options);
        var widget = child.get(console_widget_1.ConsoleWidget);
        widget.session = child.get(debug_console_session_1.DebugConsoleSession);
        return widget;
    };
    DebugConsoleContribution.bindContribution = function (bind) {
        bind(debug_console_session_1.DebugConsoleSession).toSelf().inSingletonScope();
        browser_1.bindViewContribution(bind, DebugConsoleContribution_1).onActivation(function (context, _) {
            // eagerly initialize the debug console session
            context.container.get(debug_console_session_1.DebugConsoleSession);
            return _;
        });
        bind(browser_1.WidgetFactory).toDynamicValue(function (_a) {
            var container = _a.container;
            return ({
                id: DebugConsoleContribution_1.options.id,
                createWidget: function () { return DebugConsoleContribution_1.create(container); }
            });
        });
    };
    var DebugConsoleContribution_1;
    DebugConsoleContribution.options = {
        id: 'debug-console',
        title: {
            label: 'Debug Console'
        },
        input: {
            uri: debug_console_session_1.DebugConsoleSession.uri,
            options: {
                autoSizing: true,
                minHeight: 1,
                maxHeight: 10
            }
        }
    };
    DebugConsoleContribution = DebugConsoleContribution_1 = __decorate([
        inversify_1.injectable(),
        __metadata("design:paramtypes", [])
    ], DebugConsoleContribution);
    return DebugConsoleContribution;
}(browser_1.AbstractViewContribution));
exports.DebugConsoleContribution = DebugConsoleContribution;
