// @theia/preferences/lib/browser/preferences-tree-widget.js
"use strict";
/********************************************************************************
 * Copyright (C) 2018 Red Hat, Inc. and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __values = (this && this.__values) || function (o) {
    var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
    if (m) return m.call(o);
    return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
};
Object.defineProperty(exports, "__esModule", { value: true });
var inversify_1 = require("inversify");
var preferences_menu_factory_1 = require("./preferences-menu-factory");
var preferences_decorator_1 = require("./preferences-decorator");
var algorithm_1 = require("@phosphor/algorithm");
var widgets_1 = require("@phosphor/widgets");
var browser_1 = require("@theia/core/lib/browser");
var user_preference_provider_1 = require("./user-preference-provider");
var workspace_preference_provider_1 = require("./workspace-preference-provider");
var core_1 = require("@theia/core");
var promise_util_1 = require("@theia/core/lib/common/promise-util");
var browser_2 = require("@theia/editor/lib/browser");
var common_1 = require("@theia/filesystem/lib/common");
var browser_3 = require("@theia/userstorage/lib/browser");
var uri_1 = require("@theia/core/lib/common/uri");
var PreferencesContainer = /** @class */ (function (_super) {
    __extends(PreferencesContainer, _super);
    function PreferencesContainer() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.editors = [];
        _this.deferredEditors = new promise_util_1.Deferred();
        _this.onDirtyChangedEmitter = new core_1.Emitter();
        _this.onDirtyChanged = _this.onDirtyChangedEmitter.event;
        _this.toDispose = new core_1.DisposableCollection();
        _this._preferenceScope = browser_1.PreferenceScope.User;
        return _this;
    }
    PreferencesContainer_1 = PreferencesContainer;
    PreferencesContainer.prototype.init = function () {
        this.id = PreferencesContainer_1.ID;
        this.title.label = 'Preferences';
        this.title.caption = this.title.label;
        this.title.closable = true;
        this.title.iconClass = 'fa fa-sliders';
        this.toDispose.push(this.onDirtyChangedEmitter);
    };
    PreferencesContainer.prototype.dispose = function () {
        if (this.isDisposed) {
            return;
        }
        _super.prototype.dispose.call(this);
        this.toDispose.dispose();
    };
    Object.defineProperty(PreferencesContainer.prototype, "autoSave", {
        get: function () {
            return this.editors.some(function (editor) { return editor.saveable.autoSave === 'on'; }) ? 'on' : 'off';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PreferencesContainer.prototype, "dirty", {
        get: function () {
            return this.editors.some(function (editor) { return editor.saveable.dirty; });
        },
        enumerable: true,
        configurable: true
    });
    PreferencesContainer.prototype.save = function () {
        this.editors.forEach(function (editor) { return editor.saveable.save(); });
    };
    PreferencesContainer.prototype.getTrackableWidgets = function () {
        return this.deferredEditors.promise;
    };
    Object.defineProperty(PreferencesContainer.prototype, "preferenceScope", {
        get: function () {
            return this._preferenceScope;
        },
        set: function (preferenceScope) {
            this._preferenceScope = preferenceScope;
        },
        enumerable: true,
        configurable: true
    });
    PreferencesContainer.prototype.onAfterAttach = function (msg) {
        return __awaiter(this, void 0, void 0, function () {
            var _a, _b, treePanel;
            var _this = this;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        if (this.widgets.length > 0) {
                            return [2 /*return*/];
                        }
                        _a = this;
                        return [4 /*yield*/, this.widgetManager.getOrCreateWidget(PreferencesTreeWidget.ID)];
                    case 1:
                        _a.treeWidget = _c.sent();
                        this.treeWidget.onPreferenceSelected(function (value) {
                            var preferenceName = Object.keys(value)[0];
                            var preferenceValue = value[preferenceName];
                            if (_this.dirty) {
                                _this.messageService.warn('Preferences editor(s) has/have unsaved changes');
                            }
                            else if (_this.currentEditor) {
                                _this.preferenceService.set(preferenceName, preferenceValue, _this.currentEditor.title.label === 'User Preferences'
                                    ? browser_1.PreferenceScope.User
                                    : browser_1.PreferenceScope.Workspace);
                            }
                        });
                        _b = this;
                        return [4 /*yield*/, this.widgetManager.getOrCreateWidget(PreferencesEditorsContainer.ID)];
                    case 2:
                        _b.editorsContainer = _c.sent();
                        this.toDispose.push(this.editorsContainer);
                        this.editorsContainer.activatePreferenceEditor(this.preferenceScope);
                        this.editorsContainer.onInit(function () {
                            algorithm_1.toArray(_this.editorsContainer.widgets()).forEach(function (editor) {
                                var editorWidget = editor;
                                _this.editors.push(editorWidget);
                                var savable = editorWidget.saveable;
                                savable.onDirtyChanged(function () {
                                    _this.onDirtyChangedEmitter.fire(undefined);
                                });
                            });
                            _this.deferredEditors.resolve(_this.editors);
                        });
                        this.editorsContainer.onEditorChanged(function (editor) {
                            if (_this.currentEditor && _this.currentEditor.editor.uri.toString() !== editor.editor.uri.toString()) {
                                _this.currentEditor.saveable.save();
                            }
                            _this.currentEditor = editor;
                        });
                        treePanel = new widgets_1.BoxPanel();
                        treePanel.addWidget(this.treeWidget);
                        this.addWidget(treePanel);
                        this.addWidget(this.editorsContainer);
                        this.treeWidget.activate();
                        _super.prototype.onAfterAttach.call(this, msg);
                        return [2 /*return*/];
                }
            });
        });
    };
    PreferencesContainer.prototype.onActivateRequest = function (msg) {
        if (this.currentEditor) {
            this.currentEditor.activate();
        }
        _super.prototype.onActivateRequest.call(this, msg);
    };
    PreferencesContainer.prototype.onCloseRequest = function (msg) {
        if (this.treeWidget) {
            this.treeWidget.close();
        }
        this.editorsContainer.close();
        _super.prototype.onCloseRequest.call(this, msg);
        this.dispose();
    };
    PreferencesContainer.prototype.activatePreferenceEditor = function (preferenceScope) {
        if (this.editorsContainer) {
            this.editorsContainer.activatePreferenceEditor(preferenceScope);
        }
    };
    var PreferencesContainer_1;
    PreferencesContainer.ID = 'preferences_container_widget';
    __decorate([
        inversify_1.inject(browser_1.WidgetManager),
        __metadata("design:type", browser_1.WidgetManager)
    ], PreferencesContainer.prototype, "widgetManager", void 0);
    __decorate([
        inversify_1.inject(browser_1.ApplicationShell),
        __metadata("design:type", browser_1.ApplicationShell)
    ], PreferencesContainer.prototype, "shell", void 0);
    __decorate([
        inversify_1.inject(core_1.MessageService),
        __metadata("design:type", core_1.MessageService)
    ], PreferencesContainer.prototype, "messageService", void 0);
    __decorate([
        inversify_1.inject(browser_1.PreferenceService),
        __metadata("design:type", Object)
    ], PreferencesContainer.prototype, "preferenceService", void 0);
    __decorate([
        inversify_1.postConstruct(),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", void 0)
    ], PreferencesContainer.prototype, "init", null);
    PreferencesContainer = PreferencesContainer_1 = __decorate([
        inversify_1.injectable()
    ], PreferencesContainer);
    return PreferencesContainer;
}(widgets_1.SplitPanel));
exports.PreferencesContainer = PreferencesContainer;
var PreferencesEditorsContainer = /** @class */ (function (_super) {
    __extends(PreferencesEditorsContainer, _super);
    function PreferencesEditorsContainer() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.onInitEmitter = new core_1.Emitter();
        _this.onInit = _this.onInitEmitter.event;
        _this.onEditorChangedEmitter = new core_1.Emitter();
        _this.onEditorChanged = _this.onEditorChangedEmitter.event;
        _this.toDispose = new core_1.DisposableCollection(_this.onEditorChangedEmitter, _this.onInitEmitter);
        return _this;
    }
    PreferencesEditorsContainer.prototype.dispose = function () {
        this.toDispose.dispose();
        _super.prototype.dispose.call(this);
    };
    PreferencesEditorsContainer.prototype.onCloseRequest = function (msg) {
        algorithm_1.toArray(this.widgets()).forEach(function (widget) { return widget.close(); });
        _super.prototype.onCloseRequest.call(this, msg);
    };
    PreferencesEditorsContainer.prototype.onUpdateRequest = function (msg) {
        this.onEditorChangedEmitter.fire(this.selectedWidgets().next());
        _super.prototype.onUpdateRequest.call(this, msg);
    };
    PreferencesEditorsContainer.prototype.onAfterAttach = function (msg) {
        return __awaiter(this, void 0, void 0, function () {
            var userPreferenceUri, userPreferences, _a, workspacePreferenceUri, workspacePreferences, _b, _c;
            return __generator(this, function (_d) {
                switch (_d.label) {
                    case 0:
                        userPreferenceUri = this.userPreferenceProvider.getUri();
                        return [4 /*yield*/, this.editorManager.getOrCreateByUri(userPreferenceUri)];
                    case 1:
                        userPreferences = _d.sent();
                        userPreferences.title.label = 'User Preferences';
                        _a = userPreferences.title;
                        return [4 /*yield*/, this.getPreferenceEditorCaption(userPreferenceUri)];
                    case 2:
                        _a.caption = _d.sent();
                        userPreferences.scope = browser_1.PreferenceScope.User;
                        this.addWidget(userPreferences);
                        return [4 /*yield*/, this.workspacePreferenceProvider.getUri()];
                    case 3:
                        workspacePreferenceUri = _d.sent();
                        _b = workspacePreferenceUri;
                        if (!_b) return [3 /*break*/, 5];
                        return [4 /*yield*/, this.editorManager.getOrCreateByUri(workspacePreferenceUri)];
                    case 4:
                        _b = (_d.sent());
                        _d.label = 5;
                    case 5:
                        workspacePreferences = _b;
                        if (!workspacePreferences) return [3 /*break*/, 7];
                        workspacePreferences.title.label = 'Workspace Preferences';
                        _c = workspacePreferences.title;
                        return [4 /*yield*/, this.getPreferenceEditorCaption(workspacePreferenceUri)];
                    case 6:
                        _c.caption = _d.sent();
                        workspacePreferences.scope = browser_1.PreferenceScope.Workspace;
                        this.addWidget(workspacePreferences);
                        _d.label = 7;
                    case 7:
                        this.activatePreferenceEditor(this.preferenceScope);
                        _super.prototype.onAfterAttach.call(this, msg);
                        this.onInitEmitter.fire(undefined);
                        return [2 /*return*/];
                }
            });
        });
    };
    PreferencesEditorsContainer.prototype.activatePreferenceEditor = function (preferenceScope) {
        var e_1, _a;
        this.preferenceScope = preferenceScope;
        try {
            for (var _b = __values(algorithm_1.toArray(this.widgets())), _c = _b.next(); !_c.done; _c = _b.next()) {
                var widget = _c.value;
                var preferenceEditor = widget;
                if (preferenceEditor.scope === preferenceScope) {
                    this.activateWidget(widget);
                    break;
                }
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
    };
    PreferencesEditorsContainer.prototype.getPreferenceEditorCaption = function (preferenceUri) {
        return __awaiter(this, void 0, void 0, function () {
            var homeStat, homeUri, uri;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.fileSystem.getCurrentUserHome()];
                    case 1:
                        homeStat = _a.sent();
                        homeUri = homeStat ? new uri_1.default(homeStat.uri) : undefined;
                        uri = preferenceUri;
                        if (preferenceUri.scheme === browser_3.UserStorageUri.SCHEME && homeUri) {
                            uri = homeUri.resolve(browser_3.THEIA_USER_STORAGE_FOLDER).resolve(preferenceUri.withoutScheme().path);
                        }
                        return [2 /*return*/, homeUri
                                ? common_1.FileSystemUtils.tildifyPath(uri.path.toString(), homeUri.withoutScheme().toString())
                                : uri.path.toString()];
                }
            });
        });
    };
    PreferencesEditorsContainer.ID = 'preferences_editors_container';
    __decorate([
        inversify_1.inject(common_1.FileSystem),
        __metadata("design:type", Object)
    ], PreferencesEditorsContainer.prototype, "fileSystem", void 0);
    __decorate([
        inversify_1.inject(browser_2.EditorManager),
        __metadata("design:type", browser_2.EditorManager)
    ], PreferencesEditorsContainer.prototype, "editorManager", void 0);
    __decorate([
        inversify_1.inject(browser_1.PreferenceProvider), inversify_1.named(browser_1.PreferenceScope.User),
        __metadata("design:type", user_preference_provider_1.UserPreferenceProvider)
    ], PreferencesEditorsContainer.prototype, "userPreferenceProvider", void 0);
    __decorate([
        inversify_1.inject(browser_1.PreferenceProvider), inversify_1.named(browser_1.PreferenceScope.Workspace),
        __metadata("design:type", workspace_preference_provider_1.WorkspacePreferenceProvider)
    ], PreferencesEditorsContainer.prototype, "workspacePreferenceProvider", void 0);
    PreferencesEditorsContainer = __decorate([
        inversify_1.injectable()
    ], PreferencesEditorsContainer);
    return PreferencesEditorsContainer;
}(widgets_1.DockPanel));
exports.PreferencesEditorsContainer = PreferencesEditorsContainer;
var PreferencesTreeWidget = /** @class */ (function (_super) {
    __extends(PreferencesTreeWidget, _super);
    function PreferencesTreeWidget(model, treeProps, contextMenuRenderer, preferenceSchemaProvider) {
        var _this = _super.call(this, treeProps, model, contextMenuRenderer) || this;
        _this.model = model;
        _this.treeProps = treeProps;
        _this.contextMenuRenderer = contextMenuRenderer;
        _this.preferenceSchemaProvider = preferenceSchemaProvider;
        _this.preferencesGroupNames = [];
        _this.onPreferenceSelectedEmitter = new core_1.Emitter();
        _this.onPreferenceSelected = _this.onPreferenceSelectedEmitter.event;
        _this.toDispose = new core_1.DisposableCollection();
        _this.toDispose.push(_this.onPreferenceSelectedEmitter);
        _this.id = PreferencesTreeWidget_1.ID;
        _this.properties = _this.preferenceSchemaProvider.getCombinedSchema().properties;
        for (var property in _this.properties) {
            if (property) {
                var group = property.substring(0, property.indexOf('.'));
                if (_this.preferencesGroupNames.indexOf(group) < 0) {
                    _this.preferencesGroupNames.push(group);
                }
            }
        }
        return _this;
    }
    PreferencesTreeWidget_1 = PreferencesTreeWidget;
    PreferencesTreeWidget.prototype.dispose = function () {
        this.toDispose.dispose();
        _super.prototype.dispose.call(this);
    };
    PreferencesTreeWidget.prototype.onAfterAttach = function (msg) {
        this.initializeModel();
        _super.prototype.onAfterAttach.call(this, msg);
    };
    PreferencesTreeWidget.prototype.handleContextMenuEvent = function (node, event) {
        _super.prototype.handleContextMenuEvent.call(this, node, event);
        if (node.expanded === undefined) {
            this.openContextMenu(node, event.nativeEvent.x, event.nativeEvent.y);
        }
    };
    PreferencesTreeWidget.prototype.handleClickEvent = function (node, event) {
        _super.prototype.handleClickEvent.call(this, node, event);
        if (node.expanded === undefined) {
            this.openContextMenu(node, event.nativeEvent.x, event.nativeEvent.y);
        }
    };
    PreferencesTreeWidget.prototype.handleEnter = function (event) {
        _super.prototype.handleEnter.call(this, event);
        var node = this.model.selectedNodes[0];
        if (node.expanded === undefined) {
            if (node) {
                var nodeElement = document.getElementById(node.id);
                if (nodeElement) {
                    var position = nodeElement.getBoundingClientRect();
                    this.openContextMenu(this.model.selectedNodes[0], position.left, position.bottom);
                }
            }
        }
    };
    PreferencesTreeWidget.prototype.openContextMenu = function (node, positionX, positionY) {
        var _this = this;
        if (node && browser_1.SelectableTreeNode.is(node)) {
            var contextMenu = this.preferencesMenuFactory.createPreferenceContextMenu(node.id, this.preferenceService.get(node.id), this.properties[node.id], function (property, value) {
                var _a;
                _this.onPreferenceSelectedEmitter.fire((_a = {}, _a[property] = value, _a));
            });
            contextMenu.aboutToClose.connect(function () {
                _this.activate();
            });
            contextMenu.activeItem = contextMenu.items[0];
            contextMenu.open(positionX, positionY);
        }
    };
    PreferencesTreeWidget.prototype.initializeModel = function () {
        var _this = this;
        var e_2, _a;
        var preferencesGroups = [];
        var root = {
            id: 'root-node-id',
            name: 'Apply the preference to selected preferences file',
            parent: undefined,
            visible: true,
            children: preferencesGroups,
            expanded: true,
        };
        var nodes = [];
        var _loop_1 = function (group) {
            var propertyNodes = [];
            var properties = [];
            for (var property in this_1.properties) {
                if (property.split('.', 1)[0] === group) {
                    properties.push(property);
                }
            }
            var preferencesGroup = {
                id: group + '-id',
                name: group.toLocaleUpperCase().substring(0, 1) + group.substring(1) + ' (' + properties.length + ')',
                visible: true,
                parent: root,
                children: propertyNodes,
                expanded: false,
                selected: false
            };
            properties.sort(function (a, b) { return a.localeCompare(b); });
            properties.forEach(function (property) {
                var _a;
                var node = {
                    id: property,
                    name: property.substring(property.indexOf('.') + 1),
                    parent: preferencesGroup,
                    visible: true,
                    selected: false
                };
                propertyNodes.push(node);
                nodes.push((_a = {}, _a[property] = _this.properties[property], _a));
            });
            preferencesGroups.push(preferencesGroup);
        };
        var this_1 = this;
        try {
            for (var _b = __values(this.preferencesGroupNames.sort(function (a, b) { return a.localeCompare(b); })), _c = _b.next(); !_c.done; _c = _b.next()) {
                var group = _c.value;
                _loop_1(group);
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_2) throw e_2.error; }
        }
        this.decorator.fireDidChangeDecorations(nodes);
        this.model.root = root;
    };
    var PreferencesTreeWidget_1;
    PreferencesTreeWidget.ID = 'preferences_tree_widget';
    __decorate([
        inversify_1.inject(preferences_menu_factory_1.PreferencesMenuFactory),
        __metadata("design:type", preferences_menu_factory_1.PreferencesMenuFactory)
    ], PreferencesTreeWidget.prototype, "preferencesMenuFactory", void 0);
    __decorate([
        inversify_1.inject(browser_1.PreferenceService),
        __metadata("design:type", Object)
    ], PreferencesTreeWidget.prototype, "preferenceService", void 0);
    __decorate([
        inversify_1.inject(preferences_decorator_1.PreferencesDecorator),
        __metadata("design:type", preferences_decorator_1.PreferencesDecorator)
    ], PreferencesTreeWidget.prototype, "decorator", void 0);
    PreferencesTreeWidget = PreferencesTreeWidget_1 = __decorate([
        inversify_1.injectable(),
        __param(0, inversify_1.inject(browser_1.TreeModel)),
        __param(1, inversify_1.inject(browser_1.TreeProps)),
        __param(2, inversify_1.inject(browser_1.ContextMenuRenderer)),
        __param(3, inversify_1.inject(browser_1.PreferenceSchemaProvider)),
        __metadata("design:paramtypes", [Object, Object, Object, browser_1.PreferenceSchemaProvider])
    ], PreferencesTreeWidget);
    return PreferencesTreeWidget;
}(browser_1.TreeWidget));
exports.PreferencesTreeWidget = PreferencesTreeWidget;
