// @theia/plugin-ext/lib/hosted/browser/worker/worker-main.js
"use strict";
/********************************************************************************
 * Copyright (C) 2018 Red Hat, Inc. and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
var __values = (this && this.__values) || function (o) {
    var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
    if (m) return m.call(o);
    return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
};
Object.defineProperty(exports, "__esModule", { value: true });
var event_1 = require("@theia/core/lib/common/event");
var rpc_protocol_1 = require("../../../api/rpc-protocol");
var plugin_manager_1 = require("../../../plugin/plugin-manager");
var plugin_api_1 = require("../../../api/plugin-api");
var plugin_context_1 = require("../../../plugin/plugin-context");
var plugin_protocol_1 = require("../../../common/plugin-protocol");
var env_1 = require("../../../plugin/env");
var preference_registry_1 = require("../../../plugin/preference-registry");
var debug_stub_1 = require("./debug-stub");
// tslint:disable-next-line:no-any
var ctx = self;
var pluginsApiImpl = new Map();
var pluginsModulesNames = new Map();
var emitter = new event_1.Emitter();
var rpc = new rpc_protocol_1.RPCProtocolImpl({
    onMessage: emitter.event,
    send: function (m) {
        ctx.postMessage(m);
    }
});
// tslint:disable-next-line:no-any
addEventListener('message', function (message) {
    emitter.fire(message.data);
});
function initialize(contextPath, pluginMetadata) {
    ctx.importScripts('/context/' + contextPath);
}
var envExt = new env_1.EnvExtImpl(rpc);
var preferenceRegistryExt = new preference_registry_1.PreferenceRegistryExtImpl(rpc);
var debugExt = debug_stub_1.createDebugExtStub(rpc);
var pluginManager = new plugin_manager_1.PluginManagerExtImpl({
    // tslint:disable-next-line:no-any
    loadPlugin: function (plugin) {
        if (isElectron()) {
            ctx.importScripts(plugin.pluginPath);
        }
        else {
            ctx.importScripts('/hostedPlugin/' + plugin_protocol_1.getPluginId(plugin.model) + '/' + plugin.pluginPath);
        }
        if (plugin.lifecycle.frontendModuleName) {
            if (!ctx[plugin.lifecycle.frontendModuleName]) {
                console.error("WebWorker: Cannot start plugin \"" + plugin.model.name + "\". Frontend plugin not found: \"" + plugin.lifecycle.frontendModuleName + "\"");
                return;
            }
            return ctx[plugin.lifecycle.frontendModuleName];
        }
    },
    init: function (rawPluginData) {
        var e_1, _a;
        var result = [];
        var foreign = [];
        try {
            for (var rawPluginData_1 = __values(rawPluginData), rawPluginData_1_1 = rawPluginData_1.next(); !rawPluginData_1_1.done; rawPluginData_1_1 = rawPluginData_1.next()) {
                var plg = rawPluginData_1_1.value;
                var pluginModel = plg.model;
                var pluginLifecycle = plg.lifecycle;
                if (pluginModel.entryPoint.frontend) {
                    var frontendInitPath = pluginLifecycle.frontendInitPath;
                    if (frontendInitPath) {
                        initialize(frontendInitPath, plg);
                    }
                    else {
                        frontendInitPath = '';
                    }
                    var plugin = {
                        pluginPath: pluginModel.entryPoint.frontend,
                        pluginFolder: plg.source.packagePath,
                        model: pluginModel,
                        lifecycle: pluginLifecycle,
                        rawModel: plg.source
                    };
                    result.push(plugin);
                    var apiImpl = apiFactory(plugin);
                    pluginsApiImpl.set(plugin.model.id, apiImpl);
                    pluginsModulesNames.set(plugin.lifecycle.frontendModuleName, plugin);
                }
                else {
                    foreign.push({
                        pluginPath: pluginModel.entryPoint.backend,
                        pluginFolder: plg.source.packagePath,
                        model: pluginModel,
                        lifecycle: pluginLifecycle,
                        rawModel: plg.source
                    });
                }
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (rawPluginData_1_1 && !rawPluginData_1_1.done && (_a = rawPluginData_1.return)) _a.call(rawPluginData_1);
            }
            finally { if (e_1) throw e_1.error; }
        }
        return [result, foreign];
    },
    initExtApi: function (extApi) {
        var e_2, _a;
        try {
            for (var extApi_1 = __values(extApi), extApi_1_1 = extApi_1.next(); !extApi_1_1.done; extApi_1_1 = extApi_1.next()) {
                var api = extApi_1_1.value;
                try {
                    if (api.frontendExtApi) {
                        ctx.importScripts(api.frontendExtApi.initPath);
                        ctx[api.frontendExtApi.initVariable][api.frontendExtApi.initFunction](rpc, pluginsModulesNames);
                    }
                }
                catch (e) {
                    console.error(e);
                }
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (extApi_1_1 && !extApi_1_1.done && (_a = extApi_1.return)) _a.call(extApi_1);
            }
            finally { if (e_2) throw e_2.error; }
        }
    }
}, envExt, preferenceRegistryExt, rpc);
var apiFactory = plugin_context_1.createAPIFactory(rpc, pluginManager, envExt, debugExt, preferenceRegistryExt);
var defaultApi;
var handler = {
    // tslint:disable-next-line:no-any
    get: function (target, name) {
        var plugin = pluginsModulesNames.get(name);
        if (plugin) {
            var apiImpl = pluginsApiImpl.get(plugin.model.id);
            return apiImpl;
        }
        if (!defaultApi) {
            defaultApi = apiFactory(plugin_api_1.emptyPlugin);
        }
        return defaultApi;
    }
};
// tslint:disable-next-line:no-null-keyword
ctx['theia'] = new Proxy(Object.create(null), handler);
rpc.set(plugin_api_1.MAIN_RPC_CONTEXT.HOSTED_PLUGIN_MANAGER_EXT, pluginManager);
rpc.set(plugin_api_1.MAIN_RPC_CONTEXT.PREFERENCE_REGISTRY_EXT, preferenceRegistryExt);
function isElectron() {
    if (typeof navigator === 'object' && typeof navigator.userAgent === 'string' && navigator.userAgent.indexOf('Electron') >= 0) {
        return true;
    }
    return false;
}
